--- 
title: "Quantitative Methods 2, ZHAW"
author: "JÃ¼rgen Degenfellner"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
github-repo: rstudio/bookdown-demo
description: "This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook."
---

# Preamble {#intro0}

This script is a continuation of 
[the first one](https://jdegenfellner.github.io/Script_QM1_ZHAW/) for Quantitative Methods 1 
at ZHAW. 

In the first part, we learned about the basics of probability theory,
descriptive statistics, Bayesian statistics, and hypothesis testing.

In this script, we will dive into the basics of statistical modeling - 
a world of aesthetic wonder and surprises.

This script is a first draft as you are the first group to be working with it.

Please feel free to send me suggestions for improvements or corrections.

This **should be a collaborative effort** and will (hopefully) 
never be finished as our insight grows over time.

The script can also be seen as a pointer to great sources which are
suited to deepen your understanding of the topics. Knowledge is decentralized,
and there are many great resources out there.

For the working setup with R, please 
see [this](https://jdegenfellner.github.io/Script_QM1_ZHAW/index.html#section) 
and the following sections in the first script.

The complete code for this script can be found 
[here](https://github.com/jdegenfellner/Script_QM2_ZHAW).

## Books we will heavily borrow from are:

- (older online version is Free; current version in ZHAW Library) [Statistical Rethinking](https://civil.colorado.edu/~balajir/CVEN6833/bayes-resources/RM-StatRethink-Bayes.pdf), YouTube-Playlist: [Statistical Rethinking 2023](https://youtu.be/FdnMWdICdRs?si=q2py5QVY_L299hEa)
- (Free) [Understanding Regression Analysis: A Conditional Distribution Approach](https://vdoc.pub/documents/understanding-regression-analysis-a-conditional-distribution-approach-84oqjr8sqva0)
- (Free online-access via ZHAW Library) [Data Analysis Using Regression and Multilevel/Hierarchical Models](http://www.stat.columbia.edu/~gelman/arm/)
- (Free) [Doing Bayesian Data Analysis](https://nyu-cdsc.github.io/learningr/assets/kruschke_bayesian_in_R.pdf)

```{r, echo=FALSE}
library(gridExtra)
library(ggplot2)
library(png)
library(grid)

# Load images (ensure they are in the "images/" subfolder)
img1 <- rasterGrob(readPNG("./images/Cover_Stat_Reth.png"), interpolate=TRUE)
img2 <- rasterGrob(readPNG("./images/Cover_Westfall.png"), interpolate=TRUE)
img3 <- rasterGrob(readPNG("./images/Cover_Gelman.png"), interpolate=TRUE)
img4 <- rasterGrob(readPNG("./images/Cover_Kruschke.png"), interpolate=TRUE)

# Create plots with labels
p1 <- ggplot() + annotation_custom(img1) + theme_void()
p2 <- ggplot() + annotation_custom(img2) + theme_void()
p3 <- ggplot() + annotation_custom(img3) + theme_void()
p4 <- ggplot() + annotation_custom(img4) + theme_void()

# Arrange in 2x2 grid
grid.arrange(p1, p2, p3, p4, ncol=2)
```


## If you need a good reason to buy great books...

Think of the total costs of your education. You want to extract maximum 
benefit from it. In the US, an education costs [a lot](https://www.finaid.ucsb.edu/docs/default-source/default-document-library/2024-2025-undergrad-coa.pdf?sfvrsn=fb791f8a_4).
In beautiful Switzerland, the [tuition fees](https://www.zhaw.ch/de/sml/studium/master/kosten) (if applicable) are nowhere near these figures.
Costs you could consider are opportunity costs of not working.
A comparison with both, a foreign education or opportunity costs,
justifies the investment in good books.
Or: The costs of all the good books of your education combined are probably less than an iPhone Pro.

<!--chapter:end:index.Rmd-->

---
output:
  pdf_document: default
  html_document: default
---
# Introduction {#intro}

## What is statistical modeling and what do we need this for?

Typically, one simplifies the complex reality (and loses information) 
in order to make it better
**understandable** (explainable), mathematically treatable and to make **predictions**. 

Underlying our models, there are theories which should be [falsifiable](https://en.wikipedia.org/wiki/Falsifiability)
and testable.
For instance, I would be really surprised if I pull up my multimeter and measure the voltage (V) and
electric current (I) at a resistence (R) in a circuit and find that [Ohm's law](https://en.wikipedia.org/wiki/Ohm%27s_law) $V = IR$ is not true. 
This [**law**](https://en.wikipedia.org/wiki/Scientific_law)
can be tested over and over again and if one would find a single valid counterexample, 
the law would be falsified. It is also true that the law is probably not 100% accularate,
but an extremely precise approximation of reality. Real-world measurements carry 
measurement errors and when plotting the data, one would see that the data points
might not lie exactly on a straight line. This is not a problem.

A [statistical model](https://en.wikipedia.org/wiki/Statistical_model) 
is a mathematical framework that represents the 
relationships between variables, helping us understand, infer, and
predict patterns in data. It acts as a bridge between observed data 
and the real-world processes that generated them. In health research, 
where variability and uncertainty are inherent, statistical models are 
valuable tools for making sense of complex phenomena. 
You can watch [this](https://www.youtube.com/watch?v=3d5ivs_8amQ&ab_channel=VeryNormal) as short intro.

In [QM1](https://jdegenfellner.github.io/Script_QM1_ZHAW/probs.html) we have already 
made testable predictions with respect to the probability of an event. 
In our 1000-researcher experiment we stated for instance, that the probability of
obeserving $66$ or more findings would be very unlikely. If such an event would 
occur (while not repeating the experiment many times), we would reconsider our model.
Inexactly, we could have stated something like: "We will not see more than 100 findings by chance."
With respect to our multiple choice test at the end of QM1 we could predict:
"We will not see a single person answering all questions correctly by chance in our 
lifetime (given the frequency of tests)." Note, that in this context, the word **predict**
is used with respect to a future event (chance finding or chance passing of the test).
As we will see, there does not necessarily have to be a temporal connection in order 
to *predict* something.

Depending on the task at hand, we would use different models.
In any case, logical reasoning and critical thinking comes first, 
then comes the model. **It makes no sense to estimate statistical models just for the sake of it**.

**All models are wrong, but some are useful**. 
Or to quote [George Box](https://www.tandfonline.com/doi/abs/10.1080/01621459.1976.10480949):

> "Since all models are wrong the scientist cannot obtain
> a 'correct' one by excessive elaboration. On the contrary
> following William of Occam he should seek an economical
> description of natural phenomena. Just as the ability to
> devise simple but evocative models is the signature of the
> great scientist so overelaboration and overparameterization is often the mark of mediocrity."

In my opinion, statistical modeling is an art form: difficult and beautiful.

**One goal of this course** is to improve interpretation and limitations of statistical models.
They are not magical tools turning data into truth. Firstly, the rule gargabe in, garbage out (GABA) applies.
Secondly, statistical models are based on data and their variability and have inherent limitations
one cannot overcome even with the most sophisticated models. This is expressed for instance 
in the so-called [bias-variance trade-off](https://en.wikipedia.org/wiki/Bias%E2%80%93variance_tradeoff).
You can't have it all.

### Explanatory vs. Predictive Models

I can recommend reading [this](https://projecteuclid.org/journals/statistical-science/volume-25/issue-3/To-Explain-or-to-Predict/10.1214/10-STS330.full) 
article by Shmueli et al. (2010) on this topic.

Statistical models serve different purposes depending on the research question. Two primary goals are **explanation** 
and **prediction**, and each requires a different approach:

**Explanatory models (the harder of the two)** focus on understanding causal 
relationships. 
These models aim to uncover mechanisms and answer **"why"** 
questions. For example:
   
   - Does smoking increase the risk of lung cancer? **Yes**. (If you want to see what a large effect-size looks like, check out [this study](https://bmjopen.bmj.com/content/bmjopen/8/10/e021611.full.pdf).)
   - How large is the **effect** (causal) of smoking on lung cancer? **Large**.
   - Does pain education and graded sensorimotor relearning improve disability (a question we ask in 
   our [Resolve Swiss project](https://data.snf.ch/grants/grant/220585))?

Explanatory models are **theory-driven**, designed to test hypotheses. Here, one wants to understand the underlying
mechanisms and the relationships between variables and hence often uses (parsimonious) models that are more interpretable, 
like linear regression.

**Predictive models** prioritize forecasting/predicting (future) outcomes based on patterns in the data. 
These models aim to answer **"what will happen?"** For instance:
   
   - [Gait analysis](https://www.tandfonline.com/doi/abs/10.1080/03091902.2020.1822940) using Machine Learning (ML)?
   - [Skin cancer detection](https://jamanetwork.com/journals/jamadermatology/fullarticle/2756346) using neural networks?
   - If I know the age, sex and weight of a person, can I predict his/her height? 
     Can I predict the height better with the given covariates compared to just 
     guessing by using the mean height of my sample for the next patient?

Predictive models are **data-driven**, often using complex algorithms to achieve high accuracy. 
Their success is measured using metrics like [Root Means Square Error](https://computersciencewiki.org/index.php/Root-mean-square_error_(RMSE)) 
(RMSE), [Area Unter the Curve](https://en.wikipedia.org/wiki/Receiver_operating_characteristic#:~:text=The%20area%20under%20the%20curve,ranks%20higher%20than%20'negative') 
(AUC), or **prediction error on new, unseen data**.
Any amount of model complexity is allowed. One could for instance estimate a 
[neural network](https://en.wikipedia.org/wiki/Neural_network_(machine_learning)) ("just" another statistical model) 
with many hidden layers and neurons in order to improve prediction quality. Interpretability of the model weights is not a priority here.

While explanatory and predictive goals often complement each other, 
their differences highlight the importance of clearly defining the purpose 
of your analysis. In applied health research, explanatory models help identify 
causal mechanisms, while predictive models can guide real-world decisions by 
providing actionable forecasts. Together, they enhance both our understanding 
of phenomena and our ability to make informed decisions in complex environments.

### Individual vs. Population Prediction

Another important distinction is between **individual vs. population** prediction.
In the smoking example above, we can be very sure about the mean effects that smoking has on lung cancer.
On an individual level, it is [harder to predict the outcome](https://www.liebertpub.com/doi/10.1089/rej.2019.2298). 
Nevertheless, individual predictions will be (notably) better than random guessing. We will discuss this in greater detail.

Statistical models are often much worse than one would naively expect, but
they very often better than experts. If you are interested and want to
boost your confidence in the predictive ability of statistical models,
I recommend reading chapter 21 ("Intuitions vs. Formulas") of Daniel Kahneman's book 
["Thinking, Fast and Slow"](https://en.wikipedia.org/wiki/Thinking,_Fast_and_Slow) 
(available in the ZHAW library).

### Practical Use of Statistical Models

In my optinion, we should never be afraid to test our statistical models (as honestly as possible) against reality.
We could for instance ask ourselves: 

- "How much better does this model estimate an outcome than the arithmetic mean? 
  (i.e., the linear model with just an intercept)"

- "How much better does this model classify than random guessing?"

- Is it worth the effort to collect data and estimate this model by using hundreds of hours of our time?

In some cases, these questions can be answered straightforwardly. 

- In advertising (Google, Facebook, ...), a couple of percentage points in prediction quality might make a difference of millions 
of dollars in revenue offsetting the statistitians salary by a large margin. 

- Improved forecasts of a few percentage points in the stock market or just being [slightly better](https://en.wikipedia.org/wiki/Jim_Simons) 
than the average, will make you faboulously rich.

- Improved cancer forecasting might save lives, money and pain and is of course 
  not only measured in financial gains.

### Start at the beginning

What do we actually want to do in general? Very broadly speaking we want to:
**describe** the association of variables to each other that carry variability. 
Hence, the relationship is not deterministic like $$y = 2x + 3$$ but rather we need
to "loosen up" the relationship to account for variability (in $x$ and $y$).
So, $y$ and $x$ are not fixed but aflicted with uncertainty.
Depending on your philosophical view, you might say you want to find
the "true" but unknown relationship (here, $2$ and $3$ are the true coefficients) between variables.
This is what we do in simulation studies all the time: We know the true relationship, 
simulate data by adding variability and then try to estimate the true relationship we assumed in the first place.
This is an **advantage** the pioneers of statistics did not have. We can simulate
millions of lines of data at the click of a button. 
For some practical applications, we can get a really nice and complete answer to our question
(for instance sample size for proportions).

So we are looking for a function $f$ such that

$$ \mathbf{Y} = f(\mathbf{X}) $$

where 

- $\mathbf{Y}$ is the "outcome", "dependent variable" or "response". 
- $\mathbf{X}$ are the "predictors". $\mathbf{X}$ can be a single Variable $x$ or many 
variables $x_1, x_2, \ldots, x_p$.

It is important to be aware of the notation here:
"Predict" does **not necessarily** mean that we can predict the value in 
the future. It merely means we estimate the value (or mean) of $Y$ given $X$.

- This can be done at the same time points, known as **cross-sectional** analysis ("What is the maximum jumping height 
of a person given their age at a certain point in time, whereas both variables are measured at the same time?");
- or at different time points, known as **longitudinal analysis** ("What is the maximum jumping height of a person 10 years later ($t_2$)
 given their baseline health status at time $t_1$?").

The **simplest statistical model** would be the **mean model** where $Y$ is "predicted" by a 
constant: $Y = c$ which (at least in the classical linear regression) turns out to be $c = \bar{x}$. 
This simple model is often surprisingly good, or, to put it in other words, models with more complexity
are often not that much better.

## A (simple) model for adult body heights in the Bayesian framework

As repetition, read the parts about [Bayes statistics from QM1](https://jdegenfellner.github.io/Script_QM1_ZHAW/bayes_statistics.html)
again to refresh your memory about the Bayesian framework.

It's recommendable to read the beginning of the book [Statistical rethinking](https://civil.colorado.edu/~balajir/CVEN6833/bayes-resources/RM-StatRethink-Bayes.pdf)
(hint: the online-version of the book differs a bit from the paper-version) 
up until page 39 as well. We are not completely new to the topic of Bayes thanks 
to QM1. In the Bayesian setting we use (well argued for) prior knowledge about a 
parameter or effect and update this knowledge with new data.

We want to **start building our first model** right away.

Let's begin with the example in 
[Statistical rethinking](https://civil.colorado.edu/~balajir/CVEN6833/bayes-resources/RM-StatRethink-Bayes.pdf) 
using data from the [!Kung San](https://en.wikipedia.org/wiki/%C7%83Kung_people) 
people.

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
library(rethinking)
data("Howell1")
d <- Howell1
str(d)
d2 <- d[d$age >= 18, ] # only adults
```
We want to model the adult height of the !Kun San people 
using prior knowledge (about the Swiss population) and data.

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
library(tidyverse)
d2 %>% ggplot(aes(x = height)) + geom_histogram()
```

Since we already have domain knowledge in this area, we can say that heights are usually normally distributed,
or at least a mixture of normal distrubutions (female/male).
We assume the following model:
$$h_i \sim \text{Normal}(\mu, \sigma)$$

As in [QM1](https://jdegenfellner.github.io/Script_QM1_ZHAW/), 
we want to start with a Bayesian model and hence, we need some priors.

Since we are in Switzerland and just for fun, we use the mean of
[Swiss body heights](https://www.bfs.admin.ch/asset/de/30305714) 
as expected value for the **prior for the mean**. 
According to the link (Bundesamt fÃ¼r Statistik), 
the mean height of $n=21,873$ people in the Swiss sample is
$171.1$ cm. We choose the same $\sigma$ for the prior of the normal
as in the book not to deviate too much from the example at hand.

Next comes our **model definition in the Bayesian framework**, which
I often find more intuitive than the Frequentist approach:

\[
h_i \sim \text{Normal}(\mu, \sigma)
\]
\[
\mu \sim \text{Normal}(171.1, 20)
\]
\[
\sigma \sim \text{Uniform}(0, 50)
\]

**Description of the model definition**: The heights are normally distributed with unknown mean and 
standard deviation. As our current knowledge about the mean height, we use 
a prior distribution for the mean (we do not know but want to estimate) by 
assuming the mean of a population we know and a standard deviation of $20$ cm which
allows are rather large range of possible values for $\mu$ 
(the unobserved population mean of the !Kung San people).
$\sigma$ (the unobserved standard deviation of the population of !Kun San people) 
is also unknown and a priori we restrict ourselves to values 
between $0$ and $50$ cm, whereas we assign equal plausibility to all 
values in this range (which can and should be critically discussed).

**Vizualisation of the model structure**:
```{r, echo=FALSE, , message=FALSE, warning=FALSE}
library(DiagrammeR)
grViz("
digraph BayesianModel {
  graph [layout = dot]
  
  h_i [label = 'h[i]\\n~ Normal(Î¼, Ï)', shape = oval]
  Î¼ [label = 'Î¼\\n~ Normal(171.1, 20)', shape = oval]
  Ï [label = 'Ï\\n~ Uniform(0, 50)', shape = oval]

  Î¼ -> h_i
  Ï -> h_i
}
")
```

Mind that there is a **conceptual difference** between the normal distribution of the heights 
and the normal prior distribution of the mean. The latter expresses our prior knowledge/insecurity
about the unobserved mean of the normal distribution of the heights. The normal distribution of the heights says we expect the heights to be normally distributed
but we do not know the parameters ($\mu$ and $\sigma$) yet. We will 
estimate these parameters using prior knowledge and the data.

Of course we would not need the prior here due to the large sample size, 
but let's do it anyways for demonstration purposes.
We are not completely uninformed about body heights and express our 
knowledge with the prior for $\mu$.
The $20$ in the prior for the mean expresses our range of possible true 
mean values and aknowledge
that there are a variety of different subpopulations with different means.

Using the Swiss data in the [link](https://www.bfs.admin.ch/asset/de/30305714) 
one could estimate that the standard deviation of the heights 
from $21,873$ Swiss people is around is $25.67$ cm ([Exercise 1](#exercise1_Intro)).

Remember, in the Baysian world, there is no **fixed but unknown**
parameter, but instead we define a distribution over the unobserved parameter.

We **visualize the prior for $\mu$**:

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
curve(dnorm(x, 171.1, 20), from = 100, to = 250)
```

A wide range of population means is possible. Once could discuss this distribution 
and maybe further restrict it.

The **prior for $\sigma$** is uniform between $0$ and $50$ cm. This is a very wide prior and
just constrains the values to be positive and below $50$ cm. 
This could be stronger of course.

**Visualization of the prior for $\sigma$**:

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
curve(dunif(x, 0, 50), from = -10, to = 60)
```

Note, we didnât specify a prior probability distribution of heights
directly, but once weâve chosen priors for $\mu$ and $\sigma$, these imply a 
prior distribution of individual heights. 

**Without** even having seen the **new data**, we can check what our prior
(model) for heights would predict. This is important. If the prior already 
predicts impossible values, we should reconsider our priors and/or model.

So, we simply draw $\mu$ and $\sigma$ from the priors and then draw heights 
from the normal distribution using the drawn parameters.

**Vizualisation of the prior for heights**:

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
sample_mu <- rnorm(10^4, 171.1, 20)
sample_sigma <- runif(10^4, 0, 50)
prior_h <- rnorm(10^4, sample_mu, sample_sigma)
length(prior_h)
dens(prior_h)
```

The prior is not itself a Gaussian distribution, but a distribution of
relative plausibilities of different heights, **before** seeing the data.

Note how we have created the model predictions for heights. 
We first drew $\mu$ and $\sigma$ independently 
(there is no arrow between $\mu$ and $\sigma$) from the priors.
Then we drew heights from the normal distribution using the drawn parameters.
You just follow the model structure.

Now, there are a couple of different ways to estimtate the model incorporating 
the new data. For didactic reasons, grid approximation is often used (as in the books).
For many parameters, this approach becomes more and more infeasible (due to combinatorial explosion). 

We will skip that for now and use quadratic approximation instead which
works well for many common procedures in applied statistics (like linear regression).
Later, you'll probably use (or the software in the background) mostly Markov 
chain Monte Carlo (MCMC) sampling to get the posterior. 
[Pages 39 and the following](https://civil.colorado.edu/~balajir/CVEN6833/bayes-resources/RM-StatRethink-Bayes.pdf) 
explain the 3 concepts grid approximation, quadratic approximation and MCMC.

In short, **quadratic approximation** assumes that our posterior distribution 
of body heights can be approximated well by a normal distribution, 
at least near the peak. 

Please read the [addendum](#bivariate_normal) to get a clearer picture of 
what a bivariate normal distribution is.

Using the `rethinking` package we can estimate the model using quadratic approximation.
First, we define the model in the `rethinking` syntax (see R code 4.25 in the book).

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
library(rethinking)
flist <- alist(
  height ~ dnorm(mu, sigma),
  mu ~ dnorm(171.1, 20),
  sigma ~ dunif(0, 50)
)
```

Then we estimate/fit the model using quadratic approximation.

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
m_heights <- quap(flist, data = d2)
```

Now let's take a look at the fitted model:
(Note: In the online-version of the book, they used the command `map` instead of `quap`.)

The `precis`function displays concise parameter estimate information 
(from the posterior) for an existing model fit.

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
precis(m_heights)
```
Above, we see the mean of the posterior for $\mu$ **and** $\sigma$;
and a **89% credible interval** for those parameters.
Note that these are rather tight credible intervals. We are rather confident that the mean is somewhere between
$154$ and $155$ cm and the standard deviation is between $7$ and $8$ cm.

We can now plot the posterior distribution of the mean ($\mu$) and the standard 
deviation ($\sigma$) separately by drawing from the posterior distribution.

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
post <- extract.samples(m_heights, n = 10^4)
head(post)
dens(post$mu)
dens(post$sigma)
```

Note, that **these samples come from a multi-dimensional posterior distribution**.
In our case, we approximated the **joint** posterior distribution of $\mu$ *and* $\sigma$ with a 
[bivariate normal distribution](https://en.wikipedia.org/wiki/Multivariate_normal_distribution).
They are not necessarily independent from each other, but in this case they are (see [exercise 6](#exercise6_Intro)).
We know this from the prior definition above. $\mu$ and $\sigma$ are both
defined as normal respectively uniform distributions and by definition do not
influence each other. This is also visible in the vizualisation of the model structure: 
There is no confounding variable or connection between those priors. One could
think of a common variable $Z$ that influences both $\mu$ and $\sigma$. This could
be genetic similarity which could influence both $\mu$ and $\sigma$.

Let's verify that $\mu$ and $\sigma$ are uncorrelated:

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
vcov(m_heights)
```
gives you the variance-covariance matrix of the parameters of the posterior
distribution. In the diagonal you see the variance of the parameters.

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
diag(vcov(m_heights))
```
And we can compute the correlation matrix easily:

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
cov2cor(vcov(m_heights))
```

Let's plot the posterior in 3D, because we **can**:

```{r posterior-3d-correct, echo=FALSE, message=FALSE, warning=FALSE}
library(plotly)
library(MASS)  # For kernel density estimation

# Extract posterior samples
posterior_samples <- extract.samples(m_heights)

# Calculate bivariate density using kde2d
density_est <- kde2d(posterior_samples$mu, posterior_samples$sigma, n = 100)

# Create a 3D surface plot
plot_ly(
  x = density_est$x,
  y = density_est$y,
  z = density_est$z,
  type = "surface",
  colorscale = "Viridis"
) %>%
  layout(
    title = "3D Posterior Density of Î¼ and Ï",
    scene = list(
      xaxis = list(title = "Î¼ (mu)"),
      yaxis = list(title = "Ï (sigma)"),
      zaxis = list(title = "Density")
    )
  )
```

**How beautiful ist that?**

This shows how credible each combination of $\mu$ and $\sigma$ is based on our priors and the data observed.
The higher the mountain for a certain parameter combination, the more credible this combination is.

We see in the 3D plot, that the "mountain" is not rotated, indicating 
graphically that the parameters are independent from each other.

We also see in the correlation matrix, the correlation of the parameters is $\sim 0$.
In the [context of a joint normal distribution](https://en.wikipedia.org/wiki/Correlation#Correlation_and_independence), 
this means that the parameters are also independent. 

And, it is not an accident that the posterior looks like this. Using quadratic approximation,
we used the bivariate normal distribution to **approximate** the posterior.

## Classical approach for the simplest model {#classical_simple_mean_model}

We have seen, how we could use prior knowledge to fit a very simple model
for body heights of a population (!Kung San) in the Bayesian framework.

Now, let's start at the same point in the classical framework.
Here, we do not use any prior knowledge, at least not that explicitely.

The classical approach to fit a regression line is the so-called 
**[least squares method](https://en.wikipedia.org/wiki/Least_squares)**.

There are hundreds of [videos online](https://www.youtube.com/results?search_query=linear+regression) 
explaining this method in great detail
with animations. Maybe watch these videos later, when we add a predictor to the mean model,
since most of instructional videos start at the simple linear regression using 
two parameters (intercept ($\beta_0$ or $\alpha$) and slope ($\beta_1$)).

The **(simple mean-) model** is:

$$ Y_i = height_i = c + \varepsilon_i $$

- for some $c \in \mathbb{R}$ and
- normally distributed errors $\varepsilon_i \sim \text{Normal}(0, \sigma)$.

The errors $\varepsilon_i$ are on average zero and have a constant standard deviation of $\sigma$.
So, we assume there is a fixed, but unknown, constant $c$ that we want to estimate and we 
assume that there is a special sort of error in our model that is normally distributed.
Sometimes there is a large deviation from the true $c$, sometimes there is a small deviation.
On average, the deviations are zero and the **errors should also be independent from each other**:
$$ \varepsilon_i \perp \varepsilon_j \text{ for } i \neq j$$
This means that just because I have just observed a large deviation from the true $c$
does not mean, that the probability of a large deviation in the next observation is higher/lower.
Note, that we cannot readily define different types of errors in the classical framework.

But what is $c$? We determine the shape of the model 
ourselves (constant model, or mean model) and then estimate the parameter $c$.
By defining the shape of the model ourselves and imposing a distribution where we want to 
estimate the parameter of said distribution, we are in **parametric statistics**.

We choose the $c$ which minimizes the sum of squared errors from the actual heights.
This has the advantage that deviations upper and lower from the actual height are
equally weighted. The larger the deviation the (quadratically) larger the penalty.

**Why do we do that?** Because, if the model assumptions (more on that later) 
are correct, the least squares
estimator is a really good estimator. How good? Later...

We want to miminize the following function:

$$ SSE \text{ (Sum of Squared Errors) }(c) = (height_1 - c)^2 + (height_2 - c)^2 + 
\ldots + (height_n - c)^2 =$$
$$ = \sum_{i=1}^{n} (height_i - c)^2$$

The SSE is a function of $c$ and we want to find the $c$ that minimizes the function.
Since it is a quadratic function, we can always find the minimum.
We have learnt in school how to do this (hopefully): Take the derivative of the function
and set it to zero. Solve for $c$ and you have the $c$ which yields the minimum of SSE(c).

Let's do that:

$$ \frac{d}{dc} SSE(c) =  2(height_1 - c)(-1) + 2(height_2 - c)(-1) + 
\ldots + 2(height_n - c)(-1) =$$
$$ = -2 \sum_{i=1}^{n} (height_i - c)$$

This should be zero for the minimum:
$$ -2 \sum_{i=1}^{n} (height_i - c) = 0$$
$$ \sum_{i=1}^{n} (height_i - c) = 0$$
$$ \sum_{i=1}^{n} height_i - n \cdot c = 0$$
$$ \hat{c} = \frac{1}{n} \sum_{i=1}^{n} height_i = \overline{height_i}$$

The hat over the $c$ indicates that this is the estimated value of the 
true but unknown $c$.
Everytime we estimate a parameter, we put a hat over it.

And voilÃ , we have estimated the parameter $c$ of the model, which is just the 
sample mean of all the heights. In contrast to before, we did not put 
in a lot of prior knowledge, but just estimated the parameter from the data.

In R, we can do this easily:
```{r, echo=TRUE, , message=FALSE, warning=FALSE}
mod <- lm(height ~ 1, data = d2)
summary(mod)
dim(d2)
mean(d2$height) # same as the intercept
sd(d2$height) / sqrt(nrow(d2)) # standard error of the estimator
# test-statistic for the intercept:
mean(d2$height) / (sd(d2$height) / sqrt(nrow(d2)))
# residual standard error:
sqrt(sum(mod$residuals^2) / (nrow(d2) - 1))
```

The `~1` means that there is just a so-called **intercept** in the model.
There are **no covariates**, just the constant $c$.
This is the simplest we can do. `lm` stands for linear model and with this base 
command in R we ask the software to do the least squares estimation for us.

Let's look at the **R-output** of the model estimation:

- `lm(formula = height ~ 1, data = d2)`: This is the model we estimated.
- `Residuals`: The difference between the actual height and the estimated height: 
$r_i = height_i - \hat{c}$. A univariate 5-point summary is given.
- `Coefficients`: The estimated coefficients of the model. In this case, there is just the intercept.
   We get the 
  - `Std. Error` of the estimate, i.e. the [standard error (SE)](https://en.wikipedia.org/wiki/Standard_error) of the mean, 
    which is (according to the Central Limit Theorem) 
    $$\frac{\sigma}{\sqrt{n}}$$ and can be estimated by the sample 
    standard deviation divided by the square root of the sample size.
    $\hat{SE} = \frac{s}{\sqrt{n}}$
  - the `t value` and the `Pr(>|t|)` which is the $p$-value of the (Wald-)test of 
    the null hypothesis that the 
    coefficient is zero ($H_0: \text{intercept}=0$). 
    This is a perfect example of an absolutely useless $t$-test. 
    Why? Because obviously ([exercise 2](#exercise2_Intro)) the population mean of body heights is not zero.
- `Residual standard error`: The standard deviation of the residuals $r_i = height_i - \hat{c}$.
    In this case identical with the sample standard deviation 
    of heights ([exercise 3](#exercise3_Intro)). $351$ degrees of freedom. There are $352$ observations and 
    $1$ parameter estimated (intercept/mean). Hence, there are $352-1=351$ 
    freely movable variables in the statistic of the sample standard deviation.

Let's look at the situation graphically:

```{r, echo=FALSE, , message=FALSE, warning=FALSE}
library(rethinking)
library(tidyverse)
data("Howell1")
d <- Howell1
d2 <- d[d$age >= 18, ]

# Select a few points to show vertical deviations
selected_points <- c(10, 50, 100)

mean_height <- mean(d2$height)

data.frame(height_i = 1:nrow(d2), height = d2$height) %>% # nolint: seq_linter.
  ggplot(aes(x = height_i, y = height)) +
  geom_point() +
  geom_hline(yintercept = mean_height, color = "red") +
  annotate("text", x = max(1:nrow(d2)) + 10, # nolint: seq_linter.
           y = mean_height,
           label = expression(hat(height[i])),
           color = "red", hjust = 0.6, vjust = 1.5) +
  geom_segment(data = data.frame(
    height_i = selected_points,
    height = d2$height[selected_points],
    mean_height = mean_height
  ),
  aes(x = height_i, xend = height_i,
      y = mean_height, yend = height),
  color = "red", linetype = "dotted") +
  ggtitle("Simple mean model estimation using Least Squares") +
  ylab("height estimated by model") +
  theme(plot.title = element_text(hjust = 0.5))

```

Above, the heights are plotted against the index of the observation (the order does not matter).
The variability of heights around the regression line (constant in this case) seems to 
stay constant, which is a good sign. We will call this 
[**homoscedasticity**](https://en.wikipedia.org/wiki/Homoscedasticity_and_heteroscedasticity) later.
The dashed vertical red lines show two residuals (one $>0$, the other $<0$), the difference between the actual height
and the estimated height. The model-estimated heights ($\widehat{heights_i}$) 
are all identical and nothing but the mean of all heights.

Peter Westfall explains in his excellent book a **conditional distribution 
approach** to regression, which is just what happens in the classical linear regression model.
I **highly recommend** reading the first chapters.

**What does this mean in this context?**
This means, that for every fixed value of the predictor (which we formally do not have),
the distribution of the response is normal with mean $\hat{c}$ and standard deviation $\sigma$.
Since, we do not have a predictor (apart from the intercept), we can say, that the distribution of the heights is normal with mean $\hat{c}$
and standard deviation $\sigma$. This is what we assumed in the model. It can also be directly seen
in the formula:

$$ height_i = c + \varepsilon_i $$

If you add a normally distributed random variable ($\varepsilon_i$) to a constant ($c$), 
the result is a normally distributed. No surprise here.

We can also create new data (heights) from this model and compare the distributions
of the actual data versus the model simulated data, since we have estimated the 
parameter $c$ and the standard deviation $\sigma$. 

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
c_hat <- mean(d2$height)
sigma_hat <- sqrt(sum(mod$residuals^2) / (nrow(d2) - 1))

# simulate heights from model
heights_sim <- rnorm(nrow(d2), c_hat, sigma_hat) # as many as in orig.

# Convert to data frames for plotting
df_actual <- data.frame(height = d2$height, type = "Actual Heights")
df_simulated <- data.frame(height = heights_sim, type = "Simulated Heights")

# Combine the datasets
df_combined <- rbind(df_actual, df_simulated)

ggplot(df_combined, aes(x = height, fill = type)) +
  geom_density(alpha = 0.5) +
  labs(title = "Comparison of Actual and Model-Simulated Heights",
       x = "Height",
       y = "Density") +
  scale_fill_manual(values = c("blue", "red")) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
```

One can repeat the simulation and see how the distributions changes to get 
a feeling for the variability of the model/data.

## Exercises

[E] Easy, [M] Medium, [H] Hard

(Some) solutions to exercises can be found in the git-repo [here](https://github.com/jdegenfellner/Script_QM2_ZHAW/tree/main/Solutions_Exercises).

### [E] Exercise 1 {#exercise1_Intro}

Use the [Swiss body heights](https://www.bfs.admin.ch/asset/de/30305714) data to 
determine 

- the 95% "Vertrauensintervall" for $\mu$ and
- calculate the standard deviation of the heights from $21,873$ Swiss people.
- Read the definition of the confidence interval in the footer of the table and
  explain why this is correct.

### [E] Exercise 2 {#exercise2_Intro}
Why do we **not** need a hypothesis test to know that the population mean of body heights is not zero?
Give 2 reasons.

### [H] Exercise 3 {#exercise3_Intro}
Verify analytically that the `Residual standard error` is identical with 
the sample standard deviation of the heights in [the simple mean model above](#classical_simple_mean_model).

### [M] Exercise 4 {#exercise4_Intro}
Repeat the Bayesian and Frequentist estimation of the simple model using a different data set about 
[chicken weights](https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/ChickWeight.html),
which is included in R.

- Set useful priors for the mean and standard deviation of the model 
  for the Baysian and the frequentist version considering 
  your a priori knowledge about chicken weights.

### [M] Exercise 5 {#exercise5_Intro}

- Try to understand/verify the code in the [addendum](#addendum) below.
- Change the parameters $\mu = (\mu_X, \mu_Y)$ and $\Sigma$ of the bivariate normal distribution 
  and see how the 3D plot changes.
- Plot the scatter plot of the $X$ and $Y$ values and compare it to the 3D plot.
  Does this make sense?
- Why is the scatter plot elliptical?

### [M] Exercise 6 {#exercise6_Intro}

Independence of two random variables $X$ and $Y$ implies that the 
correlation between them is zero.
A correlation of zero does not necessarily imply independence. 

- Verify this counterexample:
  For example, suppose the random variable $X$ is symmetrically distributed about zero, 
  and $Y = X^2$. Then $\mathit{Y}$ is completely determined by $X$, so that 
  $\mathit{X}$ and $\mathit{Y}$ are perfectly dependent, but their correlation is zero. 

### [H] Exercise 7 {#exercise7_Intro}
Go to the [section](#bivariate_normal) aobut the bivariate normal distribution below.

- How do you get the correlation matrix from the covariance matrix ($\Sigma$)?
  $$
  \Sigma =
  \begin{bmatrix}
  0.75 & 0.5 \\
  0.5 & 0.75
  \end{bmatrix}
  $$
  Hint: For the bivariate normal distribution, the correlation matrix is defined 
  [here](https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Bivariate_case).


## Addendum {#addendum}

### The bivariate normal distribution {#bivariate_normal}

As a refresher, you can look into the old QM1 script and read the 
chapter "4.7 Gemeinsame Verteilungen". 
Maybe [this video](https://www.youtube.com/watch?v=SP2GKq8xJ5I&ab_channel=StatisticsNinja) 
also helps.

The [bivariate normal distribution](https://en.wikipedia.org/wiki/Multivariate_normal_distribution) 
is a generalization of the normal distribution to two dimensions.
Now, we look at the distribution of two random variables $X$ and $Y$ **at the same time**.

Instead of one Gaussian curve, we have a 
[3D curve](https://en.wikipedia.org/wiki/Multivariate_normal_distribution#/media/File:Multivariate_Gaussian.png).
This curve defines how plausible different combinations of $X$ and $Y$ are.

Single points (like (3,6)) still have probability zero, because now the **volume** over a single point
($x$, $y$) is zero. The probability of a certain area is now the **volume** under the 
curve compared to the **area** under the density curve in the one-dimensional case.

**Example**: The following plot shows the density of a  bivariate normal distribution of 
two variables $X$ and $Y$
with $\mu_X = 0$, $\mu_Y = 0$, $\sigma_X = 1$, $\sigma_Y = 1$ and $\rho = \frac{2}{3}$.

Below is the correlation matrix of the bivariate normal distribution.

```{r, echo=FALSE, , message=FALSE, warning=FALSE}
library(MASS)    # For mvrnorm
library(plotly)  # For 3D plotting

# Define the parameters of the bivariate normal distribution
mu <- c(0, 0)                       # Mean
sigma <- matrix(c(0.75, 0.5, 0.5, 0.75), ncol = 2) # Covariance matrix

cov2cor(sigma)  # Convert covariance matrix to correlation matrix

# Generate a grid of x and y values
x <- seq(-3, 3, length.out = 100)
y <- seq(-3, 3, length.out = 100)
grid <- expand.grid(x = x, y = y)

# Compute the density of the bivariate normal
z <- matrix(dmvnorm(grid, mean = mu, sigma = sigma),
            nrow = length(x), ncol = length(y))

# Define the square for highlighting the probability volume
highlight_x <- c(0.5, 2)
highlight_y <- c(0.5, 2)

# Highlighted volume: Set values outside the square to NA
z_highlight <- z
z_highlight[grid$x < highlight_x[1] | grid$x > highlight_x[2] |
              grid$y < highlight_y[1] | grid$y > highlight_y[2]] <- NA

# Create a 3D plot
fig <- plot_ly() %>%
  add_surface(x = x, y = y, z = z, showscale = FALSE, opacity = 0.6) %>%
  add_surface(x = x, y = y, z = z_highlight,
              colorscale = "Viridis", opacity = 0.9) %>%
  layout(scene = list(
    xaxis = list(title = "X"),
    yaxis = list(title = "Y"),
    zaxis = list(title = "Density")
  ))

fig
```

If you move the plot around with your mouse, you see that there is a positive 
correlation between $X$ and $Y$ ($\rho = \frac{2}{3}$). This means that if $X$ is
above its mean, $Y$ is also more likely to be above its mean.
The variances of $X$ and $Y$ are both $1$. That means, that if you cut
through the plot in $X=0$ or $Y=0$, you see the same form of normal distribution.
If you look at if from above, we have hihglighted the section on the surface
over the area $X \in [0.5, 2]$ and $Y \in [0.5, 2]$. The volume over this area
under the density curve is the probability of this area: $P(X \in [0.5, 2] \text{ and } Y \in [0.5, 2])$

Calculate with R this probability with R:

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
# Load the mvtnorm package
library(mvtnorm)

# Define the parameters of the bivariate normal distribution
mu <- c(0, 0)                       # Mean
sigma <- matrix(c(0.75, 0.5, 0.5, 0.75), ncol = 2) # Covariance matrix

# Define the bounds of the square
highlight_x <- c(0.5, 2)
highlight_y <- c(0.5, 2)
# Calculate the probability using pmvnorm
pmvnorm(
  lower = c(highlight_x[1], highlight_y[1]),
  upper = c(highlight_x[2], highlight_y[2]),
  mean = mu,
  sigma = sigma
)
```

Since we do not believe everything we are told, we rather check via simulation,
if $0.1526$ is a plausible value for the probability:

```{r, echo=TRUE, , message=FALSE, warning=FALSE}
# Load necessary library
library(MASS)

# Define the parameters of the bivariate normal distribution
mu <- c(0, 0)                       # Mean
sigma <- matrix(c(0.75, 0.5, 0.5, 0.75), ncol = 2) # Covariance matrix
cov2cor(sigma)  # Convert covariance matrix to correlation matrix

# Define the bounds of the square
highlight_x <- c(0.5, 2)
highlight_y <- c(0.5, 2)

# Number of simulations
n_sim <- 10^4

set.seed(343434)
# Simulate bivariate normal samples
samples <- mvrnorm(n = n_sim, mu = mu, Sigma = sigma)

# Count how many samples fall within the square
inside_square <- sum(
  samples[, 1] >= highlight_x[1] & samples[, 1] <= highlight_x[2] &
  samples[, 2] >= highlight_y[1] & samples[, 2] <= highlight_y[2]
)

# Estimate the probability
inside_square / n_sim
```

Looks good.

<!--chapter:end:01-intro.Rmd-->

---
output:
  pdf_document: default
  html_document: default
---
# Simple Linear Regression

## Simple Linear Regression in the Bayesian Framework {#simple_lin_reg_bayes}

You can watch [this video](https://www.youtube.com/watch?v=14mkCpJ7tKs&ab_channel=VeryNormal) as primer.

We will now add one covariate/explanatory variable to the model. 
Refer to [Statistical Rethinking](https://civil.colorado.edu/~balajir/CVEN6833/bayes-resources/RM-StatRethink-Bayes.pdf) 
"4.4 Linear prediction" or "4.4 Adding a predictor" as it's called in the online version of the book.

So far, our "regression" did not do much to be honest. The mean of a list of values
was already calculated in the [descriptive statistics section](https://jdegenfellner.github.io/Script_QM1_ZHAW/descriptive_stats.html) 
before and we have mentioned how great this statistic is as measure of location and where its weaknesses are.

Now, we want to model **how** body height and weight are **related**.
Formally, one wants to *predict* body heights from body weights.

Here and in the Frequentist framework, we will see that it is **not the same**
problem (and therefore results in a different statistical model) 
**to predict body weights from body heights or vice versa**.

We remember the following:

- Regress $Y$ on $X$, which is equivalent to predict $Y$ from $X$.
  We know X and want to predict Y.
- Regress $X$ on $Y$, which is equivalent to predict $X$ from $Y$.
  We know Y and want to predict X.

The word "predictor" is important here. It is a technical term
and describes a variable that we know (in our case weight) and with 
which we want to "guess as good as possible" the value of the 
dependent variable (in our case height). "As good as possible"
means that we put a penalty on an error. The farer our prediction
is away from the true value ($y_i$), the higher the penalty.
And not only that, but if you are twice as far away from the true value,
you should be penalized four times as much. This is the idea behind
the squared error [loss function](https://en.wikipedia.org/wiki/Loss_function) 
and the core of the least squares method.
What if we would punish differently, you ask?
There are many loss functions one could use 
(for instance the [Huber loss](https://en.wikipedia.org/wiki/Huber_loss)), 
maybe we will see some later. For now, we punish quadratically.

We **always** visualize the data first to improve our understanding.
First comes descriptive statistics, then one can think about modeling.

```{r}
library(rethinking)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]
plot(d2$height ~ d2$weight)
```

It's not often that you see such a clean plot.
The scatterplot indicates a linear relationship between the two variables.
The higher the weight, the higher the height; with some deviations of course and
we decide that normally distributed errors are a good idea.
This relationsip is neither causal, nor deterministic.

- It is not causal since an increase in weight does not 
  necessarily lead to an increase in height, especially in grown-ups.
- It is not deterministic since there are deviations from the line. 
  It if was deterministic, we would not need statistical modeling. 

For simpler notation, we will call `d2$weight` $x$. $\bar{x}$
is the mean of $x$.

### Model definition

Let's write down our **model** (again with the Swiss population prior mean):

\begin{eqnarray*}
h_i &\sim& \text{Normal}(\mu_i, \sigma)\\
\mu_i &\sim& \alpha + \beta (x_i - \bar{x})\\
\alpha &\sim& \text{Normal}(171.1, 20)\\
\beta &\sim& \text{Normal}(0, 10)\\
\sigma &\sim& \text{Uniform}(0, 50)
\end{eqnarray*}

Visualization of the **model structure**:

```{r, echo=FALSE}
library(DiagrammeR)

grViz("
digraph model {
  graph [layout = dot, rankdir = TB]

  # Nodes for variables
  node [shape = ellipse, style = filled, fillcolor = lightblue]
  alpha [label = 'Î± ~ Normal(171.1, 20)']
  beta [label = 'Î² ~ Normal(0, 10)']
  sigma [label = 'Ï ~ Uniform(0, 50)']
  mu [label = 'Î¼_i = Î± + Î²(x_i - xÌ)']
  h_i [label = 'h[i] ~ Normal(Î¼_i, Ï)']

  # Connections
  alpha -> mu
  beta -> mu
  mu -> h_i
  sigma -> h_i
}
")
```

There are now additional lines for the priors of $\alpha$ and $\beta$ 
(compared to the simple mean model before).
The model structure also shows the way to simulate from the prior.
One starts at the top and ends up with the heights.

- $h_i$ is the height of the $i$-th person and we assume it is normally distributed.
- $\mu_i$ is the mean of the height of the $i$-th person and we 
  assume it is linearly dependent on the difference $x_i-\bar{x}$. 
  Compared to the intercept model, a different mean is assumed for each person
  depending on his/her weight. 
- $\alpha$ is the intercept and we use the same prior as before.
- $\beta$ is the slope of the line and we use the normal distribution as prior for it,
  hence it can be positive or negative and how plausible each value is, is
  determined by that specific normal distribution. Note, that we could 
  easily adapt the distribution to any distribution we like.
- The prior for $\sigma$ is unchanged.
- $x_i - \bar{x}$ is the deviation of the weight from the mean weight, thereby **we
  center** the weight variable. This is a common practice in regression analysis.
  A value $x_i - \bar{x} > 0$ implies that the person is heavier than the average.

The linear model is quite popular in applied statistics and one
reason is probably the rather straightforward interpretation of the coefficients:
A one unit increase in weight is on average (in the mean) associated with 
a $\beta$ unit increase/decrease (depending if $\beta$ is $>0$ or $<0$) in height.

### Priors

We want to plot our prior predictions to get a feeling **what 
the model would predict without seeing the data**.
This is a kind of "sanity check" to see if the priors and 
the model definition are reasonable.
Again, we just draw from the assumed distributions for $\alpha$ and $\beta$
100 times and draw the corresponding lines. Just as the model 
definition says.

```{r}
set.seed(2971)
N <- 100  # 100 lines
a <- rnorm(N, 171.1, 20)
b <- rnorm(N, 0, 10)

xbar <- mean(d2$weight)

# start with empty plot
plot(NULL, xlim = range(d2$weight), ylim = c(-100, 400),
     xlab = "weight", ylab = "height")
abline(h = 0, lty = 2)  # horizontal line at 0
abline(h = 272, lty = 1, lwd = 0.5)  # horizontal line at 272
mtext("b ~ dnorm(0, 10)")

# Overlay the 100 lines
for (i in 1:N) {
  curve(a[i] + b[i] * (x - xbar),
        from = min(d2$weight), to = max(d2$weight),
        add = TRUE, col = col.alpha("black", 0.2))
}
```

This linear relationship defined with the chosen priors seems rather 
non-restrictive. According to our priors,
one could see very steeply rising or falling relationships between weight and expected heights. 
We could at least make the priors for the slope ($\beta$) non-negative. 
One possibility to do this 
is to use a [log-normal distribution](https://en.wikipedia.org/wiki/Log-normal_distribution) 
for the prior of $\beta$ which can only take non-negative values.

$$\beta \sim \text{Log-Normal}(0, 1)$$

Lets plot the priors again.

```{r}
set.seed(2971)
N <- 100  # 100 lines
a <- rnorm(N, 171.1, 20)
b <- rlnorm(N, 0, 1)

xbar <- mean(d2$weight)

plot(NULL, xlim = range(d2$weight), ylim = c(-100, 400),
     xlab = "weight", ylab = "height")
abline(h = 0, lty = 2)  # horizontal line at 0
abline(h = 272, lty = 1, lwd = 0.5)  # horizontal line at 272
mtext("b ~ dlnorm(0, 1)")

# Overlay the 100 lines
for (i in 1:N) {
  curve(a[i] + b[i] * (x - xbar),
        from = min(d2$weight), to = max(d2$weight),
        add = TRUE, col = col.alpha("black", 0.2))
}
```

This seems definitely more realistic. There is some sort of positive linear relationship
between weight and expected height.

### Fit model

Now, let's **estimate the posterior/fit the model** as before:

```{r}
# load data again, since it's a long way back
library(rethinking)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]
xbar <- mean(d2$weight)
# fit model
mod <- quap(
    alist(
        height ~ dnorm(mu, sigma),
        mu <- a + b * (weight - xbar),
        a ~ dnorm(171.1, 100),
        b ~ dnorm(0, 10),
        sigma ~ dunif(0, 50)
    ) ,
data = d2)
```

Note that the model definition was now directly included in the `quap` function.
Let's look at the **marginal distributions** of the parameters:

```{r}
precis(mod)
plot(mod)
```

Note, that the credible intervals in the plot of the coefficients are 
hardly noticable. The reason is that the intercept $a$ is rather large,
whereas $b$ and sigma are comparativly small. The plot is scaled to the
largest parameter.

We can improve this by centering the height variable as well. Or,
we could standardize both weight and height. This would change the interpretation
of the $\beta$ parameter. It would then be the expected change in standard deviations
when changing weight by one standard deviation. 
See [exercise 12](#exercise12_simpl_lin_reg).

The analysis yields estimates for all our parameters of the model: $\alpha$,
$\beta$ and $\sigma$. The estimates are the mean of the posterior distribution.

See [exercise 2](#exercise2_simpl_lin_reg).

**Interpretation of $\beta$**:
The mean of the posterior distribution of $\beta$ is 0.9. A person with a weight
of 1 kg more weight can be expected to be 0.9 cm taller. A 89% credible interval
for this estimate is $[0.83, 0.97]$. We can be quite sure that the slope is 
positive (of course we designed it that way too via the prior). 

It might also be interesting to inspect the **variance-covariance matrix**, 
respectively the correlation between the parameters as we did before
in the intercept model. Remember, these are the correlations of
parameters in the multivariate (because three paremeters simulatenously) 
posterior distribution.

```{r}
diag(vcov(mod))
round(cov2cor(vcov(mod)),2)
```

- `diag(vcov(mod))` gives the variances of the parameters and 
- `cov2cor(vcov(mod))`the correlations. As we can see the correlations are (near) zero. Compare to the graphical
  display of the model structure. There is no connection.

### Result

**Graphical end result** of fitting the model:
We plot the marginal posterior distributions of $\alpha$ and $\beta$,
and also the raw data with the found regression line.

```{r}
post <- extract.samples(mod)
dens(post$a, col = rangi2)
dens(post$b, col = rangi2)

# both posteror plots seem symmetric
# we use the mean as point estimate
# for a and b.
a_quap <- mean(post$a)
b_quap <- mean(post$b)

plot(d2$height ~ d2$weight, col = rangi2)
curve(a_quap + b_quap * (x - xbar), add = TRUE)
```

### Credible bands

We could draw again and again from the posterior distribution
and calculate the means like above. Plotting the regression lines
with the respective parameters $\alpha$, $\beta$ would indicate
the variability of the estimates.
Note that we do not draw from the data (as one does in bootstrap resampling),
but from the posterior distribution.
The `link` function does this for us. It takes the posterior distribution we
have just fit, samples $\alpha$ and $\beta$ from it, calculates the mean and then
samples from this normal distribution for the mean at a given weight.
Refer to pages 98-106 in the current version of the book Statistical Rethinking
for all details.

```{r}
# Define a sequence of weights for predictions
weight.seq <- seq(from = 25, to = 70, by = 1)

# Use the model to compute mu for each weight
mu <- link(mod, data = data.frame(weight = weight.seq))
str(mu)

# Visualize the distribution of mu values
plot(height ~ weight, d2, type = "n")  # Hide raw data with type = "n"

# Loop over samples and plot each mu value
for (i in 1:100) {
  points(weight.seq, mu[i, ], pch = 16, col = col.alpha(rangi2, 0.1))
}
```

The `link` function fixes the weight at the values in `weight.seq` and
draws samples from the posterior distribution of the parameters. Above, 
the first $100$ (of $1000$) rows are used.

We can also draw a nice shade (using all 1000 rows of the $\mu$-matrix) 
for the regression line:

```{r}
# Summarize the distribution of mu
mu.mean <- apply(mu, 2, mean)
mu.PI <- apply(mu, 2, PI, prob = 0.89)
plot(height ~ weight, d2, col = col.alpha(rangi2, 0.5))
lines(weight.seq, mu.mean)
shade(mu.PI, weight.seq)
```

The function `PI` from the `rethinking` package calculates the 89% percentile interval 
for the mean of the height at a certain weight. The `apply` function 
apples this function to each columns (hence the `2`) of the matrix `mu`.

As we can see, we are pretty sure about the mean of height which 
we wanted to model in the first place.

**Mean modeling is one thing, individual prediction is another**.
Given a certain weight of a person, what is the height of the same person?
The first line in the model definition ($height_i \sim Normal(\mu_i, \sigma)$)
tells us that a person's height is distributed *around* the mean
(which linearly depends on weight) and is not necessary the mean itself.

To get to an **individual prediction**, we need to consider the uncertainty 
of the parameter estimation *and* the uncertainty from the Gaussian distribution
around the mean (at a certain weight). We do this with `sim`.

```{r}
# Simulate heights from the posterior
sim.height <- sim(mod, data = list(weight = weight.seq))
str(sim.height)

# Compute the 89% prediction interval for simulated heights
height.PI <- apply(sim.height, 2, PI, prob = 0.89)

# Plot the raw data
plot(height ~ weight, d2, col = col.alpha(rangi2, 0.5))

# Draw MAP (mean a posteriori) line
lines(weight.seq, mu.mean)

# Draw HPDI (highest posterior density interval) region for mu
shade(mu.PI, weight.seq)

# Draw PI (prediction interval) region for simulated heights
shade(height.PI, weight.seq)
```

Here, the `PI` function is applied to the simulated heights and calculates the 89%
percentile interval for each weight.

The lighter and wider shaded region is where the model expects to find 89%
of the heights of a person with a certain weight. 

This part is **sometimes a bit desillusioning** when seen for the first time:
Draw a horizontal line at 150 cm and see how many weights (according to 
the individual prediction) are compatible with this height. Weights 
from 30 to 50 kg are compatible with this height according to the 
89% prediction interval:

```{r, echo=FALSE}
# Simulate heights from the posterior
sim.height <- sim(mod, data = list(weight = weight.seq))
#str(sim.height)
##  num [1:1000, 1:46] 138 130 130 147 136 ...

# Compute the 89% prediction interval for simulated heights
height.PI <- apply(sim.height, 2, PI, prob = 0.89)

# Plot the raw data
plot(height ~ weight, d2, col = col.alpha(rangi2, 0.5))

# Draw MAP (mean a posteriori) line
lines(weight.seq, mu.mean)

# Draw HPDI (highest posterior density interval) region for mu
shade(mu.PI, weight.seq)

# Draw PI (prediction interval) region for simulated heights
shade(height.PI, weight.seq)

# Add horizontal line at height = 150
abline(h = 150, col = "red", lty = 2)

# Find weight range for height = 150
height_at_150 <- which.min(abs(height.PI[1, ] - 150)) # Closest weight for lower bound
weight_lower <- weight.seq[height_at_150]

height_at_150_upper <- which.min(abs(height.PI[2, ] - 150)) # Closest weight for upper bound
weight_upper <- weight.seq[height_at_150_upper]

# Add vertical lines for weight range
abline(v = c(weight_lower, weight_upper), col = "blue", lty = 2)

# Annotate the plot (optional)
text(weight_lower, 150, labels = paste("Upper:", round(weight_lower, 1)), pos = 4)
text(weight_upper, 150, labels = paste("Lower:", round(weight_upper, 1)), pos = 4)
```

The higher the credibility, the wider the interval,
the wider the range of compatible weights. 
In our example, more than 60% of the weight-range are 
plausible to predict a height of 150 cm.

```{r}
(50 - 30) / (range(d2$weight)[2] - range(d2$weight)[1])
```

On the other hand: We did not model the relationship this way.
We modeled **height depending on weight** and not the other way around. 
In the next chapter, we will regress weight on height (yes, this is the correct order)
and see what changes.



### Summary

- We have added a covariate (weight) to the simple mean model to predict height.
- We have centered the weight variable.
- We have defined and refined priors for the intercept and slope.
- We have estimated the posterior distribution of the parameters using quadratic approximation with `quap`.
- We have visualized the result.
- We have created credible bands for mean and individual predictions.


## Simple Linear Regression in the Frequentist Framework

We will now do the same analysis in the Frequentist framework while introducing
some foundational theory along the way.
I recommend reading the first couple of chapters from [Westfall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4).

### Model definition

Our linear model is defined as:

$$h_i = \beta_0 + \beta_1 x_i + \varepsilon_i$$

where

- $\varepsilon_i$ is the error term with $\varepsilon_i \sim N(0, \sigma^2), \forall i$
- $\beta_0$ is the unknown but fixed intercept
- $\beta_1$ is the unknown but fixed slope

#### Model Assumptions of the Classical Regression Model ([Westfall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4), 1.7): {#_model_assumptions}

The first and **most important assumption** is that the data are produced  
probabilistically, which is specifically stated as
$$Y|X = x \sim p(y|x)$$

What does this mean? 

- $Y|X = x$ is the random variable Y **conditional** on X being equal to x, i.e. the 
   distribution of $Y$ if we know the value of $X$ (in our example the weight in kg).
   [This](https://blogs.sas.com/content/iml/files/2015/09/GLM_normal_identity.png) is a nice image of what is meant here.
- $p(y|x)$ is the distribution of potentially observable $Y$ given $X = x$. 
  In our case above this was the normal distribution with mean $\mu_i$ and variance $\sigma$.

You can play with [this shiny app](https://psychmeth.shinyapps.io/Regression-NVFehler/) to improve your understanding.
It offers the option "Bedingte Verteilung anzeigen".

One always thinks about the so-called
[data generating process](https://en.wikipedia.org/wiki/Data_generating_process) 
([Westfall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4), 1.2).
How did the data come about? There is a process behind it and this process
is attempted to be modeled.

**Further assumptions**:

- **Correct functional specification**: The conditional mean function $f(x) = \mathbb{E}(Y|X=x)$.
  In the case of the linear model, the assumption is $\mathbb{E}(Y|X=x) = \alpha + \beta x$. 
  The **expectation** of $Y$ (height) depends linearly on $x$ (weight).
  This assumption is violated when the true relationship is not linear or the
  data at least suggest that it is not linear, like [here](https://www.alexanderdemos.org/Class5_files/figure-html/unnamed-chunk-2-1.png).

- **The errors are homoscedastic** (constant variance $\sigma^2$). This means the
  variances of all conditional distributions $p(y|x)$ are constant ($=\sigma^2$).
  This assumption is (for instance) violated if points are [spreading out more and more
  around the regression line](https://www.investopedia.com/thmb/n9S9lWMv6X9-sKC2DtAOtUTPSik=/1500x0/filters:no_upscale():max_bytes(150000):strip_icc():format(webp)/Heteroskedasticity22-ce5acc2acef6494d91935588b0599579.png), 
  indicating that the errors are getting larger.

- **Normality**. For the classical linear regression model all the conditional
  distributions $p(y|x)$ are normal distributions. It could well be, that
  the errors are not nicely normally distributed around the regression line,
  for instance if we have a lot of outliers upwards and the distribution 
  is skewed, like [here (Figure 5.20)](https://www.bookdown.org/rwnahhas/RMPH/mlr-normality.html).

- The **errors are independent** of each other. 
  The potentially observable $\varepsilon_i = Y_i - f(\mathbf{x_i}, \mathbf{\beta})$
  is uncorrelated with $\varepsilon_j = Y_j - f(\mathbf{x_j}, \mathbf{\beta})$ for 
  $i \neq j$. This assumption is violated if the errors are correlated,
  here is an example: The true data comes from a sine curve and we estimate a 
  linear model (green), which does not fit the data well (left plot).
  The residuals plot shows clear patterns (right plot) and indicates
  that the errors are correlated. Specifically, the errors around $x=2$ and $x=4$ 
  are negatively correlated (see [exercise 6](#exercise6_simpl_lin_reg)).
```{r, echo=FALSE, warning=FALSE}

# Set seed for reproducibility
set.seed(123)

# Generate data from a sine function
x <- seq(0, 2 * pi, length.out = 100)       # Predictor variable
true_y <- sin(x)                            # True values based on sine function

# Add noise to simulate observed data
noise <- rnorm(length(x), mean = 0, sd = 0.2)
observed_y <- true_y + noise                # Observed response variable

# Perform a simple linear regression with intercept
linear_model <- lm(observed_y ~ x)

# Display summary of the model
#summary(linear_model)

# Load necessary libraries
library(ggplot2)
library(patchwork)

# Create the first plot: Observed data and linear regression fit
plot1 <- ggplot(data = data.frame(x, observed_y, true_y), aes(x = x)) +
  geom_point(aes(y = observed_y), color = "blue", size = 2) +
  geom_line(aes(y = true_y), color = "red",
            linetype = "dashed", linewidth = 1) +
  geom_abline(intercept = coef(linear_model)[1], slope = coef(linear_model)[2],
              color = "green", linewidth = 1) +
  labs(title = "Sine Data with Linear Regression", x = "x", y = "Observed y") +
  theme_minimal()

# Create the second plot: Residuals
plot2 <- ggplot(data = data.frame(x, residuals = residuals(linear_model)),
                aes(x = x, y = residuals)) +
  geom_point(color = "blue", size = 2) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red") +
  labs(title = "Residuals vs. x", x = "x", y = "Residuals") +
  theme_minimal()

# Arrange the plots side by side using patchwork
plot1 + plot2
```

In the case above, the errors are **not conditionally independent**.
If we condition on $X=2$ and $X=4.5$, the errors are correlated ($r \sim -0.3$),
which they should not be.

These assumptions become clearer as we go along and should be checked 
for every model we fit. They are not connected, they can all be true or false.
The question is not "Are the assumptions met?" since they never are exactly met.
The question is **how** "badly" the assumptions are violated?

Remember, **all models are wrong, but some are useful**.

In full, **the classical linear regression model** can be written as:

$$Y_i|X_i = x_i \sim_{independent} N(\beta_0 + \beta_1 x_{i1} + \dots \beta_k x_{ik},\sigma^2)$$
for $i = 1, \dots, n$.

Interpretation of this formula:

The outcome variables $Y_i$ are independently (from each other, 
since the errors are indepentent from each other) conditionally normally distributed
(conditioned on $X_i=x_i$) with a mean that is linearly depending on
true but unknown parameters $\beta_0, \beta_1, \dots, \beta_k$ and a 
constant variance $\sigma^2$.

### Fit the model {#fit_model_simple_lin_reg_classic}

Again, we fit the model using the [**least squares method**](https://en.wikipedia.org/wiki/Least_squares). 
For a neat animated explanation, 
visit [this video](https://www.youtube.com/watch?v=jEEJNz0RK4Q&ab_channel=COCCmath).
There are literally hundreds of videos on the topic. Choose wisely. Not all are good.
If in doubt, use our recommended books as reading materials. This is the most reliable source.
A hint along the way: Be very sceptical if you ask GPT about information, 
although for this special case one has a good chance of getting a decent answer due to 
the vast amount of training data. 

One has to minimize the sum of squared differences between the true heights and 
the model-predicted heights in order to find $\beta_0$ and $\beta_1$.

$$SSE(\beta_0, \beta_1) = \sum_{i=1}^n (y_i - (\beta_0 + \beta_1 x_i))^2$$

We omit the technical details (set derivative to zero and solve the system) and give the results for $\beta_0$ and $\beta_1$:

\[
\hat{\beta_0} = \bar{y} - \hat{\beta_1} \bar{x},
\]
\[
\hat{\beta_1} = \frac{\sum_{i=1}^n (x_i - \bar{x})(y_i - \bar{y})}{\sum_{i=1}^n (x_i - \bar{x})^2} =
 \frac{s_{x,y}}{s_x^2} = r_{xy} \frac{s_y}{s_x}.
\]

where:

- \(r_{xy}\) is the [sample correlation coefficient](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient) between \(x\) and \(y\)
- \(s_x\) and \(s_y\) are the [uncorrected sample standard deviations](https://en.wikipedia.org/wiki/Standard_deviation) of \(x\) and \(y\)
- \(s_x^2\) and \(s_{xy}\) are the [sample variance](https://en.wikipedia.org/wiki/Variance) and [sample covariance](https://en.wikipedia.org/wiki/Covariance), respectively

Interpretation of $\hat{\beta}_0$ and $\hat{\beta}_1$: see [exercise 3](#exercise3_simpl_lin_reg).

Let's **use R again to solve the problem**:

```{r}
library(rethinking)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]
mod <- lm(height ~ weight, data = d2)
summary(mod)
```

**Interpretation of R-output**:

- `Call`: The model that was fitted.
- `Residuals`: $r_i = height_i - \widehat{height}_i$. 
   Differences between true heights and model-predicted heights.
- `Coefficients`: Names for $\beta_0$ and $\beta_1$. We call them $\hat{\beta}_0$ and $\hat{\beta}_1$.
   - `Estimate`: The (least squares) estimated value of the coefficient.
   - `Std. Error`: The standard error of the estimate. 
                   This is a measure of how precise the estimator for the coefficient is.
   - `t value`: The value of the $t$-statistic for the (Wald-) hypothesis test
        $H_0: \beta_i = 0$.
   - `Pr(>|t|)`: The $p$-value of the hypothesis test. 
- `Residual standard error`: The estimate of $\sigma$ 
   which is also a model parameter (as in the Bayesian framework).
- `Multiple R-squared`: The proportion of the variance explained by the 
   model (we will explain this below).
- `Adjusted R-squared`: A corrected version of the $R^2$ which takes into account
   the number of predictors in the model.
- `F-statistic`: The value of the $F$-statistic for the hypothesis test: 
    $H_0: \beta_1 = \beta_2 = \dots = \beta_k = 0$. Note, the alternative 
    hypotheses to this test is that *any* of the $\beta_i$ is not zero. If that is
    the case, the model explains more than the mean model with just $\beta_0$.

We could also **solve the least squares problem graphically**: We want to find the 
values of $\beta_0$ and $\beta_1$ that minimize the sum of squared differences
which can be plotted as 3D function. Since the function is a sum of squared terms,
we should expected a paraboloid form.
All we have to do is to ask R which of 
the coordinates ($\beta_0$, $\beta_1$) minimizes the sum of squared errors. 
The result confirms
the results from the `lm` function. The dot in red marks the spot 
(Code is in the [git repository](https://github.com/jdegenfellner/Script_QM2_ZHAW)):


```{r, echo=FALSE, eval=TRUE}
library(pacman)
p_load(plotly, tidyverse, rethinking)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]

d2$x <- d2$weight
d2$y <- d2$height

mod <- lm(y ~ x, data = d2)
#summary(mod)
cat("Intercept beta_0: ", coef(mod)[1], "\n")
cat("Slope beta_1: ", coef(mod)[2], "\n")

sse <- function(beta_0, beta_1, data) {
  predicted <- beta_0 + beta_1 * data$x
  sum((data$y - predicted)^2)
}

beta_0_values <- seq(100, 126, length.out = 1000) # adapt to your data
beta_1_values <- seq(-1, 3, length.out = 1000) # adapt to your data
sse_values <- outer(beta_0_values, beta_1_values, Vectorize(function(a, b) sse(a, b, d2)))
fig <- plot_ly(x = ~beta_0_values, y = ~beta_1_values, z = ~sse_values, type = "surface") %>%
  layout(title = "Sum of squared errors depending on beta_0 and beta_1")
#fig

# Highlight min of this paraboloid (fig)----
min_sse_index <- which(sse_values == min(sse_values), arr.ind = TRUE) # Find indices where error is smallest.
min_beta_0 <- beta_0_values[min_sse_index[1]]
min_beta_1 <- beta_1_values[min_sse_index[2]]
min_sse <- min(sse_values)

# check
#min_beta_0 # (= Intercept)
#min_beta_1
#coef(mod) # match

fig <- fig %>% add_trace(
  x = min_beta_0, 
  y = min_beta_1, 
  z = min_sse, 
  type = 'scatter3d', 
  mode = 'markers',
  marker = list(color = 'red', size = 7)
)

fig
```

### Confidence Intervals of coefficients (Frequentist) {#confidence_intervals_frequentist}

You can get CI's conveniently with the `confint` function:

```{r}
confint(mod, level = 0.96)
```

Remember, these are Frequentist confidence intervals. 

**If one repeats the experiment (i.e. draws new data from the true model)
many times, the true but unknown value of the parameter will 
be in the interval in 96% of the cases.**

We can also use the simple [bootstrap](https://en.wikipedia.org/wiki/Bootstrapping_(statistics)). 
The advantage of this technique is that we can basically always use it,
no matter how compliated the estimator is. We do not need formulae.
We simply

- create 1000 bootstrap samples, 
- fit the model, 
- store the coefficients. 
- The 2% and 98% quantiles of the coefficients constitute the 96% bootstrap confidence interval.

```{r}
set.seed(123)
n <- nrow(d2)
B <- 1000
boot_coefs <- matrix(NA, nrow = B, ncol = 2)
for (i in 1:B) {
  boot_idx <- sample(1:n, replace = TRUE)
  boot_mod <- lm(height ~ weight, data = d2[boot_idx, ])
  boot_coefs[i, ] <- coef(boot_mod)
}
#head(boot_coefs)
t(apply(boot_coefs, 2, quantile, c(0.02, 0.98)))
```
The CIs are quite similar to the ones from the `confint` function.

In the Bayesian setting, we used the centered weight variable. Let's to this here too for 
comparison and use 89% coverage probability.

```{r}
d2$weight_centered <- d2$weight - mean(d2$weight)
mod_centered <- lm(height ~ weight_centered, data = d2)
#summary(mod_centered)
confint(mod_centered, level = 0.89)
```

Compare with `precis` from the Bayesian model:
```{r, echo=FALSE}
library(rethinking)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]
xbar <- mean(d2$weight)
# fit model
mod <- quap(
    alist(
        height ~ dnorm(mu, sigma),
        mu <- a + b * (weight - xbar),
        a ~ dnorm(171.1, 100),
        b ~ dnorm(0, 10),
        sigma ~ dunif(0, 50)
    ) ,
data = d2)
precis(mod)
```

We are glad to see that both analyses align really nicely.

### ANOVA (Analysis of Variance) {#analysis_of_variance}

A non-obvious and very useful finding is that the total variability (SST)
in the data (our heights) can be
**decomposed** (or [analysed](https://en.wiktionary.org/wiki/analysis)) 
into two parts:

- The variability explained by the model (the regression line): SSR
- The variability not explained by the model (the residuals): SSE

$$\text{Sum of Squares in Total} = \text{Sum of Squares from Regression} + \text{Sum of Squared Errors}$$

$$SST = SSR + SSE$$

$$\sum_{i=1}^{n} (y_i - \bar{y})^2 = \sum_{i=1}^{n} (\hat{y}_i - \bar{y})^2 + \sum_{i=1}^{n} (y_i - \hat{y}_i)^2$$

If you are interested in the details, check out [this](https://en.wikipedia.org/wiki/Explained_sum_of_squares#Partitioning_in_the_general_ordinary_least_squares_model).

[This video](https://www.youtube.com/watch?v=NxRTs7sXKAQ&ab_channel=365DataScience) 
explains the concept nicely.

Let's visualize our regression result:

```{r, echo=FALSE}
library(rethinking)
library(tidyverse)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]

# Fit linear model
mod <- lm(height ~ weight, data = d2)

# Calculate predicted heights for the regression line
d2 <- d2 %>%
  mutate(predicted_height = predict(mod, newdata = d2),
         mean_height = mean(height))

# Select one point to highlight the differences
highlight_point <- d2[10, ]

# Create the plot
ggplot(d2, aes(x = weight, y = height)) +
  # Scatterplot of all points
  geom_point(alpha = 0.6) +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linewidth = 1) +
  geom_point(data = highlight_point, aes(x = weight, y = height),
            color = "red", size = 3) +
  
  # Vertical difference: Point and mean height (offset slightly left)
  geom_segment(data = highlight_point,
               aes(x = weight - 0.5, xend = weight - 0.5,
                   y = height, yend = mean_height, color = "Point-to-Mean"),
               linetype = "dashed") +
  
  # Vertical difference: Point and regression line (no offset)
  geom_segment(data = highlight_point,
               aes(x = weight, xend = weight,
                   y = height, yend = predicted_height,
                   color = "Point-to-Line"),
               linetype = "dashed") +
  
  # Vertical difference: Regression line and mean height (offset slightly right)
  geom_segment(data = highlight_point,
               aes(x = weight + 0.5, xend = weight + 0.5,
                   y = predicted_height,
                   yend = mean_height, color = "Line-to-Mean"),
               linetype = "dashed") +
  
  # Add horizontal line for the mean height
  geom_hline(yintercept = mean(d2$height), 
             color = "darkgreen", linetype = "dashed") +
  # Labels
  labs(title = "Scatterplot with Regression Line and Differences",
       x = "Weight",
       y = "Height",
       color = "Distance Type") +
  # Style
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5,
                              size = 16),  # Center and adjust title size
    legend.position = "top"  # Place legend at the top
  )
```

The blue dotted line is the distance from
the mean to the point (total variance), the red dotted line is the distance from
the mean to the regression line (explained variance) and the green dotted line
is the distance from the regression line to the point (unexplained variance).
We see that it adds up. I find this fact quite fascinating.
Thank you [Ronald Fisher](https://en.wikipedia.org/wiki/Analysis_of_variance#:~:text=ANOVA%20was%20developed%20by%20the,to%20different%20sources%20of%20variation.).

See [exercise 13](#exercise13_simpl_lin_reg).

### $R^2$ - Coefficient of Determination
[$R^2$](https://en.wikipedia.org/wiki/Coefficient_of_determination)
is the **amount of variance explained by the model**. 
You can also read [Westfall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4) 8.1.

As you can see 
[above](#analysis_of_variance), the total variance (SST) of our outcome (height) 
can be decomposed into two parts: the variance explained by the model (SSR)
and the variance not explained by the model (SSE).

Maybe the most intuitive definition of $R^2$ is:

$$R^2 = \frac{SSR}{SST} = \frac{SST - SSE}{SST} = 1 - \frac{SSE}{SST}$$

The value is between 0 and 1. The higher the value, the more variance is explained.
But be cautious. Depending on the context, a really high $R^2$ is not 
necessarily a good thing. With the data we are working with,
it could easily hint towards an error. If we are near 1,
all points in the simple linear regression model are on the line.
If we are near 0, the model does not explain much of the variance 
and we would see "noise with no slope" in the scatterplot ([exercise 4](#exercise4_simpl_lin_reg)).
The normal $R^2$ can be found in the R output under `Multiple R-squared`.

If you add a lot of variables to your regression model, you can get an
$R^2$ of 1. The $R^2$ will never decrease when adding more variables.
We will verify this when we have more than 2 explanatory variables. 
As a non-formal explanation for this: In the Sum of Squares Errors (SSE),
if you add more covariates ($\beta_2, \beta_3$), you have more freedom
to choose values that minimize the number that will be squared. Simple regression
is just a special case of multiple (more than one predictor) regression with $\beta_2=\beta_3=\dots=0$.
Hence, you will definitely not be worse off with regards to SSE when using more covariates.
A smaller SSE implies a larger SSR (sum constraint; SST remains constant) and hence a larger $R^2$.
If you have as many explanatory variables as data points, you can get an $R^2$ of 1. This is
[overfitting](https://en.wikipedia.org/wiki/Overfitting) at its "best" (which we want to avoid). 
You would just get a value for each data point by setting all other $\beta_i$ to zero and 
$\beta_i = \frac{y_i}{x_i}$. 
Since we want to find the unterlying process, we want to avoid this.

Although not perfect, one way to mitigate the influence of "too many" variables
on $R^2$ is to use the adjusted $R^2$, which an also be found in the R output (`Adjusted R-squared`).

#### Seperating property of regression due to $R^2$: {#seperating_property}
[Peter Westfall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4)
explains (in Figure 8.1 of the book) how $R^2$ influences the separation of distributions in our simple regression model.

In our regression of *height on weight* (the order is correct, that's how you say it), 
the $R^2$ is $0.5696$. The following plot shows how "well" (i.e. precise) one can predict
height if we use the 10% and 90% quantile of the weights (x_low and x_high).
In both, you see the conditional distribution of height given the weight $X = x_{low}$ or $X = x_{high}$.
Scenario 1 is the original model, scenario 2 is the same data with added noise (in Y-direction),
which reduces $R^2$ to $0.13$, much lower. In the right plot, the distributions have a large
overlap and it is hard to distinguish between weights when seeing the height.
With a very low $R^2$, the height prediction does not really change and we could just as well
use the mean model.

```{r, echo=FALSE}
library(pacman)
p_load(plotly, tidyverse, rethinking, patchwork)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]

mod <- lm(height ~ weight, data = d2)
#summary(mod)

#plot(d2$weight, d2$height)

# lets use the 10% and the 90% quantile 
# of height

x_low_high <- quantile(d2$weight, probs = c(0.1, 0.9))

# Scenario 1
# original
#coef(mod)
# Conditional distribution of height is: N(113.8793936  + 0.9050291 *weight, 5.086)

# Draw to distributions for weights = x_low_high into one diagram
# 1) N(113.8793936  + 0.9050291 *x_low_high[1], 5.086)
# 2) N(113.8793936  + 0.9050291 *x_low_high[2], 5.086)

# Parameters for the two normal distributions
mean1 <- 113.8793936 + 0.9050291 * x_low_high[1]
mean2 <- 113.8793936 + 0.9050291 * x_low_high[2]
sd <- 5.086  # Standard deviation for both distributions

# Create a data frame for plotting
x_seq <- seq(mean1 - 4 * sd, mean2 + 4 * sd, length.out = 1000)
data <- data.frame(
  x = rep(x_seq, 2),
  density = c(dnorm(x_seq, mean = mean1, sd = sd),
              dnorm(x_seq, mean = mean2, sd = sd)),
  group = rep(c(paste0("x_low = ", x_low_high[1]),
                paste0("x_high = ", x_low_high[2])), each = length(x_seq))
)

# Scenario 2
# more variability to get a worse R^2
set.seed(123)
d2$height_added_noise <- d2$height + rnorm(nrow(d2), mean = 0, sd = 15)

mod_noise <- lm(height_added_noise ~ weight, data = d2)
#summary(mod_noise)

# Extract the new coefficients
coef_noise <- coef(mod_noise)

# Conditional distribution of height with noise
# N(coef_noise[1] + coef_noise[2] * weight, Residual SD)
residual_sd <- sigma(mod_noise)

# Parameters for the two normal distributions
mean1_noise <- coef_noise[1] + coef_noise[2] * x_low_high[1]
mean2_noise <- coef_noise[1] + coef_noise[2] * x_low_high[2]

# Create a data frame for plotting (with added noise)
x_seq_noise <- seq(mean1_noise - 4 * residual_sd, mean2_noise + 4 * residual_sd, length.out = 1000)
data_noise <- data.frame(
  x = rep(x_seq_noise, 2),
  density = c(dnorm(x_seq_noise, mean = mean1_noise, sd = residual_sd),
              dnorm(x_seq_noise, mean = mean2_noise, sd = residual_sd)),
  group = rep(c(paste0("x_low = ", x_low_high[1]),
                paste0("x_high = ", x_low_high[2])), each = length(x_seq_noise))
)

# Plot for Scenario 1
plot1 <- ggplot(data, aes(x = x, y = density, color = group)) +
  geom_line(linewidth = 1.2) +
  labs(
    title = "Scenario 1",
    x = "Height",
    y = "Density",
    color = "Weight"
  ) +
  theme_minimal() +
  scale_color_manual(values = c("blue", "red")) +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    legend.position = "bottom"
  )

# Plot for Scenario 2
plot2 <- ggplot(data_noise, aes(x = x, y = density, color = group)) +
  geom_line(linewidth = 1.2) +
  labs(
    title = "Scenario 2",
    x = "Height",
    y = "Density",
    color = "Weight"
  ) +
  theme_minimal() +
  scale_color_manual(values = c("blue", "red")) +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14),
    legend.position = "bottom"
  )

# Combine the two plots side by side with a shared legend
combined_plot <- plot1 + plot2 +
  plot_layout(guides = "collect") &  # Collect the legend
  theme(
    legend.position = "bottom",
    plot.title = element_text(hjust = 0.5, size = 16)
  )

# Add a single title to the combined plot
combined_plot <- combined_plot +
  plot_annotation(
    title = "Normal Distributions for Two Scenarios",
    theme = theme(plot.title = element_text(size = 18, hjust = 0.5))
  )

# Display the combined plot
combined_plot
```

In the left plot, given $X = x_{low}$ gives a a rather strongly shifted normal distribution of
potentially observable heights for this weight compared to $X=x_{high}$.
We would have a lower missclassification error when trying to distinguish heights
of very light and very heavy people in the sample just by seeing the height. 
You can think of an even more extreme separation of these distributions,
which would happen, when $R^2$ is very high or the true slope is much higher.

See also [exercise 5](#exercise5_simpl_lin_reg).

An interesting way to look at $R^2$ is the following:
Given, that one person is in the 90% quantile of the weight,
the other is in the 10% quantile. What is the probability that the
height of the person in the 90% quantile is higher than the height of the other person?
We could calculate this relatively easy using theorems about additivity of normal distributions.
Since we are all about application, we simulate this:

```{r}
library(rethinking)
data(Howell1)
d <- Howell1
d2 <- d[d$age >= 18, ]
# Simulate heights for the two quantiles
n_sims <- 10000
set.seed(123)
mod <- lm(height ~ weight, data = d2)
summary(mod)$r.squared
x_low_high <- quantile(d2$weight, probs = c(0.1, 0.9))
x_low_high
mean1 <- 113.8793936 + 0.9050291 * x_low_high[1]
mean2 <- 113.8793936 + 0.9050291 * x_low_high[2]
sd <- 5.086  # Standard deviation for both distributions
simulated_heights <- tibble(
  # conditional normal distribution according to the model
  low_heights = rnorm(n_sims, mean = mean1, sd = sd),
  high_heights = rnorm(n_sims, mean =  mean2, sd = sd)
)

# Calculate the probability that the height of the person in the 90% quantile is higher
# than the height of the person in the 10% quantile
simulated_heights %>%
  mutate(higher_height = high_heights > low_heights) %>%
  dplyr::summarise(mean(higher_height))
```

In other words, we can be almost sure, that a person in the 90% quantile of the weight
is taller than a person in the 10% quantile of the weight given this data set. 
See also [exercise 11](#exercise11_simpl_lin_reg).

### Check regression assumptions
Everytime we fit a model, we should check the assumptions [above](#_model_assumptions).
We do this for different reasons (which will become clearer over the course). 
The assumptions are independent of each other.
They can all be true or false or some can be true and some false 
([Wesftall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4), p.21).
Chapter 4 in the book is dedicated to this topic. It is important to know
that the assumptions are usually not met exactly. The question is *how badly* they are violated,
not *if* they are violated.
Furthermore, the asssumptions refer to the data generating process, not the data itself.
Thus, the evaluation of the assumptions should involve subject matter knowledge.

$p$-values to evaluate model assumptions are not a good idea. To [quote](https://www.tandfonline.com/doi/full/10.1080/00031305.2016.1154108#d1e949) the 
American Statistical Association (ASA): "By itself, a $p$-value does not provide a good measure 
of evidence regarding a model or hypothesis."
Decision-tree thinking might not be the best idea for statistical modeling.

We will not use hypothesis tests for assumptions, because

- They are never met exactly. You cannot "prove" them.
- With small sample sizes, the statistical *power* is often low.
- With large sample sizes, the smallest deviation will be "significant".

We will follow the book (chapter 4, pages 99ff) and use graphical and simulation methods to check the assumptions.
As guidance, we could check them in the following order:

- Linearity
- Constant variance 
- Independence
- Normality

#### Linearity {#linearity}

First, we **plot the data**, as we did already above. We can add a 
a smoothing line to the raw data as well:

```{r, echo=FALSE}
ggplot(d2, aes(x = weight, y = height)) +
  geom_point(alpha = 0.6) +
  geom_smooth(method = "loess", se = FALSE, color = "blue", linewidth = 1) +
  labs(title = "Scatterplot with Regression Line",
       x = "Weight",
       y = "Height") +
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

It looks like this relationship is describable in a linear way.
No apparent curvature or patches.
A refined version of the scatterplot of the raw data is The
**residual scatter plot** ($x_i, e_i$):

```{r, echo=FALSE}
# Calculate residuals
d2$residuals <- residuals(mod)

# Create the residual scatter plot
ggplot(d2, aes(x = weight, y = residuals)) +
  geom_point(alpha = 0.6) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "red") +
  geom_smooth(method = "loess", se = FALSE, color = "blue", linewidth = 1) +
  labs(title = "Residuals vs. Weight",
       x = "Weight",
       y = "Residuals") +
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

Compared to the scatterplot above, the residuals plot magnifies
possible curvature. The reason is that the range of residuals
is smaller than the range of the heights. 

In multiple regression, we will use the ($\hat{y_i}, e_i$) plot, which is
identical to the  plot above in simple linear regression, but
very helpful in multiple regression.
You get the ($\hat{y_i}, e_i$) plot in R with `plot(mod, which = 1)`.


#### Constant variance {#constant_variance}

This assumption means that the variance of the residuals is constant.
If it is violated, the spread around a hypothetical regression line is not constant.
We look at the residual scatter plot again:

```{r, echo=FALSE}
plot(mod, which = 1)
```

Look for a changes in patterns of vertical variability. Note, that you
often do not have as many data points near the end of the range of the predictor.
Here are some examples of heteroskedasticity:
[1](https://i0.wp.com/statisticsbyjim.com/wp-content/uploads/2017/08/residuals_unfixed.png?resize=576%2C384),
[2](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQPAPMGV5U0Pe0nKnhoZWHJ8KypqVCls6ZQZud84C3KM8SJqOkuMNeJl2oTg-UDuukJRhk&usqp=CAU),
[3](https://i.sstatic.net/R5KIH.png). 
The above looks homoscedastic. If it was heteroskedastic, this is not a problem,
we just have to model it differently (later).

As always, it is a good idea to study the variability of these plots
using simulation ([exercise 7](#exercise7_simpl_lin_reg)). 

Better for detecting heteroscedasticity is the ($\hat{y_i}, |e_i|$) plot
with a smoothing line:

```{r, echo=FALSE}
library(rethinking)
data("Howell1")
d <- Howell1
str(d)
d2 <- d[d$age >= 18, ] # only adults
mod <- lm(height ~ weight, data = d2)
predicted_height <- predict(mod)
residuals <- residuals(mod)
ggplot(d2, aes(x = predicted_height, y = abs(residuals))) +
  geom_point(alpha = 0.6) +
  geom_smooth(method = "loess", se = FALSE, color = "blue", linewidth = 1) +
  labs(title = "Residuals vs. Predicted Height",
       x = "Predicted Height",
       y = "Absolute Residuals") +
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

This will probably not be a perfectly horizontal line, since
there are less points at the end of the range of the predictor.
For instance, There are less people with extreme weights in the data set.

#### Independence, uncorrelated errors

We had an example above, where the errors were correlated. 
The sine curve could stem from time series data, where the $x$-variable 
is the time and the $y$-variable is a seasonally changing variable
like temperature (purely hypothetical). In [exercise 6](#exercise6_simpl_lin_reg),
the values are autocorrelated: Correlated with previous values of the same variable.
If we would track the body heights of persons over time, we would have
an autocorrelated time series since the height of a person at time $t$ is
correlated with the height of the same person at time $t-1$, but a little
less with the height at time $t-2$ and so on. If I am tall today, it is 
very likely that I was tall yesterday.

In our case of the !Kung San data, we do not have autocorrelated data.
Just because the first person we drew is tall, does not mean that the second
person we draw is tall. We could draw a very young person for instance.

But still, we could look at the correlations the residuals with lagged residuals. 
A $lag=1$ means I compare the residuals with the ones right next to me and check if
they are correlated. 

```{r, echo=FALSE, warning=FALSE}
library(rethinking)
data("Howell1")
d <- Howell1
#str(d)
d2 <- d[d$age >= 18, ] # only adults
mod <- lm(height ~ weight, data = d2)
predicted_height <- predict(mod)
residuals <- residuals(mod)
residuals_lagged <- c(NA, residuals[1:(length(residuals)-1)])
cat("cor=",cor(residuals, residuals_lagged, use = "pairwise.complete.obs"))
plot(residuals, residuals_lagged)
```

At least with a lag of $1$, there so no large correlation between the residuals.

#### Normality {#normality_assumption}

This assumptions states that the conditional distribution $Y|X=x$ 
is a normal distribution. We do not look at the normality of $Y$ itself,
since it is not a formal requirement, that $Y$ is normally distributed.
It might well occur that the conditional distributions are all normal,
but the variance is not constant. Then we would have heteroskedasticity,
but the assumption of normality would be fullfilled.

We need to assess the normality of the residuals $e_i = y_i - \hat{y}_i$
I like doing this with a [Q-Q plot](https://jdegenfellner.github.io/Script_QM1_ZHAW/descriptive_stats.html#q-q-plots) 
from the R package `car` (command `qqPlot`):

```{r, echo=FALSE, warning=FALSE}
library(rethinking)
library(car)
data("Howell1")
d <- Howell1
d2 <- d[d$age >= 18, ] # only adults
mod <- lm(height ~ weight, data = d2)
qqPlot(residuals(mod), envelope = 0.97)
``` 

With the 97% confidence envelope, we can see if the residuals are
consistent with coming from a normal distribution.
We could again use simulation to see how the Q-Q Plot changes 
to get a better feeling (see [exercise 9](#exercise9_simpl_lin_reg)).

Another convenient way to check model assumptions (for a wide class of models) 
is to use the `check_model` function from the `performance` package:

```{r, echo=FALSE}
library(performance)
check_model(mod)
```

In this case, everything looks fine. Let`s go through the plots and explain them one by one:

##### Posterior predictive check (upper left)

Here, you create new data from the estimated model. We did that already in
the Bayesian framework using `sim` from the `rethinking` package.
You could also try the command `extract.samples()`from the `rethinking` package 
(Statistical Rethinking p. 196) to create
new data from the Frequentist model. You can try this in our exercises sessions. 
Using least squares, the estimated model was:

$$height_i = 113.87939 + 0.90503 \cdot weight_i + Normal(0, 5.086)$$

From this model, we can simulate new data (or the original weights, if we want to use fixed $X$) 
by plugging in different weights, and adding a random error from a normal distribution 
with mean 0 and standard deviation 5.086.
Then, we can compare the distribution of the simulated data with the observed data.
The blue lines in the graph (output from `check_model`) are density 
estimations of model predicted heights, the green line is the density estimate of the
observed heights. 

Let`s try to replicate this plot.
First, we want to simulate new data from the model.
A scatter plot is another nice way to check if model predictions
are in line with the observed data. One could repeat this process
multiple times to get a feeling for the variability of the model predictions.

```{r, echo=FALSE}
# Create new data
new_data <- data.frame(weight = d2$weight)
new_data$height <- rnorm(nrow(new_data), 
                   mean = coef(mod)[1] + coef(mod)[2] * new_data$weight, sd = sigma(mod))

# Create the plot
ggplot(new_data, aes(x = weight, y = height)) +
  geom_point(alpha = 0.6) +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linewidth = 1) +
  geom_point(data = d2, aes(x = weight, y = height), color = "green", alpha = 0.6) +
  labs(title = "Scatter plot of observed and simulated data",
       x = "Weight",
       y = "Height") +
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

The simulated and original data (green) fit nicely together. 
This lends some credibility to the model - at least in terms of prediction.
And now the density plots of the model created data (blue) and the observed data (green):

```{r, echo=FALSE}

# Set up an empty plot area
plot(NULL, xlim = range(d2$height), ylim = c(0, 0.07), 
     xlab = "Height", ylab = "Density", 
     main = "Posterior Predictive Density Plot")

# Simulate 10 new datasets and overlay densities
for (i in 1:10) {
  # Generate new simulated data
  new_heights <- rnorm(
    n = nrow(d2),
    mean = coef(mod)[1] + coef(mod)[2] * d2$weight,
    sd = sigma(mod)
  )
  
  # Add density for simulated data in light blue
  lines(density(new_heights), col = rgb(0.2, 0.6, 1, alpha = 0.2), lwd = 1)
}

# Add density for the original data in green
lines(density(d2$height), col = "green", lwd = 2)

# Add a legend
legend("topright", legend = c("Observed", "Simulated"),
       col = c("green", rgb(0.2, 0.6, 1, alpha = 0.5)), 
       lwd = c(2, 2), bty = "n", cex = 0.8)
```

As you can see, the densities of the observed and simulated data are broadly similar.
One could argue that heights in the range of 150-160 cm are a bit overestimated by the model.
Depending on the context, this might be a problem or not. Let`s accept the model for now.

##### Linear fit (upper right)
This is the same plot as [above](#linearity): $(\hat{y}_i, e_i)$.

##### Homogeneity of variance (middle left)
This is the same plot as [above](#constant_variance): $(\hat{y}_i, |e_i|)$.

##### Influential observations (middle right)
The standard method to check for influential observations is to compute the
so-called [Cook's distance](https://en.wikipedia.org/wiki/Cook%27s_distance). 
This is a measure of how much leverage a single observation has on the model.
We can extract the Cook's distance from the model using `cooks.distance()`.
An ugly rule of thumb would be to look at observations with a Cook's distance
greater than 1.
In the plot, the leverage $h_{ii}$ is on the x-axis, and the standardized residuals
are on the y-axis. In short, a high leverage means that the estimated value $\hat{y}_i$
is potentially far away from the original value $y_i$. 
The contours (dotted green lines) are the Cook's distance, in this 
case at a Cook's distance of 0.5.
There is a [formula](https://en.wikipedia.org/wiki/Cook%27s_distance#Relationship_to_other_influence_measures_(and_interpretation))
that relates the leverage ($h_{ii}$), the Cook's distance
and the residuals. Holding Cook`s distant constant at 0.5, 
gives you the green dotted line in the plot.

Let`s create a nasty outlier and see what happens:

```{r, echo=FALSE}
library(rethinking)
library(car)
data("Howell1")
d <- Howell1
d2 <- d[d$age >= 18, ] # only adults
mod <- lm(height ~ weight, data = d2)

# Lets's create an outlier and see what happens,
# maybe an outlier in x and y

# maybe 3 sd above the mean of weight
# and 20 sd below the mean of height
# this should tilt our regression line nicely

#dim(d2)
d2_copy <- d2
d2_copy$height[352] <- d2_copy$height[352]  - 20 * sd(d2_copy$height)
d2_copy$weight[352] <- d2_copy$weight[352]  + 3 * sd(d2_copy$weight)
mod2 <- lm(height ~ weight, data = d2_copy)
check_model(mod2)

d2_copy %>% ggplot(aes(weight, height)) + 
  geom_point() + 
  geom_smooth(method = "lm")

cat("Cook's distance = ",cooks.distance(mod2)[352])
```

One single outlier changes the diagnostic plots. Observation 352
is clearly identified as influential. The one point
changes all diagnostic plots notably.
The estimates of the regression coefficients
are also affected:

```{r, echo=FALSE}
print("Original Model")
coef(mod)
print("Model with outlier")
coef(mod2)
```

Admitted, this is a somewhat artificial example.

##### Normality of residuals (lower left)
This is basically the same plot as [above](#normality_assumption), just [detrended](https://cran.r-project.org/web/packages/qqplotr/vignettes/introduction.html). 
Let's try to replicate it:

```{r, echo=FALSE, warning=FALSE}
library(ggplot2)
library(qqplotr)

# Prepare the data
df <- data.frame(
  residuals = scale(mod$residuals) # Standardize the residuals to match `check_model()`
)

# Create the detrended QQ plot
ggplot(data = df, mapping = aes(sample = residuals)) +
  stat_qq_band(distribution = "norm", 
               bandType = "boot",
               detrend = TRUE, 
               fill = "grey", 
               alpha = 0.3) + # Confidence band
  stat_qq_line(distribution = "norm", 
  detrend = TRUE, color = "green", linewidth = 0.8) + # Detrended line
  stat_qq_point(distribution = "norm", 
  detrend = TRUE, color = "blue", linewidth = 1.5) + # Points
  labs(
    title = "Normality of Residuals",
    subtitle = "Dots should fall along the line",
    x = "Standard Normal Distribution Quantiles",
    y = "Sample Quantile Deviations"
  ) +
  coord_cartesian(ylim = c(-2, 2)) + # Match the y-range to the `check_model()` plot
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  theme(plot.subtitle = element_text(hjust = 0.5))
```

The normality assumption seems to hold. Note that in the above QQ plot, 
we detrended the residuals and used a bootrap confidence band. Depending 
on the band type, the confidence band can be wider or narrower and include
or exclude points.

**After having checked the assumptions for the classical regression model**
and we feel comfortable with the model, we get exact confidence intervals
for the effect sizes ($\beta$s) using `confint()` (p. 74 in Westfall).

**Heureka!** We have a model that fits the data well.

### Bootstrap fit
In order to get a feeling for the variability of the model with regards to the predictors as well, 
we can bootstrap the whole data set, fit the model and draw regression lines.
We create $100$ bootrap replicates of our data set `d2` by drawing with 
repetition. For every bootstrap replicate, we fit the model and draw the regression line.

```{r, echo=FALSE, warning=FALSE}
# Required Libraries
library(ggplot2)

# Define number of bootstrap replicates
n_bootstrap <- 100

# Create a list to store bootstrap models
bootstrap_models <- vector("list", n_bootstrap)

# Bootstrap and fit the model
set.seed(123) # For reproducibility
for (i in 1:n_bootstrap) {
  # Sample with replacement
  bootstrap_data <- mod$model[sample(nrow(mod$model), replace = TRUE), ]
  
  # Fit the model to the bootstrap sample
  bootstrap_models[[i]] <- lm(as.formula(mod$call), data = bootstrap_data)
}

# Extract regression lines
bootstrap_lines <- lapply(bootstrap_models, function(model) {
  data.frame(
    x = mod$model[[2]], # Predictor variable
    y = predict(model, newdata = mod$model) # Predicted values
  )
})

# Combine all bootstrap lines into a single data frame
bootstrap_lines_df <- do.call(rbind, lapply(seq_along(bootstrap_lines), function(i) {
  data.frame(x = bootstrap_lines[[i]]$x, y = bootstrap_lines[[i]]$y, replicate = i)
}))

# Scatterplot of the original data with bootstrap regression lines
ggplot(mod$model, aes(x = .data[[names(mod$model)[2]]], y = .data[[names(mod$model)[1]]])) +
  geom_point(color = "blue", alpha = 0.7) + # Scatterplot points
  geom_line(data = bootstrap_lines_df, aes(x = x, y = y, group = replicate), 
            color = "grey", alpha = 0.3, linewidth = 0.5) + # Bootstrap regression lines
  labs(
    title = "Bootstrap Regression Lines",
    x = "Predictor",
    y = "Response"
  ) +
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

The result is very stable. Neither intercept nor slope change much.

### Regression towards the mean

There are great explanations for "[regression towards the mean](https://en.wikipedia.org/wiki/Regression_toward_the_mean)" in 
[Gelman](http://www.stat.columbia.edu/~gelman/arm/) p.58 and 
[Westfall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4) p.36.
This [video](https://www.youtube.com/watch?v=1tSqSMOyNFE&ab_channel=Veritasium) might be interesting to watch.

It describes the phenomenon that the predicted value (y) is closer to its 
mean than the predictor (x) to its mean.
In our case this means, if the weight of a person is 1 standard deviation above the mean of body weights,
the (model-)predicted height is less than 1 standard deviation above the mean of body heights, but
still larger than the average height. Let's verify:

```{r, echo=FALSE, warning=FALSE}
# Standardize the data
d2$weight_scaled <- scale(d2$weight)
d2$height_scaled <- scale(d2$height)

mod <- lm(height_scaled ~ weight_scaled, data = d2)
cat("predicted height=", as.numeric(coef(mod)[1] + coef(mod)[2] * 1))

# Compute principal components
pca <- prcomp(d2[, c("weight_scaled", "height_scaled")], scale. = FALSE)

# Extract principal component directions
pc1 <- pca$rotation[, 1] * pca$sdev[1]
pc2 <- pca$rotation[, 2] * pca$sdev[2]

# Adjust scaling factor for visualization
scale_factor <- 2  # Adjust as necessary for visual clarity

# Updated ggplot with properly scaled and aligned PC arrows
ggplot(d2, aes(x = weight_scaled, y = height_scaled)) +
  geom_point(alpha = 0.6) +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linewidth = 1) +
  geom_point(data = data.frame(weight_scaled = 1, height_scaled = coef(mod)[1] + coef(mod)[2] * 1),
             aes(x = weight_scaled, y = height_scaled),
             color = "red", size = 3) +
  # Line from the red point to the x-axis
  geom_segment(aes(x = 1, xend = 1, y = 0, yend = coef(mod)[1] + coef(mod)[2] * 1), 
               color = "red", linetype = "dotted") +
  # Line from the red point to the y-axis
  geom_segment(aes(x = 0, xend = 1, y = coef(mod)[1] + coef(mod)[2] * 1, yend = coef(mod)[1] + coef(mod)[2] * 1), 
               color = "red", linetype = "dotted") +
  # Add principal component 1
  geom_segment(aes(x = 0, xend = scale_factor * pc1[1], 
                   y = 0, yend = scale_factor * pc1[2]),
               color = "green", linetype = "dotted", arrow = arrow(length = unit(0.2, "cm"))) +
  # Add principal component 2
  geom_segment(aes(x = 0, xend = scale_factor * pc2[1], 
                   y = 0, yend = scale_factor * pc2[2]),
               color = "green", linetype = "dotted", arrow = arrow(length = unit(0.2, "cm"))) +
  geom_vline(xintercept = 0, color = "black", linetype = "dashed") +
  geom_hline(yintercept = 0, color = "black", linetype = "dashed") +
  labs(title = "Regression towards the mean with Principal Components",
       x = "Weight (standardized)",
       y = "Height (standardized)") +
  coord_fixed(ratio = 1) +  # Fix the aspect ratio to ensure equal scaling
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

As Gelman points out in his book (Figure 4.2), there is a fine detail:
The regression line is not the line most people would draw through the data.
They would draw a line through the main directions of variability 
(directions are the eigenvectors of the covariance matrix).
In dotted green, you can see these directions. The regression line is a solution 
to another problem (as we have seen): It minimizes the sum of squared residuals, 
which are defined via the **vertical** distances to the line.
See [exercise 8](#exercise8_simpl_lin_reg).

### Random X vs. fixed X

A detail that is not mentioned often in introductory statistics courses
is the question of whether the predictor variable $X$ is random or fixed.
In our case, we did not specify the weights of the people in the !Kung San data set.
*Obervational* data was collected and we have no control over the weights.
In an *experimental* setting, we could have controlled the weights of the people.
We could have for instance only included people with weights at certain steps (50 kg, 60 kg, 70 kg).
In the latter case, we could consider X as fixed. In the former case, we consider X as random.
If we would draw another sample from the population, we would get different weights again.
Further reading: [Westfall](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4) 1.5.

There are many more details we could look into, but we wanted to give a first 
practical introduction to regression analysis with less emphasis on theory and proofs behind it.

We want to make clear what the confidence intervals from `confint` of a model mean. For this see
[exercise 14](#exercise14_simpl_lin_reg).

## Exercises

[E] Easy, [M] Medium, [H] Hard

(Some) solutions to exercises can be found in the git-repo [here](https://github.com/jdegenfellner/Script_QM2_ZHAW/tree/main/Solutions_Exercises).

### [E] Exercise 1 {#exercise1_simpl_lin_reg}

In the model from above:

\begin{eqnarray*}
h_i &\sim& \text{Normal}(\mu_i, \sigma)\\
\mu_i &\sim& \alpha + \beta (x_i - \bar{x})\\
\alpha &\sim& \text{Normal}(171.1, 20)\\
\beta &\sim& \text{Normal}(0, 10)\\
\sigma &\sim& \text{Uniform}(0, 50)
\end{eqnarray*}

- What ist the expected height when $x_i = \bar{x}$?
- What is the expected height when $x_i$ changes by 1 unit?

### [E] Exercise 2 {#exercise2_simpl_lin_reg}

Look at the marginal distrubutions of the parameters in the Bayesian model.

- Plot the posterior distribution of all 3 parameters.
- Include in the plot a 99% credible interval (HDI).

### [H] Exercise 3 {#exercise3_simpl_lin_reg}

Go to the coefficient estimates in the simple linear regression setting 
above ([Fit the model](#fit_model_simple_lin_reg_classic))
in the classical framework.

- Simulate data from the simple linear regression model 
  using the estimated coefficients ($\hat{\beta_0} = 113.87939$,
  $\hat{\beta_1} = 0.90574$, $\hat{\sigma} = 5.086$).
- Use the formulae from above to calculate $\beta_0$ and $\beta_1$ by "hand".
- Change your true parameters ($\beta_0, \beta_1, \sigma$) and see how the 
  estimates for the standard deviation of simulated heights, weights
  and the correlation between simulated heights and weights change.
- Create a scatterplot to see the data as well.

### [H] Exercise 4 {#exercise4_simpl_lin_reg}

Verify the statement above in the text for high and low values of $R^2$.

### [H] Exercise 5 {#exercise5_simpl_lin_reg}

Verify with simulation in R that the separation of the distributions in the simple linear regression model
improves if the true (but usually unknown) slope increases.
The code for the plots can be found 
[here](https://github.com/jdegenfellner/Script_QM2_ZHAW/blob/main/02-Simple_Linear_Regression.Rmd).
Hint: Use GPT for coding.

### [H] Exercise 6 {#exercise6_simpl_lin_reg}

Go to the model assumptions in the classical regression model ([Model Assumptions](#_model_assumptions)).

- Use the code from [github](https://github.com/jdegenfellner/Script_QM2_ZHAW/blob/main/02-Simple_Linear_Regression.Rmd) 
  to recreate the regression model with the sine-curve.

- Use the autocorrelation function `acf` to check for autocorrelation in the residuals.

### [M] Exercise 7 {#exercise7_simpl_lin_reg}

- Simulate data from the regression of heights on weights in our !Kung San data set.
- Draw the $\hat{y_i}, e_i$ plot.
- Draw the $\hat{y_i}, |e_i|$ plot.
- Repeat the simulation and look at the variability of the plot.

### [E] Exercise 8 {#exercise8_simpl_lin_reg}

- Go to p.36 in Westfall's book and read Appendix A.
- Pay close attention to the explanation about regression toward the mean.

### [M] Exercise 9 {#exercise9_simpl_lin_reg}

Go to the residuals [above](#normality_assumption) where we tested the normality assumption.

- Calcuate mean and standard deviation from the residuals of the model
  that regresses height on weight.
- Simulate from a normal distribution using these parameters.
- Get a feeling how the QQ plot changes by drawing the QQ plot repeatedly.

### [M] Exercise 10 {#exercise10_simpl_lin_reg}

Using our !Kung San data, 

- show that the regression of height on weight (`lm(height ~ weight)`)
  is not the same as the regression of weight on height (`lm(weight ~ height)`).
- Draw both regression lines in one diagram.
- Can you simulate data where the two regressions deliver (almost) identical results?
- Explain why the results differ and what consequences this would have for a research question.
  Which question do I answer with each?

### [E] Exercise 11 {#exercise11_simpl_lin_reg}

Go back to the [section](#seperating_property) about $R^2$ and the separation of the distributions.
How would the probability that a person in the 90% quantile of the weight
is taller than a person in the 10% quantile of the weight change if you change 

- the true slope of the regression line
- the true $\sigma$, i.e. if you add more noise and have a lower $R^2$?

### [M] Exercise 12 {#exercise12_simpl_lin_reg}

Go back to the Bayesian simple regression model for height [above](#simple_lin_reg_bayes).

- Standardize weights and heights in the data set `d2`.
- Estimate the regression model as we did in the section using `quap`.
- Plot the posterior distribution using `plot(model)`.
- Interpret the regression coefficient $\beta$.

### [M] Exercise 13 {#exercise13_simpl_lin_reg}

Go back to the [ANOVA section](#analysis_of_variance).

- Calculate SSE, SST and SSR for the regression of height on weight.

- How many degrees of freedom does each term have?

### [M] Exercise 14 {#exercise14_simpl_lin_reg}

When we fit a simple linear regression model (Frequentist), we get confidence intervals for the coefficients
using the command `confint`. As stated previously, these are Frequentist CIs. That means,
if we draw data from the true (but usually unknown) data generating process, we would expect
that the true value of the coefficient (intercept and slope) lies in the CI in X% of the cases.

- Simulate data from the true data generating process of the simple linear regression model.
- Fit the model and get the confidence intervals for the coefficients.
- Repeat this process 1000 times and see how many times the true value of the coefficients
  lies in the CI.

Changing the distribution of the covariate $x$ should not change the results, since
we only get a different amount of data points agross the range of $x$.

- Verify this statement by changing the distribution of $x$.


## eLearning 1

eLearning Assignments:

-	Review all content from the script, up to and including section 3.1.
- Reading assignment on frequentist statistics (exam-relevant!): 
  Chapter 1 (Introduction to Regression Models) from the book 
  Understanding Regression Analysis - A Conditional Distribution Approach.
- Research task: Find at least two scientific articles in your field that apply 
  regression (either simple regression with one predictor or multiple regression). 
  We have already searched for journals relevant to your field in QM1. 
  If you send me one of these studies in advance, we could also discuss the 
  applied methods in more detail during the practice sessions.

<!--chapter:end:02-Simple_Linear_Regression.Rmd-->

# Multiple Linear Regression

So far, we have dealt with the simple mean model and the model with one predictor
in the Bayesian and Frequentist framework.
We will now add another predictor and subsequently an interaction term to the model.
Finally, we will add more than two predictors to the model.

If you feel confused at any point: As Richard McElreath repeatedly says:
This is normal, it means you are paying attention. I also refer to the
great [Richard Feynman](https://www.youtube.com/watch?v=lytxafTXg6c&ab_channel=sdfhsfh).

## Linear Regression with 2 predictors in the Bayesian Framework

### Meaning of "linear"
What is a linear model? The term "linear" refers to the relationship of the predictors 
with the dependent variable (or outcome). The following model is also linear:

$$height_i = \beta_0 + \beta_1 x_i + \beta_2 x_i^2$$

The model is linear in the parameters $\beta_0, \beta_1, \beta_2$ but not in the predictors $x_i$.
The term $x_i^2$ is ok, since the heights are just sums of multiples of the predictors (which can be nonlinear).
This model is not a linear model anymore:

$$height_i = \beta_0 + \beta_1 x_i + e^{\beta_2 x_i^2}$$

$\beta_2$ is now is the exponent of $e$. It would also not be linear,
if the coefficients are in a square root or in the denominator of a fraction,
or in a sine or in a logarithm. You get the idea.

Here is an easy way to check if the model is linear: If I change the predictor-value 
(i.e.,the value of $x_i$, $x_i^2$ or whatever your predictor is) by one unit,
the change in the (expected value of the) dependent variable is the coefficient in front of 
the predictor ($\beta_i$).

### Adding a transformed predictor to the model {#adding_transformed_predictor_bayes}

Around 4.5. in the book [Statistical Rethinking](https://civil.colorado.edu/~balajir/CVEN6833/bayes-resources/RM-StatRethink-Bayes.pdf)
there is are lineare regression using a quadratic term for weight. 
It is a principle, called the "**variable inclusion principle**", that we always include the lower order terms when fitting a model
with higher order terms. See [Westfall](https://vdoc.pub/documents/understanding-regression-analysis-a-conditional-distribution-approach-84oqjr8sqva0), 
p. 213. If we do not include the lower order terms, the coefficient does not measure what
we want it to meausure (curvature in our case). For instance, if we want to model a quadratic relationship (parabola) between
weight and height, we also have to include the linear term for weight ($x_i$).
Since we do not assume the relationship between weight and height to be linear but
quadratic (which is a polynomial of degree 2), we call this a 
[polynomial regression](https://en.wikipedia.org/wiki/Polynomial_regression#:~:text=In%20statistics%2C%20polynomial%20regression%20is,nth%20degree%20polynomial%20in%20x.).
This [video](https://www.youtube.com/watch?v=QptI-vDle8Y&ab_channel=MikeXCohen) could be instructive. 
One has to be careful with fitting polynomials to data points since the regression coefficients
can become quite large. Using a polynomial of high degree implies to have a lot of parameters 
to estimate. Increasing the degree of the polynomial increases $R^2$ but also the risk of overfitting.
(see Statistical Rethinking p. 200). So this is - of course - not the final solution to regression problems.

This time, lets look at the whole age range of data from the !Kung San people
including people with age $\le 18$ years.

```{r}
library(rethinking)
library(tidyverse)
data(Howell1)
d <- Howell1
d %>% ggplot(aes(x = weight, y = height)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue") +
  geom_smooth(method = "loess", se = FALSE, color = "red")
```

It would not be a good idea to fit a linear trend through this data,
because we would not caupture the relationship adequately. 
The red line is a [loess smothing](https://en.wikipedia.org/wiki/Local_regression) line
which is often used to capture non-linear relationships.
The blue line is the usual line from classic linear regression (from the previous chapter).
Which one describes the data more accurately? 
In this case it is obvious, a non-linear relationship is present and it might be a good idea 
to model it. Modeling the relationship with a linear trend leads to bad residuals with structure.
We will demonstrate this in the Frequentist setting.
Unfortunately, in more complex settings, with more predictors, it is not always so easy to see.

This time, we use the mean for the prior from the book ($178 cm$).
The model equations are (see [exercise 2](#exercise2_multiple_regression)):

\begin{eqnarray*}
h_i &\sim& \text{Normal}(\mu_i, \sigma) \\
\mu_i &=& \alpha + \beta_1 x_i + \beta_2 x_i^2 \\
\alpha &\sim& \text{Normal}(178, 20) \\
\beta_1 &\sim& \text{Log-Normal}(0, 1) \\
\beta_2 &\sim& \text{Normal}(0, 1) \\
\sigma &\sim& \text{Uniform}(0, 50)
\end{eqnarray*}

The prior for $\beta_1$ is log-normal, because we can reasonably assume
the the overall linear trend is positive. 
The prior for $\beta_2$ is normal, because
we are not so sure about the sign yet. If we thought back to our school days to the topic of 
"curve discussion" or parabolas, we could probably also assume that $\beta_2$ is negative.
But, data will show.

How can we interpret the model equations? 
The model assumes that the **expected** height $\mu_i$ of a person $i$
depends non-linearly (quadratically) on the (standardized) weight $x_i$ of the person.
We are in the business of mean-modeling. 
The prior for $\sigma$ is uniform as before.
The prior for $\alpha$ is normal with mean $178$ and standard deviation $20$
because this is what we can expect from body heights in our experience.

Let's **fit the model**:

We standardize the weight again and add the squared weights to the data set.
Standardizing the predictors is a good idea, especially in polynomial regression
since squares and cubes of large numbers can get huge and cause numerical problems.

Let's fit the model with the quadratic term for weight:

```{r}
# Standardize weight
d$weight_s <- (d$weight - mean(d$weight)) / sd(d$weight)
# Square of standardized weight
d$weight_s2 <- d$weight_s^2
m4.1 <- quap(
  alist(
    height ~ dnorm(mu, sigma),
    mu <- a + b1*weight_s + b2*weight_s^2,
    a ~ dnorm(178, 20),
    b1 ~ dnorm(0, 10),
    b2 ~ dnorm(0, 10),
    sigma ~ dunif(0, 50)
  ), data = d)
precis(m4.1)
plot(precis(m4.1, pars = c("b1", "b2", "sigma")))
```

$\beta_2$ is indeed negative. 
In the estimate-plot above I have left out the intercept $\alpha$ to 
make the other coefficients more visible. As you can see, the credible intervals
are very tight. Using prior knowledge and data, the model is very sure about the 
coefficients.
We get our **[joint distribution](https://en.wikipedia.org/wiki/Joint_probability_distribution)** 
of the **four model parameters**.
Let's look at the fit using the mean estimates of the posterior distribution:

```{r}
# Summarize the model parameters
model_summary <- precis(m4.1)
params <- as.data.frame(model_summary)

# Extract parameter values
a <- params["a", "mean"]       # Intercept
b1 <- params["b1", "mean"]     # Coefficient for standardized weight
b2 <- params["b2", "mean"]     # Coefficient for squared standardized weight

# Generate a sequence of standardized weights for the fitted curve
weight_seq <- seq(min(d$weight_s), max(d$weight_s), length.out = 200)

# Calculate the fitted values using the quadratic equation
height_fitted <- a + b1 * weight_seq + b2 * weight_seq^2

# Plot the scatterplot
plot(d$weight_s, d$height, pch = 16, col = "blue",
     xlab = "Standardized Weight", ylab = "Height (cm)",
     main = "Scatterplot with Fitted Curve (Standardized Weight)")

# Add the fitted curve
lines(weight_seq, height_fitted, col = "red", lwd = 2)

# Add a legend
legend("topright", legend = c("Observed data", "Fitted curve"),
       col = c("blue", "red"), pch = c(16, NA), lty = c(NA, 1), lwd = 2)

       # ======== Simulate Heights from Posterior ========

# Prepare new data with the same number of rows as the original data
new_data <- data.frame(weight_s = d$weight_s, weight_s2 = d$weight_s2)

# Simulate height values from the posterior (same number as original data)
sim_heights <- sim(m4.1, data = new_data, n = nrow(d))  # Posterior samples

# Extract random samples from simulated heights
height_samples <- apply(sim_heights, 2, function(x) sample(x, 1))

# ======== Plot Observed vs. Simulated Heights ========

# Plot observed data
plot(d$weight_s, d$height, pch = 16, col = "blue",
     xlab = "Standardized Weight", ylab = "Height (cm)",
     main = "Observed vs. Simulated Heights")

# Add simulated height points
points(d$weight_s, height_samples, pch = 16, col = "red")

# Add legend
legend("topright", legend = c("Observed Data", "Simulated Data"),
       col = c("blue", "red"), pch = c(16, 16))

```

The second plot shows the original heights and simulated heights from the 
posterior distribution in one plot. This fits quite well.

The quadratic model fits much better than the linear model without the quadratic term. 
In the book, there is also a polynomial regression with a cubic term for weight.
Maybe this fits even better (see [exercise 1](#exercise1_multiple_regression)).

The quadratic model has a very high $R^2$ value. In this case,
this is a good sign. The height values "wiggle" around the fitted curve
very narrowly. We do not need other variables to explain the heights.
In short: If you know a person's weight, you can predict the height very well
(in this population at least). One might hypothesize that the population
behaves rather similarly with respect to nutritional and life style factors.
If different groups within the !Kung San people would have high calorie intake
and few physical activity, the model would probably not fit so well.

### Adding another predictor to the model {#adding_predictor_bayes}

Since the !Kung San data set has already such a high $R^2$ ($=0.95$!) with the quadratic term 
(and possibly higher with the cubic term), we will use the created data set from 
[below](#adding_predictor_freq) in the frequentist setting
to estimate the coefficients of the model with two predictors here as well.
We *have* the true but (usually) unknown data generating 
mechnisms (for didactic reasons).

We use rather uniformative priors and fit the model using `quap`:


```{r}
library(rethinking)
set.seed(123)
n <- 100
X1 <- rnorm(n, 0, 5)
X2 <- rnorm(n, 0, 5)
Y <- 10 + 0.5 * X1 + 1 * X2 + rnorm(n, 0, 2) # true model
df <- data.frame(X1 = X1, X2 = X2, Y = Y)

# fit model
m4.2 <- quap(
  alist(
    Y ~ dnorm(mu, sigma),
    mu <- a + b1*X1 + b2*X2,
    a ~ dnorm(10, 10),
    b1 ~ dnorm(0, 10),
    b2 ~ dnorm(0, 10),
    sigma ~ dunif(0, 50)
  ), data = df)
precis(m4.2)
```


#### Checking model assumptions {#check_model_bayes}
Andrew Gelman mentions in some of his talks (see [here](https://sites.stat.columbia.edu/gelman/research/published/philosophy_chapter.pdf) for more details) 
that many Bayesians he met do not check their models, since they reflect subjective probability.
As I said in the introduction, one should not be afraid to check model predictions against the observed
and probably new data. If a model for predicting BMI performs much worse on a new data set,
we should adapt. We **do not ask** the question if a model is true or false, but if it is useful or
how badly the model assumptions are violated.

For further, more detailed information on model checking, refer to chapter 6 of 
[Gelman's book](https://sites.stat.columbia.edu/gelman/book/BDA3.pdf).

Anyhow, we plot two posterior predictive checks here.
We test the model within the same data set. 
In order to do this, we create new observations
by drawing from the posterior distribution and compare these with the acutally 
observed values. This is called **posterior predictive checks**.

**First, we plot the observed $Y$ values against the predicted $Y$ values ($=\hat{Y}$)** 
from the model (as in Statistical rethinking, Chapter 5).
Although these practically never lie on the line $y=x$, they should be sufficiently 
close to it. We could also compare these two plots with the mean-model (see [exercise 6](#exercise6_multiple_regression)).

```{r}
# 1) Posterior predictive checks Y vs Y_hat
# see Statstical Rethinking p 138.
# call link without specifying new data
# so it uses the original data
mu <- link(m4.2)

# summarize samples accross cases
mu_mean <- apply(mu, 2, mean)
mu_PI <- apply(mu, 2, PI, prob = 0.89)

# simulate observations
# again, no new data, so uses original data
D_sim <- sim(m4.2, n = 1e4)
D_PI <- apply(D_sim, 2, PI, prob = 0.89)

plot(mu_mean ~ df$Y, col = rangi2, ylim = range(mu_PI), 
     xlab = "Observed Y", ylab = "Model-Predicted Y")
abline(a = 0, b = 1, lty = 2)
for(i in 1:nrow(df)) lines(rep(df$Y[i], 2), mu_PI[,i], col = rangi2)
```

As we can see, the model fits the data quite well. The points are close to the dashed line ($y=x$).
No under- or overestimation is visible. The model seems to capture the relationship between the predictors $X_1$ and $X_2$
and the dependent variable $Y$ quite well - at least in a predictive sense. If there were patches of data points above or below the dashed line,
we would probably have to reconsider the model definition and think about why these points are not captured by the model.

**Next, we plot the posterior predictive plots** analog to the upper left in the `check_model` output.

```{r}
library(scales)  # For the alpha function to adjust transparency

# 2) Posterior predictive densities
# Simulate observations using the posterior predictive distribution
D_sim <- sim(m4.2, n = 1e4)  # Generate 10,000 simulated datasets

# Calculate densities for all samples
densities <- apply(D_sim, 1, density)

# Find the maximum density value for setting the y-axis limits
max_density <- max(sapply(densities, function(d) max(d$y)))

# Create the density plot with predefined ylim
plot(NULL, xlim = range(df$Y), ylim = c(0, max_density),
     xlab = "Y", ylab = "Density",
     main = "Comparison of Observed and Predicted Densities")

# Add 100 posterior predictive density lines
set.seed(42)  # For reproducibility
n_lines <- 100
samples <- sample(1:1e4, n_lines)  # Randomly sample 100 posterior predictive datasets
for (s in samples) {
  lines(density(D_sim[s, ]), col = alpha("lightblue", 0.3), lwd = 1)
}

# Add the density line for the observed Y values
obs_density <- density(df$Y)
lines(obs_density$x, obs_density$y, col = "green", lwd = 2)

# Add legend
legend("topright", legend = c("Posterior Predictive Densities", "Observed Density"),
       col = c("lightblue", "green"), lty = 1, lwd = c(1, 2))
```

The light blue lines show distributions of model predicted $Y$ values. 
The green line shows the distribution of the observed $Y$ values.
As we can see, there seem to be no systematic differences between the observed and predicted values. 
The model seems to capture the relationship well.
If we see systematic deviations here, we need to reconsider the model definition. 

**Example**: If you want to predict pain ($Y$ variable) and you have a lot of zeros (pain-free participants)
you will probably see a discrepancy between the observed and predicted values in this plot.
What could you do? You could use a two step process (model the probability that a person is pain-free 
and then model the pain intensity for the people who have pain) or use a different model (like a zero-inflated model).

Note that we did not explicitely assume normally distributed errors in the model definition above,
so we won't check this here but in the Frequentist framework below.

## Linear regression with 2 predictors in the Frequentist Framework 

To reiterate from the last chapter:
In full, the **classical linear regression model** can be written as 
(see p. 21-22 in Westfall):

$$ Y_i|X_i = x_i \sim_{independent} N(\beta_0 + \beta_1 x_{i1} + \dots \beta_k x_{ik},\sigma^2)$$
for $i = 1, \dots, n$.

The $Y_i$ are independently normally distributed *conditioned* on the predictors 
having the values $X_i = x_i$. Each conditional distribution has an expected value ($\mu$)
that is a linear function of the predictors and a constant variance $\sigma^2$.

If the assumptions of the classical linear regression model are met, the least squares
estimators (OLS) are the best (smallest variance) linear unbiased (on average correct) estimators 
- so-called: **BLUE** - of the parameters.

### Adding a transformed predictor to the model {#adding_transformed_predictor_freq}

No, let's fit the same model as [above](#adding_transformed_predictor_bayes) in the Bayesian framework.

The model is:

$$height_i = \alpha + \beta_1 weight_i + \beta_2 weight_i^2 + \varepsilon_i$$
whereas
$$\varepsilon_i \sim N(0, \sigma)$$

And if you build the expectation on both sides for fixed $weight_i$, you get:

$$\mathbb{E}(height_i|weight_i) = \alpha + \beta_1 weight_i + \beta_2 weight_i^2$$

The last line means, the expected height of a person given a certain weight depends 
quadratically on the weight. The error term $\varepsilon_i$ is on average zero, hence it goes away here.
Remember the [law of large numbers](https://en.wikipedia.org/wiki/Law_of_large_numbers#Weak_law): 
The sample mean $\bar{\varepsilon_i}$ approaches the expected value $\mathbb{E}(\varepsilon_i)=0$ 
as the sample size increases. If you drew many samples (from the true model) 
and average over the error terms, the average will approach zero. 
Think of [this](https://github.com/jdegenfellner/ZHAW_Teaching/blob/main/Wiggle_in_simple_linear_regression.R) animated 
graph if you need a dynamic image of the regression model.
The weights are considered fixed and therefore do not change when building the expectation.

We are looking for fixed, but unknown, parameters $\alpha$, $\beta_1$, $\beta_2$ and $\sigma$.
The fixed $\sigma$ indicates that the observations *wiggle* around the expected value
equally strong not matter which weight we have. This is called **homoscedasticity**.

Let's fit the model using the `lm` function in R which uses 
[least squares](https://en.wikipedia.org/wiki/Least_squares) to estimate the parameters.
At this point I could torture you with [matrix algebra](https://en.wikipedia.org/wiki/Matrix_(mathematics)) 
and show you the [normal equations](https://en.wikipedia.org/wiki/Linear_least_squares) for linear regression,
but I will spare you for now.
Note that the least squares algorithm for fitting the curve works for all
kinds of functional forms. For example, we could also fit an exponential curve using the same
technique (see [exercise 9](#exercise9_multiple_regression)).

```{r}
# scale weight
d$weight_s <- scale(d$weight)
# Fit the model
m4.2 <- lm(height ~ weight_s + I(weight_s^2), data = d)
summary(m4.2)
mean(d$height)
confint(m4.2, level = 0.94)
```

See `?I` in R. This command is used so that R knows that it should 
treat the "^2" as "square" and not as formula syntax.
We could also create a new variable as before. Whatever you prefer.

#### Interpretation of output and coefficients {#interpretation_output_freq_quadratic}

- The intercept $\alpha$ is the **model-predicted height** of a person of **average weight** 
  ($weight_s=0$ for a person of average weight).
  Note that this is not equal to the average height ($138.2636~cm$) of the people in the data set 
  (see [exercise 12](#exercise12_multiple_regression)).
- The residuals have range from $-19.97$ to $19.51$. So, the model maximally
  overestimates the heights by $19.97$ cm and underestimates by $19.51$ cm.
  These numbers are plausible when you look at the scatterplot with the fitted
  curve.
- The coefficients $\beta_1$ and $\beta_2$ agree with the Bayes estimates.
  Specifically, $\beta_2$ is non-zero indicating curvature. You cannot directly interpret the coefficients
  as in the non-quadratic case since, for instance, you cannot change $weight^2$ by one unit
  and hold $weight$ constant at the same time. Refer to Peter Westfall's book section 9.1. for all the details.
- If you like $p$-values: All the hypotheses that the coefficients are zero
  are rejected. The $p$-values are very small. The values of the test statistics can not be explained
  by chance alone. On the other hand, for at least $\beta_1$ and
  and the global test this is not a surprise when you look at the scatterplot.
  After having fit many models, you would have guessed that all three parameters 
  are solidly non-zero. The intercept is not zero since a person of average weight probably
  has non-zero height. $\beta_1$ is non-zero since you can easily imagine a linear
  trend line with positive slope going through the data, and $\beta_2$ is non-zero
  since there is clearly (non-trivial) curvature in the scatterplot.
- The $R^2$ is a whopping $0.96$ which could be a sign of overfitting, but
  in this case we conclude that the true relationship is caputured rather well.
  [Overfitting](https://en.wikipedia.org/wiki/Overfitting) would occur if 
  our curve would rather fit the noise in the data than
  the underying trend.

#### Checking model assumptions

```{r, warnings=FALSE}
library(performance)
check_model(m4.2)
```

If we want to be perfectionists, we could remark that (upper right plot)
in the lower fitted values the residuals are more negative, 
meaning that the model overestimates the heights in this region.
In the middle region the model underestimates a bit and we can see
a positive tendency in the residuals. Apart from that,
the diagnostic plots look excellent.

### Adding another predictor to the model {#adding_predictor_freq}

Now, we add another predictor to the model. We use $X_1$ and $X_2$ 
**simultaneously** to predict $Y$. We are now in the lucky situation that
we can still visualize the situation in 3D. The regression line from simple
linear regression
becomes a [plane](https://stackoverflow.com/questions/47344850/scatterplot3d-regression-plane-with-residuals).
The vertical distances between the data points and the plane are the residuals.
See [here](https://rpubs.com/pjozefek/576206) or 
[here](https://www.sthda.com/english/wiki/scatterplot3d-3d-graphics-r-software-and-data-visualization)
at the end for examples.
Minimizing the sum of the squared errors gives again 
the estimates for the coefficients. 

For demonstration purposes, we can **create data ourselves** with known
coefficients. This is the same as [above](#adding_predictor_bayes).
This is the true model, which we usually do not know:

$$ Y_i = \beta_0 + \beta_1 X_{1i} + \beta_2 X_{2i} + \varepsilon_i$$
$$ \varepsilon_i \sim N(0, \sigma^2)$$
$$ \mathbb{E}(Y_i|X_1 = x_1; X_2 = x_2) = \beta_0 + \beta_1 x_1 + \beta_2 x_2$$
$$ i = 1 \ldots n$$

for example:

$$ Y_i = 10 + 0.5 \cdot X_{1i} + 1 \cdot X_{2i} + \varepsilon_i$$
$$ \varepsilon_i \sim N(0, 5)$$
$$ \mathbb{E}(Y_i|X_1 = x_1; X_2 = x_2) = 10 + 0.5 x_1 + 1 x_2$$
$$ i = 1 \ldots n$$


According to the model, the conditional expected value of $Y_i$ given $X_1 = x_1$ and $X_2 = x_2$
is a linear function of $x_1$ and $x_2$. Note, that small letters are realized values
of random variables. Also note, that in the expectation the error term goes away, since
$\mathbb{E}(\varepsilon_i) = 0$.

- If $X_1$ increases by one unit, $Y$ increases by $0.5$ units on average (in expectation).
- If $X_2$ increases by one unit, $Y$ increases by $1$ unit on average (in expectation).
- If $X_1$ and $X_2$ are zero, $Y$ is $10$ on average (in expectation).

Why in expectation? Because there is still the error term which makes the whole thing random!
We can see that an increase in $X_1$ does not influence the relationship between $X_2$ and $Y$.
Hence, there is **no interaction** between $X_1$ and $X_2$ with respect to $Y$.

Now lets's draw 100 points from this model, fit the model and add the plane:

```{r}
library(plotly)

set.seed(123)
n <- 100
X1 <- rnorm(n, 0, 5)
X2 <- rnorm(n, 0, 5)
Y <- 10 + 0.5 * X1 + 1 * X2 + rnorm(n, 0, 2)
d <- data.frame(X1 = X1, X2 = X2, Y = Y)

# Fit the model
m4.3 <- lm(Y ~ X1 + X2, data = d)
summary(m4.3)

# Create a grid for the plane
X1_grid <- seq(min(d$X1), max(d$X1), length.out = 20)
X2_grid <- seq(min(d$X2), max(d$X2), length.out = 20)
grid <- expand.grid(X1 = X1_grid, X2 = X2_grid)

# Predict the values for the grid
grid$Y <- predict(m4.3, newdata = grid)

# Convert the grid into a matrix for the plane
plane_matrix <- matrix(grid$Y, nrow = length(X1_grid), ncol = length(X2_grid))

# Create the interactive 3D plot
plot_ly() %>%
  add_markers(
    x = d$X2, y = d$X1, z = d$Y,
    marker = list(color = "blue", size = 5),
    name = "Data Points"
  ) %>%
  add_surface(
    x = X1_grid, y = X2_grid, z = plane_matrix,
    colorscale = list(c(0, 1), c("red", "pink")),
    showscale = FALSE,
    opacity = 0.7,
    name = "Fitted Plane"
  ) %>%
  plotly::layout(
    scene = list(
      xaxis = list(title = "X1"),
      yaxis = list(title = "X2"),
      zaxis = list(title = "Y")
    ),
    title = "Interactive 3D Scatterplot with Fitted Plane"
  )
```

This is, of course, a very idealized situation. There is no curvature in the plane,
no interaction, no outliers, no heteroscadasticity. It's the simplest case of multiple regression
with 2 predictors. Reality is - usually - more complicated.

Let's look at the summary output and check model assumptions:

```{r}
summary(m4.3)
check_model(m4.3)
```

We could repeat this simulation to get a feeling for the variability.
The posterior predictive checks look nice. In this case, we *know* that the model is true.

#### Adding variables to the model and why

This is a very complex question. We will go into it in later chapters and the next course (Methodenvertiefung).
At this point we can say this:
We add variables to the model (and probably use 
other models apart from linear regression) *depending* on the goal at hand (prediction or explanation).

- **Prediction** (i.e. guessing the outcome as good as possible) seems to be easier 
  than explanation. For instance, within linear models and 
  just a handful of predictors, one can even brute force the problem by searching through
  all subsets of predictors. If that is not possible, one could use clever algortithms, like 
  [best subest selection](https://www.sthda.com/english/articles/37-model-selection-essentials-in-r/155-best-subsets-regression-essentials-in-r/). 

- **Explanation** (looking for causal relationships between variables) is more difficult.
  - What is **not** a good idea if you want to find *causal* relationships,
  is to throw all variables into the model and hope for the best.
  - What is also not a good idea is to select variables depending on the $p$-values of the coefficients (Westfall Chapter 11).
  - **Leaving variables out**, that are important, can lead to biased estimates of the coefficients 
  ([omitted variable bias](https://en.wikipedia.org/wiki/Omitted-variable_bias)).
  - Importantly, also **adding variables** can hurt conclusions from the model (see Statistical Rethinking 6.2).

### Interaction Term $X_1 \times X_2$ {#interaction_term}

I recommend reading the excellent explanations about interactions
in John Kruschke's book [Doing Bayesian Data Analysis](https://nyu-cdsc.github.io/learningr/assets/kruschke_bayesian_in_R.pdf),
15.2.2 und 15.2.3. Peter Westfall also has a nice explanation in his [book](https://www.routledge.com/Understanding-Regression-Analysis-A-Conditional-Distribution-Approach/Westfall-Arias/p/book/9780367493516?srsltid=AfmBOore3O_Ciecl0TTkr9AjPIY1d6OmbQa7o7IAdKpTSkD8s9HkwzD4) 
in section 9.3.

Our statistical model is now:

$$ Y_i = \beta_0 + \beta_1 X_{1i} + \beta_2 X_{2i} + \mathbf{\beta_3 X_{1i} \times X_{2i}} + \varepsilon_i$$
$$ \varepsilon_i \sim N(0, \sigma^2)$$
$$ \mathbb{E}(Y_i|X_1 = x_1; X_2 = x_2) = \beta_0 + \beta_1 x_{1} + \beta_2 x_{2} + \beta_3 x_{1} \times x_{2}$$
$$ i = 1 \ldots n$$

for example:

$$ Y_i = 10 + 0.5 \cdot X_{1i} + 1 \cdot X_{2i} + 0.89 \cdot X_{1i} \times X_{2i} + \varepsilon_i$$
$$ \varepsilon_i \sim N(0, 5)$$
$$ \mathbb{E}(Y_i|X_1 = x_1; X_2 = x_2) = 10 + 0.5 x_1 + 1 x_2 + 0.89 x_1 \times x_2$$
$$ i = 1 \ldots n$$

The second equation states that the conditional expectation of $Y_i$ given $X_1=x_1$ and $X_2=x_2$ 
is a function of $x_1$ and $x_2$ and their interaction $x_1 \times x_2$ (i.e., the product). 

We are in a different situation now.
Set for instance $x_2$ to a certain value, say $x_2 = 7$. Then the relationship (in expectation) 
between $Y$ and $X_1$ is:

$$ \mathbb{E}(Y_i|X_1 = x_1; X_2 = 7) = 10 + 0.5 x_1 + 1 \cdot 7 + 0.89 x_1 \cdot 7$$
$$ \mathbb{E}(Y_i|X_1 = x_1; X_2 = 7) = 10 + (0.5 + 0.89 \cdot \mathbf{7}) \cdot x_1 + 1 \cdot 7$$

Depending on the value of $x_2$, the *effect* of $X_1$ on $Y$ changes. 
Hence, $X_2$ **modifies** the relationship between $X_1$ and $Y$, or stated otherwise,
$X_1$ and $X_2$ **interact** with respect to $Y$. Remember, the word *effect* is 
used in a strictly technical/statistical sense and **not in a causal** sense.
It does not mean that if we *do* change $X_1$ by one unit, 
$Y$ will also change in an experiment. We are purely describing the relationship
in an associative way. We will probably touch causality later.
Bayesian statistics and causal inference are gaining popularity. Hence, we should try to keep up.

Let's draw 100 points from this model, fit the model and add the plane (see also [exercise 4](#exercise3_multiple_regression)):

```{r}
set.seed(123)
n <- 100
X1 <- rnorm(n, 0, 5)
X2 <- rnorm(n, 0, 5)
Y <- 10 + 0.5 * X1 + 1 * X2 + 0.89 * X1 * X2 + rnorm(n, 0, 5)
d <- data.frame(X1 = X1, X2 = X2, Y = Y)

# Fit the model
m4.4 <- lm(Y ~ X1 * X2, data = d)
summary(m4.4)

# Create a grid for the plane
X1_grid <- seq(min(d$X1), max(d$X1), length.out = 20)
X2_grid <- seq(min(d$X2), max(d$X2), length.out = 20)
grid <- expand.grid(X1 = X1_grid, X2 = X2_grid)

# Predict the values for the grid
grid$Y <- predict(m4.4, newdata = grid)

# Convert the grid into a matrix for the plane
plane_matrix <- matrix(grid$Y, nrow = length(X1_grid), ncol = length(X2_grid))

# Create the interactive 3D plot
plot_ly() %>%
  add_markers(
    x = d$X2, y = d$X1, z = d$Y,
    marker = list(color = "blue", size = 5),
    name = "Data Points"
  ) %>%
  add_surface(
    x = X1_grid, y = X2_grid, z = plane_matrix,
    colorscale = list(c(0, 1), c("red", "pink")),
    showscale = FALSE,
    opacity = 0.7,
    name = "Fitted Plane"
  ) %>%
  plotly::layout(
    scene = list(
      xaxis = list(title = "X1"),
      yaxis = list(title = "X2"),
      zaxis = list(title = "Y")
    ),
    title = "Interactive 3D Scatterplot with Fitted Plane"
  )
```

The term `X1 * X2` is a shortcut for `X1 + X2 + X1:X2` where `X1:X2` is the interaction term.
R automatically includes the main effects of the predictors when an interaction 
term is included (variable inclusion principle).
The true but usually unknown $\beta$s are estimated quite precisely.

#### Formal test for interaction
We could apply a formal test for the interaction term by model comparison.
The command `anova(., .)` would compare the two models and test if the change in the
residual sum of squares is statistically interesting.

```{r}
m4.5 <- lm(Y ~ X1 + X2, data = d) # without interaction
anova(m4.5, m4.4)
```

One can show that the following test statistic is $F$ distributed under the null hypothesis (that $\beta_3=0$):

$$ F = \frac{\left(RSS_{\text{Model 1}} - RSS_{\text{Model 2}}\right) / \left(df_{\text{Model 1}} - df_{\text{Model 2}}\right)}{RSS_{\text{Model 2}} / df_{\text{Model 2}}}$$

where $RSS$ is the residual sum of squares, 
$df$ are the degrees of freedom of the residual sum of squares for both models.

The output of the `anova` command shows us the residual degress of freedom (`Res.Df`)
of both models, the residual sum of squares errors of both models (`RSS`), 
the sum of squared errors between model 1 and model 2 (`Sum of Sq`), the value of the 
F-statistic and the $p$-value for the hypothesis, that the coefficient for
the interaction term is zero ($\beta_3=0$). Model 1 RSS has 97 degrees of freedom, since we have 100 data points
and 3 parameters to estimate ($\beta_0, \beta_1, \beta_2$). Model 2 has 96 degrees of freedom, since
we have 100 data points and 4 parameters to estimate ($\beta_0, \beta_1, \beta_2, \beta_3$).

Let's verify the value of the $F$ statistic:

```{r}
RSS_model1 <- sum(residuals(m4.5)^2)
RSS_model2 <- sum(residuals(m4.4)^2)
df_model1 <- n - length(coef(m4.5))
df_model2 <- n - length(coef(m4.4))
F <- ((RSS_model1 - RSS_model2) / (df_model1 - df_model2)) / (RSS_model2 / df_model2)
F
# Sum of Sq
RSS_model1 - RSS_model2
```

In the numerator of the $F$ statistic, we have the change in the residual sum of squares 
(from the small (model 1) model to the larger one (model 2), `Sum of Sq`)
per additional parameter in the model (one additional parameter $\beta_3$). 

In the denominator, we have the residual sum of squares per residual degree of freedom of 
the larger model (model 2). Hence, in the numerator we have the information on how much
better we get with respect to the number of variables added, and in the denominator
we have information on how good the full model is with respect to its degrees of freedom.

The $p$-value is the probability of observing a value of the F statistic as extreme or more 
extreme than the one we observed, given that the null hypothesis is true. Here, 
the $p$-value is extremely small. So, statistically we would see an improvement in RSS
which is not explainable by chance alone.
But **let's be careful with $p$-values** and especially with fixed cutoff values for $p$, 
which we will **never** use in this script.
Even for a rather small effect $\beta_3$, we would reject the null hypothesis, if only the sample
size is large enough. Since a very small effect relative to $\beta_1$ and $\beta_2$ would
probably not be of practical interest, one should be careful with looking at $p$-values alone.
For instance, in Richard McElreath's book [Statistical Rethinking](https://xcelab.net/rm/statistical-rethinking/),
there are no $p$-values at all. I like that.

If you again look at the comparison of the RSS between the two models, you would 
immediately see that the model with the interaction term is better (at least with respect to this metric). 
The difference is huge. We have already mentioned in the context of $R^2$ not to overinterpret 
such metric, because RSS is monotonically descreasing with number of variables added and reaches
zero when the number of variables equals the number of data points (see [exercise 3](#exercise3_multiple_regression)).

### Using an interaction plot to suspect a potential interaction {#interaction_plot}

Chronologically before we include an interaction term in the model, we can use an interaction plot
to see if there is a potential interaction between the predictors.
We can just create a categorical predictor out of the continuous predictors.
Just categorize the predictors into quartiles and plot the means of the dependent variable ($Y$).
If the lines are parallel, there is probably no interaction. If the lines are not parallel, 
there might be an interaction.

```{r}
n <- 100
X1 <- rnorm(n, 0, 5)
X2 <- rnorm(n, 0, 5)
Y <- 10 + 0.5 * X1 + 1 * X2 + 0.89 * X1 * X2 + rnorm(n, 0, 5)
d <- data.frame(X1 = X1, X2 = X2, Y = Y)

# Create categorical variables based on quartiles
d$X2_cat <- cut(d$X2, 
                breaks = quantile(d$X2, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE), 
                include.lowest = TRUE, 
                labels = c("Q1", "Q2", "Q3", "Q4"))

d$X1_cat <- cut(d$X1, 
                breaks = quantile(d$X1, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE), 
                include.lowest = TRUE, 
                labels = c("Q1", "Q2", "Q3", "Q4"))

# Create the interaction plot
interaction.plot(d$X2_cat, d$X1_cat, d$Y)
```

There seems to be an interaction of the predictors with respect to $Y$. The lines are not parallel.
If there was no interaction, the change in $Y$ with respect to $X_2$ would be the same 
for all levels of $X_1$.
This seems not to be the case here.
See [exercise 5](#exercise5_multiple_regression).

If we had one or both predictors already categorical, we would not have to discretize them before.

### Simpsons Paradox {#simpsons_paradox}

The [Simpsons paradox](https://en.wikipedia.org/wiki/Simpson%27s_paradox) is a phenomenon, 
in which a trend appears in several different groups of data but disappears 
or reverses when these groups are combined. I agree with the criticism that this is not really 
a paradox but a failure to consider confounding variables adequately. Let's quickly invent an example.
We are interested in the relationship *hours of muscle training* and *strength* (not based on evidence)
in children vs. adults. Within both groups there will be an increasing relationship. The more training,
the more muscle strength. But if we combine the groups, we will see a decreasing relationship.

```{r}
library(tidyverse)
n <- 100
age <- c(rep("child", n/2), rep("adult", n/2))
training <- c(rnorm(n/2, 0, 5) + 30, rnorm(n/2, 0, 5)+ 10)
strength <- c(
  10 + 0.5 * training[1:(n/2)] + rnorm(n/2, 0, 2), # For children
  25 + 0.5 * training[(n/2 + 1):n] + rnorm(n/2, 0, 2) # For adults
)

d <- data.frame(age = age, training = training, strength = strength)

ggplot(d, aes(x = training, y = strength, color = age)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE) + # Group-specific regression lines
  geom_smooth(data = d, aes(x = training, y = strength), 
              method = "lm", se = FALSE, color = "black", linetype = "dashed", linewidth = 1.2) + # Overall regression line
  labs(title = "Regression Lines for Training and Strength",
       x = "Training",
       y = "Strength") +
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

- **Group-Specific Trends**:
   - In the group of **children** (blue line), strength **increases** with training, as indicated by the positive slope of the regression line.
   - Similarly, in the group of **adults** (red line), strength also **increases** with training.

- **Overall Trend**:
   - When both groups are combined and the categorical variable *age* is neglected, 
     the overall regression line (black, dashed) shows a **negative slope**, suggesting that **strength decreases** with training. 
   - This overall trend is opposite to the trends observed within the individual groups.

- **Why Does This Happen?**
   - This paradox occurs because the relationship between the grouping variable (`age`) and the independent variable (`training`) creates a confounding effect.
   - In this case: Children tend to have higher training values overall, 
     while adults tend to have lower training values. *Age* is associated with both *strength* and *training* and 
     is therefore a [*confounder*](https://en.wikipedia.org/wiki/Confounding) in the relationship between *training* and *strength*. 
     Not adjusting (=including it in the regression model as predictor) for *age* leads to a misleading association.

## What happens when you just throw variables into multiple regression? {#throwing_variables}

This sub-chapter is **important**. I can guarantee you that not too many applied 
scientists using regression models know about this.

A first taste of causality.

Richard McElreath has 3 cool examples on 
[github](https://github.com/rmcelreath/causal_salad_2021/blob/main/1_causal_salad.r)
that show what happens in the context of *explanation* if you include variables the wrong 
way and explains this in a 
[video](https://www.youtube.com/watch?v=KNPYUVmY3NM&ab_channel=RichardMcElreath).

We will look at these below - **the pipe, the fork and the collider**. These are causal graphs showing the 
relationships between the variables. One is interested in the **effect** of X on Y.
In this case, it is truly an *effect*, since we create the models in such a way that changing
one variable ($X$), changes the other ($Y$) - which is indicated by an arrow in the graph.
These graphs are called DAGs - directed acyclic graphs. *Directed* because of the arrows,
*acyclic* because there are no cycles in the graph. One nice tool for drawing them is
[dagitty](https://www.dagitty.net/), which is also 
[implemented in R](https://cran.r-project.org/web/packages/dagitty/index.html).
The [online drawing tool](https://www.dagitty.net/dags.html) is handy.

**Remember**: Not only can it hurt to add variables to the model in an explanatory (causal) context,
but also in a predictive context. The model can become unstable and the predictions
can become worse. This is called [overfitting](https://en.wikipedia.org/wiki/Overfitting).

### Pipe
In this setting $X$ is associated with $Z$ and $Z$ is associated with $Y$. 
$X$ and $Y$ are not directly associated, but through $Z$ (see graph below).
If we condition on $Z$, the association between $X$ and $Y$ *disappears*.
This means, if we **know** the value of $Z$, $X$ does not give us any *additional* information about $Y$.

This can be seen in the scatterplot: Once we are within $Z=0$ (black dots) or $Z=1$ (red dots), 
$X$ does not give us any information about $Y$,
i.e., the point cloud is horizontal and there is no correlation.

The `inv_logit` function is the inverse of the logit function.
It assigns higher probability of $Z$ being $1$ if X has a higher value.
Let's plot this for understanding:

```{r}
x <- seq(-3, 3, 0.1)
plot(x, inv_logit(x), type = "l", col = "blue", lwd = 2, xlab = "X", ylab = "P(Z=1|X)")
```

Higher $X$ values lead to higher probabilities of $Z=1$. 
As you can see, more red dots are on the right side of the scatterplot below.

Now to the pipe and a mini simulation for it:

```{r}
# pipe
library(dagitty)
library(tidyverse)
library(ggdag)

dag <- dagitty( 'dag {
  X -> Z -> Y
}' )

dagitty::coordinates( dag ) <-
  list( x=c(X=0, Y=2, Z=1),
        y=c(X=0, Y=0, Z=0) )

ggdag(dag) + 
  theme_dag()

a <- 0.7
cols <- c( col.alpha(1,a) , col.alpha(2,a) )

# pipe
# X -> Z -> Y
N <- 1000
X <- rnorm(N)
Z <- rbern(N,inv_logit(X))
Y <- rnorm(N,(2*Z-1))

plot( X , Y , col=cols[Z+1] , pch=16 )
abline(lm(Y[Z==1]~X[Z==1]),col=2,lwd=3)
abline(lm(Y[Z==0]~X[Z==0]),col=1,lwd=3)
abline(lm(Y~X),lwd=3,lty=3)

cor(X[Z==1],Y[Z==1])
cor(X[Z==0],Y[Z==0])
```

Or in the framework of Simpsons paradox:

```{r}
library(rethinking)

N <- 1000
X <- rnorm(N)
Z <- rbern(N,inv_logit(X))
Y <- rnorm(N,(2*Z-1))
mod1 <- lm(Y ~ X) # without conditioning on Z
summary(mod1)
mod2 <- lm(Y ~ X + Z) # with conditioning on Z
summary(mod2)
```

Adding $Z$ to the model (i.e., conditioning on $Z$) makes the coefficient for $X$ disappear. 
Knowing $Z$ means that $X$ does not give us any additional information about $Y$.
See [exercise 7](#exercise7_multiple_regression).

$Z$ is also called a [**mediator**](https://en.wikipedia.org/wiki/Mediation_(statistics)).

We could easily verify that $Z$ constitutes a mediator by using Baron and 
Kenny's 1986 approach (see Wiki):

```{r}
summary(lm(Y ~ X))# Step 1
summary(lm(Z ~ X))# Step 2
summary(lm(Y ~ X + Z))# Step 3
```

Both coefficients for $X$ are "significant" in steps 1 and 2.
The coefficients for $Z$ in step 3 is "significant" and the coefficient for
$X$ is smaller compared to step 1. And not to forget: the very small $p$-values are strongly
related to the very large smaple size and a strict cutoff ($\alpha = 0.05$) makes no sense.

**Example**: *Hours of studying* is associated with *exam performance*.
This observation makes sense. A mediator in this context: *understanding/knowledge*.
One could study for hours inefficiently with low levels of concentration
and achieve low exam performance or one could study for a comparatively
low number of hours and achieve good results. If we know that a person has gained the
*understanding/knowledge*, the *hours of studying* do not give us any additional 
information with respect to *exam performance*.

So, in order to see the effect of $X$ on $Y$, we do *not* condition on $Z$.

### Fork {#fork}

In health science, this is the classical [confounder](https://en.wikipedia.org/wiki/Confounding).
$Z$ is associated with $X$ and $Y$. $X$ and $Y$ are *actually* not associated (no arrow) but an association is 
still shown: `cor(X,Y) = 0.5097437`. 
But if we condition on $Z$, the association between $X$ and $Y$ disappears. The assocation is 
spurious in this case and *adjusting* for the confounder yields the correct result. 

```{r}
# fork
# X <- Z -> Y

dag <- dagitty( 'dag {
  X <- Z -> Y
}' )

dagitty::coordinates( dag ) <-
  list( x=c(X=0, Y=2, Z=1),
        y=c(X=0.5, Y=0.5, Z=0) )

ggdag(dag) + theme_dag()

N <- 1000
Z <- rbern(N)
X <- rnorm(N,2*Z-1)
Y <- rnorm(N,(2*Z-1))

plot( X , Y , col=cols[Z+1] , pch=16 )
abline(lm(Y[Z==1]~X[Z==1]),col=2,lwd=3)
abline(lm(Y[Z==0]~X[Z==0]),col=1,lwd=3)
abline(lm(Y~X),lwd=3,lty=3)

cor(X[Z==1],Y[Z==1])
cor(X[Z==0],Y[Z==0])
cor(X,Y)
```

In the example, we know that $Z$ is associated with both $X$ and $Y$ (per construction).
As you can see in the definitions of $X$, $Y$ and $Z$, $X$ and $Y$ would not be associated 
weren't it for $Z$. 

**Example**: *Carrying lighters* is (spuriously) associated with *lung cancer*. Obviously, *carrying a lighter*
does *not* cause lung cancer. *Carrying a lighter* is associated with *smoking*, since smokers need to light their
cigarrettes and how often do non-smokers carry a lighter just for fun? 
*Smoking* is (causally) associated with *lung cancer*. If you just look at the association between
*carrying a ligher* and *lung cancer*, you would find an association. But if you condition on *smoking*, the association
disappears. See [exercise 13](#exercise13_multiple_regression).

### Collider {#collider}
This is rather interesting. $X$ and $Y$ are independently associated 
with $Z$. This can be seen in the toy example below, where $X$ and $Y$ are used
for the definition of $Z$. $Z$ is defined as a **sum 
score** of $X$ and $Y$. Sum scores are *very* often used in health sciences (and others).
The higher the sum score, the higher the probability that $Z=1$.
Now, *if* we know the value of $Z$, $X$ and $Y$ are negatively associated (see graph). 
The reason for this association is that there is a compensatory effect. In order to get a high 
score, you can either have a high value of $X$ or $Y$ or both.
This induces the negative correlation.

```{r}
# collider
# X -> Z <- Y

#dag
dag <- dagitty( 'dag {
  X -> Z <- Y
}' )

dagitty::coordinates( dag ) <-
  list( x=c(X=0, Y=2, Z=1),
        y=c(X=0, Y=0, Z=1) )

ggdag(dag) + theme_dag()

N <- 1000
X <- rnorm(N)
Y <- rnorm(N)
Z <- rbern(N,inv_logit(2*X+2*Y-2))

plot( X , Y , col=cols[Z+1] , pch=16 )
abline(lm(Y[Z==1]~X[Z==1]),col=2,lwd=3)
abline(lm(Y[Z==0]~X[Z==0]),col=1,lwd=3)
abline(lm(Y~X),lwd=3,lty=3)

cor(Y[Z==1], X[Z==1])
cor(Y[Z==0], X[Z==0])

summary(lm(Y ~ X)) # mod1
summary(lm(Y ~ X + Z)) # mod2
```

If you think of Simpsons paradox again, conditioning on $Z$ *creates* an association between $X$ and $Y$,
which would otherwise be independent (by definition). So by learning the value of $Z$, we can learn something
from $X$ about $Y$. On the one hand, adding $Z$ to the model creates an association where there is none,
on the other hand, *prediction* of $Y$ is better with $Z$ in the model! In prediction, we largely
do not care about the causal structure. We just want to predict $Y$ as accurately as possible.

The sum score example is also called [Berkson's paradox](https://en.wikipedia.org/wiki/Berkson%27s_paradox).
Let's check if this also works for three variables in a sum score 
in [exercise 10](#exercise10_multiple_regression).

### Multicollinearity {#multicollinearity}
Westfall section 8.4 and McElreath section 6.1 cover this topic.
[Multicollinearity](https://en.wikipedia.org/wiki/Multicollinearity) means that there is a strong
correlation between two or more predictors. It is *perfect* multicollinearity when the correlation is 1.
For example, if you accidentally include the same variable twice in the model, or when you include
the elements of a sum score and the sum score itself as predictors in the model. R just gives you 
an *NA* if you do this in `lm`. 

#### Example from McElreath 6.1 {#mcElreath_6.1}
The code creates body heights and leg lengths (left and right) for 100 people.

`runif` draws a uniform random number between 40 and 50% of height as leg length, hence on average 45%.
The slope in the linear regression should therefore be around the average height divided by the average leg length:
$\frac{10}{0.45 \cdot 10} \sim 2.2$.

```{r}
library(rethinking)

N <- 100
set.seed(909)
height <- rnorm(N, 10, 2)
leg_prop <- runif(N, 0.4, 0.5)
leg_left <- leg_prop * height + rnorm(N, 0, 0.02)
leg_right <- leg_prop * height + rnorm(N, 0, 0.02)

d <- data.frame(height = height, leg_left = leg_left, leg_right = leg_right)
cor(d$leg_left, d$leg_right)

m4.6 <- quap(
  alist(
    height ~ dnorm(mu, sigma),
    mu <- a + bl * leg_left + br * leg_right,
    a ~ dnorm(10, 100),
    bl ~ dnorm(2, 10),
    br ~ dnorm(2, 10),
    sigma ~ dexp(1)
  ),    data = d
)
precis(m4.6)
plot(precis(m4.6))
```

With the given regression model, one asks the question: "What is the value of knowing each leg's length, after 
already knowing the other leg's length?" The answer is: "Not much.", since they are highly correlated.
Both coefficients are not around the expected $\beta$ and the credible intervals are wide and include the 
credible value $0$.

```{r}
post <- extract.samples(m4.6)
plot( bl ~ br, post, col=col.alpha(rangi2,0.1), pch = 16 )
```

Since the two coefficients are almost perfectly multicollinear 
(`cor(d$leg_left, d$leg_right)=0.9997458`), knowing one leg's length does not 
give us any additional information about the other leg's length.
Leaving out the right leg length would give the correct result ([exercise 11](#exercise11_multiple_regression)).


#### Example from Westfall 8.4

In the example below, the second prector $X2$ is a perfect linear function of the first predictor $X1$.

```{r}
# Westfall 8.4.
set.seed(12345)
X1 = rnorm(100)
X2 = 2*X1 -1     # Perfect collinearity
Y = 1 + 2*X1 + 3*X2 + rnorm(100,0,1)
summary(lm(Y~X1+X2))
```

Let's look at a 3D plot:

```{r}
library(plotly)

# Generate data
set.seed(42)
X1 <- rnorm(100)
X2 <- 2 * X1 - 1  # Perfect collinearity
Y  <- 1 + 2 * X1 + 3 * X2 + rnorm(100, 0, 1)

# Create the 3D scatter plot with vertical lines
plot_ly() %>%
  add_markers(x = X1, y = X2, z = Y, 
              marker = list(color = Y, colorscale = "Viridis", size = 5),
              name = "Data Points") %>%
  layout(title = "3D Scatter Plot",
         scene = list(xaxis = list(title = "X1"),
                      yaxis = list(title = "X2"),
                      zaxis = list(title = "Y")))
#VIF(lm(Y ~ X1 + X2)) # error
#check_model(lm(Y ~ X1 + X2)) # error
```


There is no unique solution for a plane in this case.
Infinitely many planes can be defined using the "line" in space.
The problem collapses into the simple linear regression problem.
One can just plug in the formula for $X2$ into the model, which yields the identical result:

```{r}
set.seed(12345)
X1 = rnorm(100)
# Y = 1 + 2*X1 + 3*(2*X1 -1) + rnorm(100,0,1) = 
Y = -2 + 8*X1 + rnorm(100,0,1)
summary(lm(Y ~ X1))
```

In reality, you only have one parameter ($X1$ or $X2$) in the case of perfect multicollinearity.


## More than 2 predictors

In practice, you often want more than two predictors in the model.
If you listen closely to the research question raised by yourself or 
your colleagues, often it won't be clearly stated what goal you want to achieve
with the regression model: 
[prediction or explanation](https://projecteuclid.org/journals/statistical-science/volume-25/issue-3/To-Explain-or-to-Predict/10.1214/10-STS330.full).

If we want to be rigorous about the true relationships between variables, 
we need to think about the causal structure of the variables, 
as [Richard McElreath](https://www.youtube.com/watch?v=YvhuYONl1o0&ab_channel=KI-Campus)
argues. 

In order not to just throw in variables into our regression model and hope for the best,
in order to have a fighting chance, we use the information provided
in Chapter 5 in the Statistical Rethinking book.

**Overall strategy for modeling:**

- **Have a research question** and decide how this question can be answered: 
  Quantitatively (experiment or observational study) or 
  qualitatively. 
  Remember: Statistical modeling is not a catch-all approach to
  turn data into truth. 
  And: We never want to play *find-the-right-statistical-test-Bingo*.
  Statistical tests should make sense and not just thrown at data.
  What does the current literature say about the topic?

- **Define goal**: Prediction or explanation? 
  *Prediction*: Use everything and every model type you can get your hands 
  on to "best" predict the outcome (neural nets, random forests, etc.).
  *Explanation*: How do the variables relate to each other, 
  how do they influence each other? Temporal relationships are important. 
  And we want interpretable models.

- **Do** exploratory data analysis (**EDA**) to get a feeling for the data:
  - This includes (among others) scatterplots, histograms, boxplots,
    correlation matrices, etc.
  - How are the raw correlations/associations between the variables?
  - How are the variables distributed?
  - Are there outliers?
  - Are there missing values and why are the values missing?
  - Think about the data generating process. How did the data come about?
  - There is no shame in fitting exploratory models too.

- **Draw a DAG** (directed acyclic graph) of the hypthesized relationships between 
  the variables, even though we will not do formal 
  [causal inference](https://miguelhernan.org/whatifbook) in this lecture yet
  (hopefully in the next).
  The drawn DAG has testable implications (more below). See 5.1.2 in Rethinking.
  Nicely enough, `dagitty`spits out:
  - The [adjustment sets](https://rdrr.io/cran/dagitty/man/adjustmentSets.html) 
    for the regression model based on a DAG. Which variables should I include
    as covariates in my model, *if* the DAG is correct?
  - The [implied conditional dependencies](https://search.r-project.org/CRAN/refmans/dagitty/html/impliedConditionalIndependencies.html) 
    coming from the DAG. These
    can be checked with the data. It is not a proof that we have the true
    relationships depicted by the DAG, but it is a good start.

- **Decide on a statistical model** (more on that later).

- **Define priors** for the parameters of the model:
  If we do not know much, we choose vague priors (i.e., wide range of plausible 
  parameter values). In the best case,
  priors should be well argued for, especially in a *low-data* setting
  (if we do not have many observations).

- **Prior predictive checks**: Does the model produce outcomes ($Y$) that
  are at all plausible? If not, we might have to rethink the model and/or priors.

- **Fit and check the model**:
  If the models makes certain assumptions, check if these are met.
  For the classical regression model, see Chapter 4 in Westfall.
  Posterior predictive check: Check if the model produces new data that looks 
  like the observed data.

- **Interpret and report** the results.

### Example in NHANES data {#example_nhanes}

We will now try to invent a not too exotic example with 
[NHANES](https://www.cdc.gov/nchs/nhanes/index.html) data.
The National Health and Nutrition Examination Survey (NHANES) is a large, 
ongoing study conducted by the CDC to assess the health and nutritional status 
of the U.S. population. It combines interviews, physical examinations, and 
laboratory tests to collect data on demographics, diet, chronic diseases, and 
physical activity. NHANES uses a complex, nationally representative sampling 
design, making it a valuable resource for public health research and policy 
development. 

Open data is **great** and NHANES provides all the data to download for free
without any restrictions or even a registration. For convenience, Randall Pruim
created an [R package](https://cran.r-project.org/web/packages/NHANES/NHANES.pdf) 
called `NHANES` that contains a cleaned-up version of the
NHANES data from 2009-2012. We will use this and ignore the complex sampling
design for now. 

```{r}
#install.packages("NHANES")
library(pacman)
p_load(NHANES, tidyverse)
data(NHANES)
head(NHANES)
unique(NHANES$SurveyYr)
colnames(NHANES)
```

The variable descriptions can be found 
[here](https://cran.r-project.org/web/packages/NHANES/NHANES.pdf).
As an exercise, please do exploratory data analysis; 
see [exercise 14](#exercise14_multiple_regression).

#### Research question 
Does *Physical activity* (PhysActive) **influence** the *average systolic 
blood pressure* (BPSysAve) in adults ($\ge 20$ years)?

We propose the following relationships among the variables 
(and limit ourselves to 4 covariates):

```{r}
df <- NHANES # shorter

# Define the DAG
dag <- dagitty('dag {
  PhysActive -> BPSysAve
  Age -> PhysActive
  Age -> BPSysAve
  PhysActive -> BMI
  BMI -> BPSysAve
  Gender -> PhysActive
  Gender -> BMI
  Gender -> BPSysAve
}')

# Set node coordinates for a nice layout
dagitty::coordinates(dag) <- list(
  x = c(PhysActive = 0, BPSysAve = 2, Age = 1, BMI = 1, Gender = 0.5),
  y = c(PhysActive = 1, BPSysAve = 1, Age = 2, BMI = 1.5, Gender = 2)
)

# Plot the DAG with larger node labels and bubbles
ggdag(dag) + 
  theme_minimal() + 
  geom_dag_point(size = 20, color = "black") +  # Increase node size
  geom_dag_text(size = 2.5, color = "white") +    # Increase label size
  ggtitle("Hypothesized Relationships") + 
  theme(plot.title = element_text(hjust = 0.5))
```

This DAG illustrates the hypothesized relationships between physical activity, 
age, BMI, gender, and blood pressure (BPSysAve). Physical activity potentially directly 
influences blood pressure, as regular exercise is hypothesized to improve 
cardiovascular health and lower blood pressure. Age has a direct effect on 
blood pressure, as arterial stiffness and other age-related physiological changes 
contribute to higher blood pressure over time. BMI also plays a role, with higher 
BMI being associated with increased blood pressure due to greater vascular 
resistance and metabolic factors. Additionally, gender affects blood pressure, 
as men and women often have different baseline levels due to hormonal and 
physiological differences.

There are also indirect pathways that contribute to these relationships. 
Age influences both physical activity and BMI, as older individuals tend to 
be less active and may experience weight gain due to metabolic changes (confounder). 
BMI and physical activity are also interconnected, physical activity influences BMI, 
further influencing blood 
pressure (mediation). Gender plays a role in both BMI and physical activity, 
as men and women tend to have different average BMI distributions 
and physical activity levels due to both biological and societal factors (confounder).

Note, that we have a very large sample size (6,919 people). As exercise for later,
we randomly select 50 to 100 participants and repeat the analysis.
This would also give us the opportunity to study how well we can
infer the relationships in the larger sample from the smaller one,
which could be an eye-opener.

**Which variables should be included in the model?**

```{r}
adjustmentSets(dag, exposure = "PhysActive", outcome = "BPSysAve")
impliedConditionalIndependencies(dag)
```

The one adjustment set (using the command `adjustmentSets`) proposed by `dagitty` tells us that 
we should include age and gender as covariates in the model (in addition to *PhysActive*), 
*if* the DAG is correct.
Note that **we do not include all three covariables in the model here**, just two of them.

*If* the DAG is correct, this would imply some conditional independencies
(using the command `impliedConditionalIndependencies`) .
We should probably check them to see how this is done:

First, we need to test, if age and BMI are independent, 
*if* we condition on gender and physical activity;
i.e., just add the to the regression model.
This is the case:

```{r}
# Age _||_ BMI | Gnder, PhyA
summary(lm(Age ~ BMI + Gender + PhysActive, data = df %>% dplyr::filter(Age >= 20)))
```

The implication can be confirmed. 

**Interpretation**: If we know gender and physical activity (yes/no) of a person,
BMI does not offer any *additional* information about age.

Second, I would go out on a limb and say that there is no relationship 
between age and gender, at least not in a statistically relevant sense.


#### Goal

The **goal** is to better understand (explain) the relationship between 
physical activity and blood pressure while considering age, BMI and gender. 
I would suggest that in most cases relevant for us, one wants to understand
relationships rather to predict an outcome.

#### EDA
[Exercise 14](#exercise14_multiple_regression)

#### Statistical model 

The outcome is a continuous variable (blood pressure), the exposure is a binary 
variable (physical activity yes/no). 
Surprise, surprise, we could try a **multiple linear regression model** here.

As a first approximation, let's assume *BPSysAve* is (sufficiently) normally distributed,
which can be critized. Let's quickly visualize the distribution of*BPSysAve* and overlay
a normal density with paramaters estimated from the data ($\mu$ sample mean of *BPSysAve*, 
$\sigma$ sample standard deviation of *BPSysAve*):

```{r, warning=FALSE}
df_age <- df %>% dplyr::filter(Age >= 20)
df_age %>%
  ggplot(aes(x = BPSysAve)) +
  geom_histogram(aes(y = after_stat(density)), 
  bins = 30, fill = "lightblue", alpha = 0.6) +  
  geom_density(color = "blue", linewidth = 1) +  
  stat_function(
    fun = dnorm, 
    args = list(mean = mean(df_age$BPSysAve, na.rm = TRUE), 
    sd = sd(df_age$BPSysAve, na.rm = TRUE)), 
    color = "red", linewidth = 1, linetype = "dashed"
  ) +  # Theoretical normal curve
  labs(
    x = "Systolic Blood Pressure (BPSysAve)", 
    y = "Density", 
    title = "Distribution of Systolic Blood Pressure with Normal Curve"
  ) +
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

Not quite normal, but we will assume it for now.

#### Priors

Now, let's write down the statistical model.
We choose vague priors since we have a vast data set.
The advantage of the Bayesian approach remains (that we have a fully 
probabilistic model). Admitted, using Bayes is much more technical, at least
at first glance.

We center age again, which makes the interpretation of the intercept easier.

- The intercept ($\beta_0$) is the expected value of *BPSysAve* for a person 
  of average age, not physically active (reference level of the factor) and 
  female gender (reference level of the factor).

- The coefficient for *PhysActive* ($\beta_1$) is the expected difference in 
  *BPSysAve* between a physically active person and a non-physically active person, 
  holding all other predictors constant. *BPSysAve* ranges from 78 to 226 mmHg.
  If we do not know enough, we can just take a very vague prior for now:
  $\beta_1 \sim \text{Normal}(0, 50)$. With 95% probability, the effect of
  physical activity on *BPSysAve* is between -100 and 100.

- $beta_2$: The expected change in *BPSysAve* for a one-year increase in age 
  (above the mean age), holding all other predictors constant. 
  $\beta_2 \sim \text{Normal}(0, 10)$. This effect should be comparatively small,
  since we are dealing with a one-year increase in age. We could further 
  standardize age, then the effect would be in relation to one standard 
  deviations increase in age.

- $\sigma$: The standard deviation of the residuals. 
  $\sigma \sim \text{Uniform}(0, 50)$. We use a vague prior.
  95% of the residuals are between -100 and 100. This seems plausible enough.

\begin{eqnarray*}
BPSysAve_i &\sim& \text{Normal}(\mu_i, \sigma)\\
\mu_i &=& \beta_0 + \beta_1 \cdot PhysActive + \beta_2 \cdot Age_{centered} + \beta_3 \cdot Gender\\
\beta_0 &\sim& \text{Normal}(140, 20)\\
\beta_1 &\sim& \text{Normal}(0, 50)\\
\beta_2 &\sim& \text{Normal}(0, 10)\\
\beta_3 &\sim& \text{Normal}(0, 10)\\
\sigma &\sim& \text{Uniform}(0, 50)
\end{eqnarray*}

Note that age and gender are not influenced by each other.

#### Prior Predictive Checks {#prior_predictive_checks_NHANES}
Now we simply draw from the priors and thereby generate systolic blood pressure values.

```{r}
# Prior predictive checks
set.seed(123)
n_sims <- dim(df_age)[1] # 7235 participants
beta_0_vec <- rnorm(n_sims, 140, 20)
beta_1_vec <- rnorm(n_sims, 0, 50)
beta_2_vec <- rnorm(n_sims, 0, 10)
beta_3_vec <- rnorm(n_sims, 0, 10)
sigma_vec <- runif(n_sims, 0, 50)
BPSysAve_sim <- rnorm(n_sims, beta_0_vec + beta_1_vec + beta_2_vec + beta_3_vec, sigma_vec)
```

This yields also negative values (please verify) for *BPSysAve*, which is not plausible.

In [exercise 15](#exercise15_multiple_regression) we will play around with the priors 
until the prior predictive check producues plausible values. 
**Admitted**, one should probably not look at the distribution of data while doing this.
On the other hand, we could have just Googled the distribution of blood pressure values.
This is the result:

```{r}
set.seed(123)
n_sims <- dim(df_age)[1] # 7235
beta_0_vec <- rnorm(n_sims, 120, 7)
beta_1_vec <- rnorm(n_sims, 0, 10)
beta_2_vec <- rnorm(n_sims, 5, 10)
beta_3_vec <- rnorm(n_sims, 0, 10)
sigma_vec <- runif(n_sims, 0, 20)
BPSysAve_sim <- rnorm(n_sims, beta_0_vec + beta_1_vec + beta_2_vec + beta_3_vec, sigma_vec)
length(BPSysAve_sim) # 7235

df_sim_vs_obs <- data.frame(
  BPSysAve_sim = BPSysAve_sim,
  BPSysAve_obs = df_age$BPSysAve
)

# Combine observed and simulated values into one long-format data frame
df_long <- df_sim_vs_obs %>%
  tidyr::pivot_longer(cols = everything(), 
  names_to = "Type", values_to = "BPSysAve") %>%
  mutate(Type = factor(Type, levels = c("BPSysAve_obs", "BPSysAve_sim"), 
  labels = c("Observed", "Simulated")))

# Plot densities of both observed and simulated values
ggplot(df_long, aes(x = BPSysAve, fill = Type)) +
  geom_density(alpha = 0.5) +  # Semi-transparent density curves
  labs(
    x = "Systolic Blood Pressure", 
    y = "Density", 
    title = "Observed vs. Simulated Systolic Blood Pressure"
  ) +
  scale_fill_manual(values = c("Observed" = "blue", "Simulated" = "red")) + 
  theme_minimal() + 
  theme(plot.title = element_text(hjust = 0.5))
```

Looks decent. Not quite there yet. 
The blood pressure values have a smaller variances and are skewed.
Let's nevertheless fit the model.

#### Fit and check model
 We use `quap` for the model fitting.

```{r}
library(data.table)
df <- NHANES
df <- as.data.table(df)

# Compute age mean for centering
Age_mean <- mean(df[Age >= 20,]$Age, na.rm = TRUE) # adults

# Filter age
df <- df %>% dplyr::filter(Age >= 20) %>% 
  dplyr::select(BPSysAve, PhysActive, Age, Gender, BMI) %>%
  drop_na()
dim(df) # 6919    7
sum(is.na(df)) # 0

# Fit model
m_NHANES <- quap(
  alist(
    BPSysAve ~ dnorm(mu, sigma), 
    mu <- beta_0 + beta_1[PhysActive] + beta_2 * (Age - Age_mean) + beta_3[Gender],
    beta_0 ~ dnorm(120, 7),
    beta_1[PhysActive] ~ dnorm(0, 10),  
    beta_2 ~ dnorm(5, 10),
    beta_3[Gender] ~ dnorm(0, 10),  
    sigma ~ dunif(0, 20) 
  ),
  data = df
)

precis(m_NHANES, depth = 2)

# We want to know the expected difference for the levels of PhysActive and Gender:
# (which is not directly visible in the summary)
post <- extract.samples(m_NHANES)
post$diff_PhysActive <- post$beta_1[,2] - post$beta_1[,1]
post$diff_G <- post$beta_3[,2] - post$beta_3[,1]

library(conflicted)
conflicts_prefer(posterior::sd)

#as.data.frame(precis(post, depth = 2, digits = 3))
post_summary <- as.data.frame(precis(post, depth = 2, digits = 3))  # Convert to dataframe
post_summary[, -which(names(post_summary) == "histogram")]  # Remove histogram column
```

The summary tells us that `diff_PhysActive` is $-0.74$ with a large proportion
of the posterior $<0$. Please research if this is a clinically relevant difference.
My guess is: no.

#### Fit model using `lm` (Frequentist approach)

We can also fit the model using the Frequentist approach.
No priors. The results should be very similar due to the large sample size.

```{r}
library(car)
dim(df) 
df <- df %>% mutate(Age_center = Age - mean(Age))
mod <- lm(BPSysAve ~ PhysActive + Age_center + Gender, data = df)
summary(mod)
check_model(mod)
qqPlot(mod) # bad
```

The results are similar to the Bayesian approach.
In the summary output we see that the effect of physical activity is $-0.73693$
with a small $p$-value ($0.051$). This is one example for the absurdity of strict cutoffs
for $p$-values. The other coefficients including the estimate for $\sigma$ are also similar.
It's always a good idea to double-check. 

Model fit could be better, the model thinks the blood pressure is normally distributed.
There also seems to be some heteroscadasticity and the residuals are not normally distributed.

**By the way, usually**, one would fit this model by throwing in all variables into the model, including BMI.
Let's see if this approach changes the results:

```{r}
mod <- lm(BPSysAve ~ PhysActive + Age + Gender  + BMI, data = df) # add BMI
summary(mod)
check_model(mod)
qqPlot(mod) 
```

**This changed the results notably**. Before, the effect of physical activity was $-0.73693$
with a rather small $p$-value,
now it is $-0.22870$ with a $p$-value ten times as large.

**Remember this**: The interpretation of a whole paper could change by falsely including
an additional variable (in this case BMI).

#### Improve model {#improve_model_NHANES}
Now, that we have seen that the model fit is probably not sufficient, we could try to improve it.

**First** order of business is the **non-normality** of the blood pressure values (outcome).
The model predicts normally distributed values, which is not the case.
It seems that the blood pressure values are right-skewed.

Let's try a log-normal distribution for the outcome variable, which only 
takes positive values and is skewed. 

We would **further** try to model $\sigma$ as a linear function of $\mu$.
For higher expected values of *BPSysAve*, we would expect higher variability.
Hence, we try to **explicitely model heteroscadasticity**.

Note, that we have no problem with adding parameters to the model 
because of the large sample size.

See [exercise 16](#exercise16_multiple_regression).

\begin{eqnarray*}
BPSysAve_i &\sim& \text{Log-Normal}(\mu_i, \sigma_i)\\
\sigma_i &=& \text{exp}(\beta_4 + \beta_5 \cdot \mu_i)\\
\mu_i &=& \beta_0 + \beta_1 \cdot PhysActive + \beta_2 \cdot Age_{centered} + \beta_3 \cdot Gender\\
\beta_0 &\sim& \text{Normal}(140, 20)\\
\beta_1 &\sim& \text{Normal}(0, 50)\\
\beta_2 &\sim& \text{Normal}(0, 10)\\
\beta_3 &\sim& \text{Normal}(0, 10)\\
\beta_4 &\sim& \text{Normal}(0, 10)\\
\beta_5 &\sim& \text{Uniform}(0, 50)
\end{eqnarray*}

We now allow the standard deviation $\sigma_i$ to be different for different values of the mean.
With `exp()` we ensure that $\sigma_i$ is positive.
This time, we assume: $log(BPSysAve_i) \sim \text{Normal}(\mu_i, \sigma_i)$. That's what it means
to be log-normally distributed.  

**Fit improved model**

```{r}
sum(is.na(df)) # 0
set.seed(122)
m_NHANES_lnorm <- quap(
  alist(
    BPSysAve ~ dlnorm(lmu, lsd), 
    lsd <- exp(beta_4 + beta_5 * lmu),
    lmu <- beta_0 + beta_1[PhysActive] + beta_2 * (Age - Age_mean) + beta_3[Gender],
    beta_0 ~ dnorm(140, 10),  # 
    beta_1[PhysActive] ~ dnorm(0, 10),  # 
    beta_2 ~ dnorm(0, 10),
    beta_3[Gender] ~ dnorm(0, 10),  # 
    beta_4 ~ dnorm(0, 10),  # 
    beta_5 ~ dnorm(0, 10)  #
  ),
  data = df,
  start = list(beta_0 = 140, beta_1 = c(0.5, 0), beta_2 = 0, beta_3 = c(0.5, 0), beta_4 = 1, beta_5 = 0)
)

#precis(m_NHANES_lnorm, depth = 2)
nhanes_summary <- as.data.frame(precis(m_NHANES_lnorm, depth = 2))  # Convert to dataframe
nhanes_summary[, -which(names(nhanes_summary) == "histogram")]  # Remove histogram column

# difference in levels of PhysActive and Gender (catorical variables):
post_lnorm <- extract.samples(m_NHANES_lnorm)
post_lnorm$diff_PhysActive <- post_lnorm$beta_1[,2] - post_lnorm$beta_1[,1]
post_lnorm$diff_G <- post_lnorm$beta_3[,2] - post_lnorm$beta_3[,1]
#precis(post_lnorm, depth = 2)
post_lnorm_summary <- as.data.frame(precis(post_lnorm, depth = 2))  # Convert to dataframe
post_lnorm_summary[, -which(names(post_lnorm_summary) == "histogram")]  # Remove histogram column
```

`diff_PhysActive` is now  $-0.003350291$ with a very tight credible interval
containing $0$ as a credible effect. The model is pretty sure, that the effect
is practically zero.

Let's vizualize the posterior predictive distributions of 100 samples:

```{r}
# posterior predictive checks
sample_BP <- sim(m_NHANES_lnorm, n = 1000)

# Convert the first 100 rows of the posterior samples into a long format for ggplot
df_posterior <- as.data.frame(t(sample_BP[1:100,])) %>%
  pivot_longer(cols = everything(), names_to = "Simulation", values_to = "BPSysAve_sim")

# Create the plot
ggplot() +
  geom_density(data = df_posterior, aes(x = BPSysAve_sim, group = Simulation), 
               color = "lightblue", alpha = 0.05) +
  geom_density(data = df, aes(x = BPSysAve), color = "green", linewidth = 1.2) +
  labs(title = "Density Estimation: Original vs. Posterior Samples",
       x = "Systolic Blood Pressure",
       y = "Density") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
```

With respect to the distribution of blood pressure, the model looks much better now.
Note, that we did not assume that the residuals are normally distributed, neither
did we assume homoscedasticity (on the contrary).

See [exercise 17](#exercise17_multiple_regression).

#### Interpret and report results

Summarizing the results, we find the expected direction of association in our first model.
The effect of physical activity on blood pressure was negative. 
After improving the model fit, we concluded that the small but clinically irrelevant effect
from the first model dissapeared.
It would be interesting to see how other variables containing more information 
influence blood pressure, like *PhysActiveDays* (Number of days in a typical week 
that participant does moderate or vigorous intensity activity).

### Concluding remarks

The NHANES example was a somewhat realistic instance of a multiple regression model
which could easily be encountered in practice. In a Master thesis, we are often in 
a low-data setting with 20-150 observations. Here, the priors are much more important.

[Exercise 18](#exercise18_multiple_regression) asks you to repeat the analysis 
with a smaller sample size.

We saw that we can readily adapt the model, if assumptions of the classical 
linear regression model are violated. In our case, we had to model the outcome
as log-normal in order to allow for the model to predict skewed blood pressure values. 
In addition, we explicitly modeled heteroscedasticity by allowing the standard deviation
to grow with the mean. Theoretically, one could build arbitrarily complex models
this way. Of course, syntax problems and convergence issues could arise.
Even in this example some starting values were not fit for purpose, which is why
I had to add the `set.seed()` command in the improved model. 

## Exercises

[E] Easy, [M] Medium, [H] Hard

(Some) solutions to exercises can be found in the git-repo [here](https://github.com/jdegenfellner/Script_QM2_ZHAW/tree/main/Solutions_Exercises).

### [M] Exercise 1 {#exercise1_multiple_regression}

- Fit a model with a cubic term for weight and height of the !Kung San people.
- Add the prediction bands as seen in the book.
- Come up with an explanation for the functional form of this relationship.
- Could there be reasons to for taking a less complicated model 
  ([1](https://en.wikipedia.org/wiki/Statistical_model_specification), [2](https://en.wikipedia.org/wiki/Occam%27s_razor))?

### [E] Exercise 2 {#exercise2_multiple_regression}

Consider the model equations from [above](#adding_transformed_predictor_bayes) 
where we used polynomial regression to model the relationship between 
weight and height:

- Draw the model hierarchy for the model.

### [H] Exercise 3 {#exercise3_multiple_regression}

Invent a data set (or use the first 3 lines of a previous data set) 
with 3 observations of $Y$ and $X_1, X_2$ and $X_3$. You have a data frame
with 3 rows and 4 columns. 

- Fit a model with $Y$ as the dependent variable and $X_1, X_2, X_3$ as predictors.
- How big is $R^2$?
- Could you have calulated this without `lm`and R?

### [E] Exercise 4 {#exercise4_multiple_regression}

Go back to the [section about the interaction term](#interaction_term) in the linear model.

- Use the code provided.
- Standardise the predictors. How are the $\beta$s changing and what is their interpration now?
- Change the relative sizes of the true but usually unknown $\beta$s. 
  What happens to the estimates and the graph?
- What happens if you change the error term and increase or decrease its variance?

### [E] Exercise 5 {#exercise5_multiple_regression}

Draw the interaction plot from the [section about the interaction plot](#interaction_plot) 
for the case when there is no interaction, i.e. $\beta_3 = 0$.

### [M] Exercise 6 {#exercise6_multiple_regression}

Go back to the model assumptions checks [above](#check_model_bayes).

- Create the same two plots for the simple mean model without predictors, just with the intercept.
- Which model fits the data better according to these posterior predictive checks?

### [E] Exercise 7 {#exercise7_multiple_regression}

Go back to the Simpson's paradox [section](#simpsons_paradox).

- Invent your own example for the pipe, fork and collider.

### [M] Exercise 8 {#exercise8_multiple_regression}

- Take a data set of your choosing with many columns, say 10 or so, either from the internet 
  or from R (if available)
- Fit a model for an arbitrary outcome, add more and more variables to predict the outcome
  and verify that the $R^2$ increases.

### [M] Exercise 9 {#exercise9_multiple_regression}

Exponential curve fitting. Go back to the [section](#adding_transformed_predictor_freq) about adding a transformed predictor in the Frequentist
setting. Hint: You can use the `optim` function in R.

- Assume the relationship between weight and height looks like this: 
  $height_i = \alpha + \beta_1 e^{\beta_2 weight_i}$.
- Use R and the least squares method to estimate the parameters $\alpha, \beta_1, \beta_2$.
- Note that the sum of squared errors is this: 
  $\sum_{i=1}^n (height_i - \alpha - \beta_1 e^{\beta_2 weight_i})^2$.
- What happens if you do not constrain the parameters $\beta_1$ and $\beta_2$ to be negative?
- Calculate the $R^2$ for this model.

### [E] Exercise 10 {#exercise10_multiple_regression}

Let's try to verify if Berkson's paradox (which we have mentioned in the 
[collider-section](#collider)) also works for three variables in a sum score.

- Now, we assume some college admits only applicants in the top 20% of a score 
  consisting of the sum of three variables: $W, X, Z$ (grade point average, math score, verbal score).
- All three scores are individually normally distributed with mean 100 and standard deviation 15.
- Calculate the correlation matrix `cor()` of all students and the admitted students.
- Are any of the three variables correlated?
- Plot a scatterplot of the math score and the verbal score und color the points according 
  to being admitted or not. You can add trendlines for the two groups.

### [E] Exercise 11 {#exercise11_multiple_regression}

Go back to the multicollinearity [section](#multicollinearity) and the [example](#mcElreath_6.1) 
from McElreath 6.1.

- Verify that the coefficient is correct when leaving out the right leg length from the model.

### [M] Exercise 12 {#exercise12_multiple_regression}

Go backt to the interpretation of the intercept in the quadratic height model [above](#interpretation_output_freq_quadratic).

- Why is the intercept not equal to the sample mean of the heights?

### [H] Exercise 13 {#exercise13_multiple_regression}

Go back to the section about the [fork](#fork) and the example about smoking at the end.

- Draw a DAG for this example.
- Create data in R, where you assume that the probability of carrying a lighter is higher in smokers,
  the probability of lung cancer is higher in smokers.
- Show that the association between carrying a lighter and lung cancer disappears when conditioning on smoking.
- You may also invent an example relevant to the field of physiotherapy.

### [H] Exercise 14 {#exercise14_multiple_regression}

- Perform extensive EDA on the NHANES data set implemented in R. 
  See [above](#example_nhanes).

### [H] Exercise 15 {#exercise15_multiple_regression}

Go back to the NHANES model [above](#prior_predictive_checks_NHANES).

- Play around with the priors until the prior predictive check produces plausible values.

### [M] Exercise 16 {#exercise16_multiple_regression}

Go to the improved model fit of our NHANES example [above](#improve_model_NHANES).

- Draw the model hierarchy for the model.

### [H] Exercise 17 {#exercise17_multiple_regression}

Go back to the improved model fit of our NHANES example [above](#improve_model_NHANES).

- Draw observed vs. model-predicted blood pressure values.
- add the $y=x$ line to the plot.
- Is the fit better compared to the first model?

### [H] Exercise 18 {#exercise18_multiple_regression}

Go back to the NHANES example [above](#example_nhanes).

- Repeat the NHANES analysis with a smaller sample size.
- Draw 50-100 rows randomly from the adult NHANES data set we have used for the full
  analysis.
- What do you observe?

<!--chapter:end:03-Multiple_Linear_Regression.Rmd-->

# Reliability and Validity

For this chapter we refer to the book 
[Measurement in Medicine](https://www.cambridge.org/core/books/measurement-in-medicine/8BD913A1DA0ECCBA951AC4C1F719BCC5).

I invite you to read the introductory chapters 1 and 2 about concepts,
theories and models, and types of measurement.

In general, when conducting a measurement of any sort 
(laboratory measurements, scores from questionnaires, etc.), 
we want to be reasonably sure

- that we actually **measure what we intend to measure**; 
  ([validity](https://en.wikipedia.org/wiki/Validity_(statistics)); 
  chapter 6 in the book);
- that the measurement does **not change too much** if the 
  underlying **conditions are the same**
  ([reliability](https://en.wikipedia.org/wiki/Reliability_(statistics)); 
  chapter 5 in the book); and
- that we are able to detect a **change** if the underlying conditions change 
  ([responsiveness](https://tinyurl.com/3vdcxy49); chapter 7 in the book); and
- that we understand the meaning of a change in the measurement
  (interpretability; chapter 8 in the book).

In this [video](https://www.youtube.com/watch?v=KuT2n1w0Ixc&ab_channel=Physiotutors), 
Kai jump starts you on reliability and validity.


## Reliability
You can watch this [video](https://www.youtube.com/watch?v=9HSoWaRpcys&ab_channel=Physiotutors) 
to get started.

Imagine, you measure a patient (pick your favorite measurement), for example,
the range of motion (ROM) of the shoulder. 

- If you are interested in how
  similar your measurements are in comparison to your colleagues, you are 
  trying to determine the so-called **inter-rater reliability**.
- If you are interested in how similar your measurements are when you measure
  the same patient twice, you are trying to determine the so-called 
  **intra-rater reliability**.

Assuming there is a true (but unknown) underlying value (of Range of Motion, ROM), 
it is clear that measurements will not be *exactly* the same. 
Possible influences (potentially) causing different results are:

- the measurement instrument itself (e.g., the goniometer),
- the patient (e.g., mood/motivation),
- the examiner (e.g., mood, influence on patient),
- the environment (e.g., the room temperature).

Note that the **true score** is defined in our context as the average of all measurements
if we would measure repeatedly an infinite number of times.

### Peter and Mary's ROM measurements

The data can be found [here](http://www.clinimetrics.nl/answers-to-the-assignments-in-textbook_22_0.html).
We randomly select 50 measurements from Peter and Mary in 50 different patients,
plot their measurements and annotate the absolutely largest one(s). At first the
*not affected shoulder* (nas) and then the *affected shoulder* (as).

```{r}
library(pacman)
p_load(tidyverse, readxl)

# Read file
url <- "https://raw.githubusercontent.com/jdegenfellner/Script_QM2_ZHAW/main/data/chapter%205_assignment%201_2_wide.xls"
temp_file <- tempfile(fileext = ".xls")
download.file(url, temp_file, mode = "wb")  # mode="wb" is important for binary files
df <- read_excel(temp_file)

head(df)
dim(df)

# As in the book, let's randomly select 50 patients.
set.seed(123)
df <- df %>% sample_n(50)
dim(df)

# "as" = affected shoulder
# "nas" = not affected shoulder

df <- df %>%
  mutate(diff = abs(ROMnas.Peter - ROMnas.Mary))  # Compute absolute difference

max_diff_point <- df %>%
  dplyr::filter(diff == max(diff, na.rm = TRUE))  # Find the row with the max difference

df %>%
ggplot(aes(x = ROMnas.Peter, y = ROMnas.Mary)) +
  geom_point() + 
  geom_point(data = max_diff_point, aes(x = ROMnas.Peter, y = ROMnas.Mary), 
             color = "blue", size = 4) +  # Highlight max difference point
  geom_abline(intercept = 0, slope = 1, color = "red") +
  theme_minimal() +
  ggtitle("ROMnas.Peter vs. ROMnas.Mary") +
  theme(plot.title = element_text(hjust = 0.5)) +
  annotate("text", x = max_diff_point$ROMnas.Peter, 
           y = max_diff_point$ROMnas.Mary, 
           label = paste0("Max Diff: ", round(max_diff_point$diff, 2)), 
           vjust = -1, color = "blue", size = 4)

# average abs. difference:
mean(df$diff, na.rm = TRUE) # 7.2

cor(df$ROMnas.Peter, df$ROMnas.Mary, use = "complete.obs") 
```

The red line represents the line of equality ($y=x$). If the measurements are 
exactly the same,
all points would lie on this line. The blue point represents the largest 
absolute difference in measured Range of Motion (ROM) values between Peter and Mary
from the randomly chosen 50 people. Note that the maximum difference in 
all 155 patients is 35 degrees.

The first simple measure of agreement we could use is the (Pearson) correlation, which
measures the **strength and direction of a linear relationship between two variables**.
But correlation does not exactly measure what we want. If there was a bias 
(e.g., Mary systematically measures 5 degrees more than Peter), correlation would not 
notice this. (-> exercise later...). It actually is too optimistic about 
the agreement since it only cares about the linearity and not about a potential bias.
$r=0.2403213$ which indicates a weak positive correlation. Higher values of Peter's
are associated with higher values of Mary's measurements.
But: Knowing Peter's measurement does not help us to [*predict*](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#Interpretation_of_the_size_of_a_correlation) 
Mary's measurement at such a low correlation (-> exercise later).
So, on the *not affected shoulder* (nas), agreement is really bad.

What about the affected shoulder (as)?

```{r}
library(ggExtra)
df <- df %>%
  mutate(diff = abs(ROMas.Peter - ROMas.Mary))  # Compute absolute difference

max_diff_point <- df %>%
  dplyr::filter(diff == max(diff, na.rm = TRUE))  # Find the row with the max difference

p <- df %>%
ggplot(aes(x = ROMas.Peter, y = ROMas.Mary)) +
  geom_point() + 
  geom_point(data = max_diff_point, aes(x = ROMas.Peter, y = ROMas.Mary), 
             color = "blue", size = 4) +  # Highlight max difference point
  geom_abline(intercept = 0, slope = 1, color = "red") +
  theme_minimal() +
  ggtitle("ROMas.Peter vs. ROMas.Mary") +
  theme(plot.title = element_text(hjust = 0.5)) +
  annotate("text", x = max_diff_point$ROMas.Peter, 
           y = max_diff_point$ROMas.Mary, 
           label = paste0("Max Diff: ", round(max_diff_point$diff, 2)), 
           vjust = -1, color = "blue", size = 4)
# Add marginal histograms
ggMarginal(p, type = "density", fill = "gray", color = "black")

# average abs. difference:
mean(df$diff, na.rm = TRUE) # 7.2
cor(df$ROMas.Peter, df$ROMas.Mary, use = "complete.obs") 

# mean difference
mean(df$ROMas.Peter - df$ROMas.Mary, na.rm = TRUE) 
```

In the affected side, the average absolute difference is even larger ($7.78$)
with a maximum absolute difference of 37 degrees,
but the correlation is much higher ($r=0.8516653$). See Figure 5.2 in the book.

Btw, this is an an example for using the correlation coefficient even though
the marginal distributions are not normal: There are much more measurements in the higher
values around 80 than below, say, 60. But the correlation coefficient makes sense
for descriptive purposes.

In this case, knowing Peter's measurement *does* help us to predict 
Mary's measurement (-> exercise later).

### Intraclass Correlation Coefficient (ICC)

One way to measure reliability is to use the intraclass correlation coefficient (ICC).

This measure is based on the idea that the observed score $Y_i$ consists of the true score 
(ROM) ($\eta_i$) and a measurement error (for each person).
The proportion of the true score variability to the total variability is the ICC.

In the background one thinks of a statistical model from the 
[Classical Test Theory (CTT)](https://en.wikipedia.org/wiki/Classical_test_theory).
There is 

- a true underlying score $\eta_i$ (for each patient i) and 
- an error term $\varepsilon_i \sim N(0, \sigma_i)$ which is the difference between the true
score and 
- the observed score $Y_i$.

$$Y_i = \eta_i + \varepsilon_i$$

It is assumed that $\eta_i$ and $\varepsilon_i$ are independent: ($\mathbb{C}ov(\eta_i, \varepsilon_i)=0$).
This is a nice assumption because now we know (see [here](https://en.wikipedia.org/wiki/Variance#Addition_and_multiplication_by_a_constant)) 
that the variance
of the observed score $Y_i$ is just the sum of the variance of the true score $\eta_i$
and the variance of the error term $\varepsilon_i$:

$$\mathbb{V}ar(Y_i) = \mathbb{V}ar(\eta_i) + \mathbb{V}ar(\varepsilon_i)$$
$$\sigma_{Y_i}^2 = \sigma_{\eta_i}^2 + \sigma_{\varepsilon_i}^2$$

We want most of the variability in our observed scores $Y_i$ to be explained by the 
true but unobservable scores $\eta_i$. The measurement error $\varepsilon_i$ should be
be comparatively small. If it is large, we are mostly measuring noise or at least not
what we want to measure. 

**How do we get to the theoretical definition of the ICC?**

If you either pull two people with the same true but unobservable score $\eta$ out of the population
or measure the same person twice and the score ($\eta$) does not change in between, we can 
**define reliability as correlation between these two measurements**:

$$Y_1 = \eta + \varepsilon_1$$
$$Y_2 = \eta + \varepsilon_2$$

$$cor(Y_1, Y_2) = cor(\eta + \varepsilon_1, \eta + \varepsilon_2) = \frac{Cov(\eta + \varepsilon_1, \eta + \varepsilon_2)}{\sigma_{Y_1}\sigma_{Y_2}}  =$$

If we use 

- the [properties of the covariance](https://en.wikipedia.org/wiki/Covariance#Properties), 
- the fact that the true score $\eta$ and the errors $\varepsilon_i$ are independent, and 
- the fact that the errors $\varepsilon_1$ and $\varepsilon_2$ are independent, we get:

$$\frac{Cov(\eta, \eta) + Cov(\eta, \varepsilon_2) + Cov(\varepsilon_1, \eta) + Cov(\varepsilon_1, \varepsilon_2)}{\sigma_{Y_1} \sigma_{Y_2}}  =$$
$$\frac{\sigma_{\eta}^2 + 0 + 0 + 0}{\sigma_{Y_1} \sigma_{Y_2}}$$

Since $\eta$ is a random variable (we draw a person randomly from the population),
it is well defined to talk about the variance of $\eta$ (i.e., $\sigma_{\eta}^2$).
I think this aspect may not come across in the book quite so clearly.

Furthermore, it does not matter if I call the measurement $Y_1$, $Y_2$ or more general
$Y$, since they have the same variance and true score: 

$$\sigma_{Y} = \sigma_{Y_1} = \sigma_{Y_2}$$

Hence, it follows that:

$$cor(Y_1, Y_2) = \frac{\sigma_{\eta}^2}{\sigma_{Y_1}^2} = \frac{\sigma_{\eta}^2}{\sigma_{Y}^2} =  \frac{\sigma_{\eta}^2}{\sigma_{Y}^2} = \frac{\sigma_{\eta}^2}{\sigma_{\eta}^2 + \sigma_{\varepsilon}^2}$$

This is the **intraclass correlation coefficient (ICC)**. 
It is (as seen in the formula above) the proportion of the
true score variability to the total variability. 
It ranges from 0 and 1 (think about why!).

A little manipulation to improve understanding:

Let's look again at the term for the ICC above and divide the numerator and the denominator by
$\sigma_{\eta}^2$, which we can do, since it is a positive number:

$$\frac{\sigma_{\eta}^2}{\sigma_{\eta}^2 + \sigma_{\varepsilon}^2} = \frac{1}{1 + \frac{\sigma_{\varepsilon}^2}{\sigma_{\eta}^2}}$$

We could call the term $\frac{\sigma_{\varepsilon}^2}{\sigma_{\eta}^2}$ the noise-to-signal ratio.
The higher this ratio, the lower the ICC. The lower the ratio, the higher the ICC.

- If you increase the noise (measurement error $\sigma_{\varepsilon}^2$) for fixed 
  true score variability $\sigma_{\eta}^2$, the ICC decreases, because the denominator
  increases.
- If you increase the true score variability $\sigma_{\eta}^2$ for fixed noise $\sigma_{\varepsilon}^2$, 
  the ICC increases, since the denominator decreases.

Btw, we could also divide by $\sigma_{\varepsilon}^2$ and get the signal-to-noise ratio.

At first glance, the following statement seems wrong:

In a very **homogeneous population** (patients have very similar scores/measurements),
the **ICC might be very low**. The reason is that the patient variability $\sigma_{\eta}^2$ is low
and you probably have some measurement error $\sigma_{\varepsilon}^2$.
Hence, if you look at the formula, ICC must be low (for a given measurement error).

On the other hand, if you have a very **heterogeneous population** (patients have rather different 
scores/measurements), the **ICC might be very high**.
The reason is that the patient variability $\sigma_{\eta}^2$ is high and you probably 
have some measurement error $\sigma_{\varepsilon}^2$.

**What matters is the ratio of the two**, as can be seen from the formula above.

Let's try to calculate the ICC for our data using a statistical model. There are a couple of different
R packages to do this. We will use the `irr` package.

```{r}
library(irr)
irr::icc(as.matrix(df[, c("ROMas.Peter", "ROMas.Mary")]), 
    model = "oneway", type = "consistency")
```

We get the result: $ICC(1) = 0.851$, which is identical to the correlation coefficient
because there is no systematic difference between Peter and Mary.

Since we are forward looking and modern regression model experts, 
we would like to see if we can get the result using the Bayesian 
framework. 

Below is the model structure. 

\[
\begin{array}{rcl}
ROM_i &\sim& N(\mu_i, \sigma_{\varepsilon}) \\
\mu_i &=& \alpha[ID] \\
\alpha[ID] &\sim& \text{Normal}(\mu_{\alpha}, \sigma_{\alpha}) \\
\mu_{\alpha} &\sim& \text{Normal}(66, 20) \\
\sigma_{\alpha} &\sim& \text{Uniform}(0,20) \\
\sigma_{\varepsilon} &\sim& \text{Uniform}(0,20)
\end{array}
\]

**Model details:**

- $ROM_i$ is the observed ROM-score for patient $i$. 
  Every patient has two observations (one each from Mary and Peter). 
  So, for instance $i=1,2$ could be patient $ID=1$.
- $\mu_i$ is the expected value of the observed score for patient $ID$.
- $\sigma_{\varepsilon}$ is the standard deviation of the measurement error.
- $\alpha[ID]$ is the patient-specific intercept ($=\eta_i$). 
  Since every patient has a different intercept and they
  come from a normal distribution, we have a **random intercepts model**.
- $\mu_{\alpha}$ is the mean of the prior for the patient-specific intercepts.
  This is the overall mean of the scores.
- $\sigma_{\alpha}$ is the standard deviation of the patient-specific intercepts.
  **This is the patient variability**! The nice thing about presenting a model in this 
  way is that it's easier to interpret. $\sigma_{\alpha}$ says how much 
  the scores of the patients vary in relation to their respective level $\alpha[ID]$.
- The prior distributions express (as always) our prior beliefs about the parameters.

The **ICC** is then calculated as the ratio of the between-patient variance and
the total variance: 

  $$\frac{\sigma_{\alpha}^2}{\sigma_{\alpha}^2 + \sigma_{\varepsilon}^2}$$

We did not even notice it, but this was our first **multilevel regression model**.
It is multilevel due to the extra layer of patient-specific intercepts.
The **observations** are obviously **clustered within patients**, since observations
from the **same patient** are **more similar than observations from different patients**.
If one would run a normal linear regression model, one would ignore this clustering
and the assumption of independent error terms would be violated.

Draw model structure ... exercise..

This time we fire up the `rethinking` package and use the `ulam` function 
to fit the model.
This uses Markov Chain Monte Carlo (MCMC) to sample from the posterior 
distribution of the parameters.

- The `chains` argument specifies how many chains we want to run.
  A *chain* is a sequence of points in a space with as many dimensions as there 
  are parameters in the model. It jumps from one point to the next in this parameter 
  space and in doing so, visits the points of the posterior approximately in the correct 
  frequency. [Here](https://blog.revolutionanalytics.com/2013/09/an-animated-peek-into-the-workings-of-bayesian-statistics.html) 
  is an excellent visualization.

- The `cores` argument specifies how many CPU cores we want to use. 
  For larger jobs, one can try to parallelize
  the chains, which saves some time.

```{r}
library(rethinking)
library(tictoc)

df_long <- df %>% 
  mutate(ID = row_number()) %>%
  dplyr::select(ID,ROMas.Peter, ROMas.Mary) %>% 
  pivot_longer(cols = c(ROMas.Peter, ROMas.Mary), 
               names_to = "Rater", values_to = "ROM") %>% 
  mutate(Rater = factor(Rater))

tic()
m5.1 <- ulam(
  alist(
    # Likelihood
    ROM ~ dnorm(mu, sigma),
    
    # Patient-specific intercepts (random effects)
    mu <- a[ID],  
    a[ID] ~ dnorm(mu_a, sigma_ID),  # Hierarchical structure for patients
    
    # Priors for hyperparameters
    mu_a ~ dnorm(66, 20),  # Population-level mean
    sigma_ID ~ dunif(0,20),  # Between-patient standard deviation
    sigma ~ dunif(0,20)  # Residual standard deviation
  ), 
  data = df_long, 
  chains = 8, cores = 4
)
toc() # 7s

precis(m5.1, depth = 2)

post <- extract.samples(m5.1)
var_patients <- mean(post$sigma_ID^2)  # Between-patient variance
var_residual <- mean(post$sigma^2)     # Residual variance
var_patients / (var_patients + var_residual) # ICC
# 0.846
# not too bad; very close to the result from the irr package
```

In the output from `precis(m5.1, depth = 2)` above we see 

- all 50 intercept estimates for each patient: `a[ID]` 
- `mu_a`is the overall intercept.
- `sigma_ID` is the **patient variability**.
- `sigma` is the **residual variability**.

We just square the sigmas to get the variances. 

Remember: In the background, there is just a statistical model to predict
the outcome. Depending on the predictors, we get different models and 
probably different ICCs.

We can also estimate a **random intercept model** with the `lme4` package using
the command `lmer`in the Frequentist framework. No priors.

```{r}
library(lme4)
m5.2 <- lmer(ROM ~ (1|ID), data = df_long)
summary(m5.2)
print(VarCorr(m5.2), comp = "Variance")

# ICC = 
270.99 / (270.99 + 47.35) # 
# 0.8512597
# -> exactly the same result as the irr package
```

The expression `Formula: ROM ~ (1 | ID)` specifies that we want to fit a model with 
a random intercept. This means that every patient (ID) gets its own intercept
which is drawn from a normal distribution. We will probably talk about this in the
next lecture (Methodenvertiefung) in greater detail.

So far, we have only looked at the general **$ICC$** (ICC1 in the `psych`output) 
(see also page 106 in the book). 

There, we have not yet explicitely considered a bias (=systematic difference 
between the raters) that the raters could have. In the book,
they introduce a bias of 5 degrees (Mary measures 5 degrees more than Peter on average).

The model fÃ¼r ICC 2 and 3 in the `psych` output explicitely considers this
(systematic) difference that could occur between the raters. 
This results in an extra term in the denominator
of the ICC, an additional variance component.

From the same statistical model (!) we take the variance components to calculate:

$$ICC_{agreement} = \frac{\sigma_{\alpha}^2}{\sigma_{\alpha}^2 + \mathbf{\sigma_{rater}^2} + \sigma_{\varepsilon}^2}$$

where $\sigma_{rater}^2$ is the variance due to systematic rater differences.

$$ICC_{consistency} = \frac{\sigma_{\alpha}^2}{\sigma_{\alpha}^2 + \sigma_{\varepsilon}^2}$$

We will now introduce the 5 degree bias and use our Bayesian 
framework to estimate the ICC.
By introducing a bias, we should see
a lower ICC. Note, that the prediction quality of Mary's scores
given Peter's scores should not change, since we would only shift Mary's scores
down by 5 degrees, which would not disturb the linear regression model. We can
always shift the points to where we want them to be. We do that for instance
when we scale or standardize the data.

Admitted, the Bayesian version in this case takes longer and is more complex.
The advantage is still that it's fully probabilistic and one could work with 
detailed prior information, especially for smaller smaple sizes.

Anyhow, let's try to give the model equations considering
the introduced bias. This is the model for both 
$ICC_{agreement}$ and $ICC_{consistency}$!

\[
\begin{array}{rcl}
ROM_i &\sim& N(\mu_i, \sigma_{\varepsilon}) \\
\mu_i &=& \alpha[ID] + \beta[Rater] \\
\alpha[ID] &\sim& \text{Normal}(\mu_{\alpha}, \sigma_{\alpha}) \\
\beta[Rater] &\sim& \text{Normal}(0, \sigma_{\beta}) \\
\mu_{\alpha} &\sim& \text{Normal}(66, 20) \\
\sigma_{\alpha} &\sim& \text{Exp}(0.5) \\
\sigma_{\beta} &\sim& \text{Exp}(1) \\
\sigma_{\varepsilon} &\sim& \text{Exp}(1)
\end{array}
\]

As you can see, $\mu_i$ now consists of the patient-specific intercept $\alpha[ID]$
(everyone of the 50 patients gets one)
and the rater-specific effect $\beta[Rater]$ (Mary and Peter get one). 
So, in total, we have **three sources of variability**: 

- the patient variability $\sigma_{\alpha}$, 
- the rater variability $\sigma_{\beta}$,
- and the residual variability $\sigma_{\varepsilon}$.

Note, that if Peter measures each of the 50 patients twice, 
the systematic difference between Peter's measurements would 
be zero. Of course, one could be creative and think of 
a learning effect or something.

Draw model structure ... exercise..

```{r}
library(rethinking)
library(conflicted)
conflicts_prefer(posterior::sd)
df_long_bias <- df_long %>%
  mutate(ROM = ROM + ifelse(Rater == "ROMas.Mary", 5, 0))
head(df_long_bias)

set.seed(123)
m5.2 <- ulam(
  alist(
    # Likelihood
    ROM ~ dnorm(mu, sigma_eps),
    
    # Model for mean ROM with patient and rater effects
    mu <- alpha[ID] + beta[Rater],
    
    # Patient-specific random effects
    alpha[ID] ~ dnorm(mu_alpha, sigma_alpha),
    
    # Rater effect (Peter/Mary)
    beta[Rater] ~ dnorm(0, sigma_beta),
    
    # Priors for hyperparameters
    mu_alpha ~ dnorm(66, 10),  # Population mean ROM
    sigma_alpha ~ dexp(0.5),  # Between-patient SD (less aggressive shrinkage)
    sigma_beta ~ dexp(1),   # Rater SD (better regularization)
    sigma_eps ~ dexp(1)     # Residual SD (prevents over-shrinkage)
  ), 
  data = df_long_bias, 
  chains = 8, cores = 4
)

precis(m5.2, depth = 2)
precis(m5.2)

# check systematic difference for rater in posterior
post <- extract.samples(m5.2)
mean(post$beta[,1] - post$beta[,2])

# ICC agreement:
post <- extract.samples(m5.2)
(var_patients <- mean(post$sigma_alpha^2))  # Between-patient variance
(var_raters <- mean(post$sigma_beta^2))     # Rater variance
(var_residual <- mean(post$sigma_eps^2))    # Residual variance

# ICC_agreement = 
var_patients / (var_patients + var_raters + var_residual)
# 0.8033613 (sigma_alpha ~ dexp(1))
# 0.83 (sigma_alpha ~ dexp(0.5))

# ICC (Single_fixed_raters) = ICC3 in psych output = 
var_patients / (var_patients + var_residual)
# 0.8415256
```

It should be noted that this ICC is very sensitive to the choice of the prior.
If you choose too agressive priors for the standard deviations $\sigma_{\alpha}, 
\sigma_{\beta}, \sigma_{\varepsilon}$, you will get a too low ICC.

We will probably talk about this in the next lecture (Methodenvertiefung) in greater detail.
I have played around a little with the parameters in the exponential priors 
to get the desired result which compares nicely to the two alternative methods below:
using the `psych` package and with the `lmer`
package. Both use a Frequentist random intercept model in the background.
Using a package like `psych` just gives a more convenient interface to 
elicit the ICC. 

**`psych` package**:
```{r}
library(psych)
library(conflicted)
# needs wide format
#conflicts_prefer(dplyr::select)
df_wide <- df_long_bias %>%
  pivot_wider(names_from = Rater, values_from = ROM)
df_wide_values <- df_wide %>% dplyr::select(-ID)
psych::ICC(df_wide_values) # ICC1 = 0.83
```

**`lmer` package**:
```{r}
# _lmer------
m5.3 <- lmer(ROM ~ (1 | ID) + (1 | Rater), data = df_long_bias)
summary(m5.3)
print(VarCorr(m5.3), comp = "Variance")
# Groups   Name        Variance
# ID       (Intercept) 270.882 
# Rater    (Intercept)   6.193 
# Residual              47.557 


# ICC (Single_random_raters) = ICC2 in psych output
270.882 / (270.882 + 6.193 + 47.557) # 
# 0.8344279

# ICC (Single_fixed_raters) = ICC3 in psych output
270.882 / (270.882 + 47.557) #
# 0.85
```

### Explanation of ICCs in the `psych` output

If you want to know all the details, refer to [Shrout and Fleiss (1979)](https://d1wqtxts1xzle7.cloudfront.net/50483847/syarat_reliabilitas_icc-libre.pdf?1479841049=&response-content-disposition=inline%3B+filename%3DIntraclass_Correlations_Uses_in_Assessin.pdf&Expires=1740684631&Signature=hHiFbcQD3PDVIyWDJ-bUhcm3WtsK19YhHm6FKtnafNdqsm9NhR6cr9lbCf~gVV5SYG1XlTwLlcfJkQ9Z-ahjmmNV893aWi5plo~yL4oZBEjrmFa9WCd7k6vzFTkri1Xbgfh~GyPARWXBtqABytovtL-RD1420Kw9qk150nw3-kUWcuvRiIc~r0y65XQaXf-V9mm~uXRFdUqec4Vs-Bwh~IrJfHWQASGgp8wZjzh2130MCP3-iaorxNn~79c~nm2f1aIl5WRqRXB6EIy8HlrNFpxNSt1pgTPZoZadEECM4qH395KLY5ijUnhoCDT9AmcOplPnFiC5t8dKW-n25ziofQ__&Key-Pair-Id=APKAJLOHF5GGSLRBV4ZA).
The help-function `?psych::ICC` contains a relatively good and much shorter explanation.
The variance formulae given in the help-file are probably somewhat confusing.
We try to stick to the notation of the book.

Let's talk about the first three **ICCs in the `psych` output**:.

- **Single_raters_absolute ICC1:**
  According to the help file: "Each target [i.e. patient] is rated by a different 
  judge [i.e. rater] and the judges are selected at random." So, variability due to raters
  is implied and cannot be disentangled. 
  This is formally not our situation, since we have only two raters and
  50 patients. But for this case, we do not care who measures, since we do not
  model it, hence, we cannot know if there are systematic differences between
  the raters. There might as well be 50 raters doing their thing, or just 2 as in our case. 
  This is the ICC we calculated above; the overall ICC. 
  in the book and based on the following model:
  
  $$Y_{ij} = \eta_i + \varepsilon_i$$
  where $i \in {1,...,50}$ is the patient and $j \in {1,2}$ is the *measurement* 
  ($=50*2=100$ rows in long format).
  Note that we do not mention a rater here, since we do not care who took the
  measurement. It is not part of the model. The ICC is then calculated as:

  $$ICC = \frac{\sigma_{\eta}^2}{\sigma_{\eta}^2 + \sigma_{\varepsilon}^2}$$
  
  whereas we could get the variance components from either the posterior in the Bayesian
  setting or from the `lmer` output in the Frequentist setting.

- **Single_random_raters ICC2:** 
  ICC2 ($=ICC_{agreement}$ in the book) and ICC3 ($=ICC_{consistency}$ in the book) 
  are based on the **same** statistical model. The only difference is
  that ICC2 assumes that the (in our case) 2 raters are randomly selected from a larger pool of raters,
  hence, the rater variability must be explicitely considered and yields a potentially smaller
  value for the ICC. Compared to ICC1, we have repeated measurements from the same raters
  in 50 patients. That's why we can model their bias. One observation would not be enough.
  The help file says: "A random sample of k judges rate each target. 
  The measure is one of absolute agreement in the ratings."
  A random sample of k (2 in our case) judges means that we cannot rule out the variability
  due to raters (you get a variety of them and their biases are different).
  
  $$Y_{ij} = \eta_i + \beta_j + \varepsilon_i$$
  where $i \in {1,...,50}$ is the patient, $j \in {1,2}$ is the **rater** 
  (doing one measurement in each patient). The ICC is then calculated as:
  
  $$ICC_{agreement} = \frac{\sigma_{\eta}^2}{\sigma_{\eta}^2 +\mathbf{\sigma_{rater}^2} + \sigma_{\varepsilon}^2}$$

- **Single_fixed_raters ICC3:**
  ICC3 ($=ICC_{consistency}$ in the book) is based on the **same model as ICC2**, 
  but assumes that the raters are fixed. 
  This means that **the raters are the same for all patients** in the future study. 
  So, we have considered the rater variability in the model 
  (which was possible due to the repeated measurements from the same raters in 50 patients), 
  but do not care since Mary and Peter will be the people doing the future
  measurements, not other therapists. If you fix a random variable 
  (raters in this case), variance is zero. The help file says: 
  "A fixed set of k judges rate each target. 
  There is no generalization to a larger population of judges."
  The ICC is then calculated as:

  $$ICC_{consistency} = \frac{\sigma_{\eta}^2}{\sigma_{\eta}^2 + \sigma_{\varepsilon}^2}$$

  ICC1 and ICC3 are **not** identical, since ICC1 does not consider the rater variability
  in the model. They are based on *different* statistiacal models.
  ICC2 and ICC3 are based on the *same* model.

If there is no systematic difference between raters, all 3 ICCs and the Pearson 
correlation (r) are the same (see Figure 5.3 in the book).

$ICC_{consistency}$ vs. $ICC_{agreement}$:
The latter is used, when we need Peter and Mary to concur in their measurements.
Patients coming to Peters practice will get the same (or very similar) "diagnosis" (ROM-value)
from Mary. When there is systematic difference (line is shifted downwards in Figure 5.3), 
this cannot be guaranteed.
If we only need Peter and Mary to *rank* the patients in the same order,
we can use $ICC_{consistency}$.

### Summary Peter and Mary, with and without bias

Below, we summarize the results for the ICCs (calculated with `psych`)
for the unbiased and biased case (Mary measures on average 5 degrees more than peter).

```{r}
library(pacman)
p_load(conflicted, tidyverse, flextable)

# Ensure select() from dplyr is used
#conflicted::conflicts_prefer("select", "dplyr")

# Unbiased ICC Calculation
df_wide_unbiased <- df_long %>%
  pivot_wider(names_from = Rater, values_from = ROM)
df_wide_values_unbiased <- df_wide_unbiased %>% dplyr::select(-ID)
icc_results_unbiased <- psych::ICC(df_wide_values_unbiased)

# Extract relevant ICC values
icc_unbiased_df <- icc_results_unbiased$results %>%
  dplyr::select(type, ICC) %>%
  rename(`Unbiased ICC` = ICC)

# Biased ICC Calculation
df_wide_biased <- df_long_bias %>%
  pivot_wider(names_from = Rater, values_from = ROM)
df_wide_values_biased <- df_wide_biased %>% dplyr::select(-ID)
icc_results_biased <- psych::ICC(df_wide_values_biased)

# Extract relevant ICC values
icc_biased_df <- icc_results_biased$results %>%
  dplyr::select(type, ICC) %>%
  dplyr::rename(`Biased ICC` = ICC)

icc_merged_df <- left_join(icc_unbiased_df, 
                           icc_biased_df, 
                           by = "type") %>%
  slice(1:3)

ft <- flextable(icc_merged_df) %>%
  flextable::set_header_labels(type = "ICC Type") %>%
  flextable::set_caption("Intraclass Correlation Coefficients - Unbiased vs. Biased") %>%
  flextable::set_table_properties(width = .5, layout = "autofit")
ft
```

The **left column** shows that the ICCs are identical for the **unbiased case**.
Specifically, ICC2 (=$ICC_{agreement}$ in the book) and 
ICC3 ($ICC_{consistency}$ in the book) are based on the same model which explicitely 
considers a potential bias between the raters. Since there is none,
the ICCs are the same.

In the **biased case**, there *is* as systematic difference between Mary and Peter.

**ICC1** does not care about it and shows a somewhat lower value compared to 
before ($0.833$). The reason is because the agreement line is in a plot with 
Mary on Y and Peter on X shifted upwards by 5 degrees.
If you would introduce a bias of 15 degrees, the ICC would
be even lower ($ICC1 = 0.61$, $ICC2 = 0.65$ -> verify as exercise). 
The unbiased column would of course stay the same.

**ICC2** now considers the bias of 5 degrees. The model knows about the shift.
If we compare the variance components of ICC1 and ICC2, we see:
  
```{r, echo=FALSE}
print("Model for ICC1: ROM ~ (1 | ID)")
m_ICC1 <- lmer(ROM ~ (1 | ID), data = df_long_bias)
print(VarCorr(m_ICC1), comp = "Variance")

print("Model for ICC2 (and 3): ROM ~ (1 | ID) + (1 | Rater)")
m_ICC2 <- lmer(ROM ~ (1 | ID) + (1 | Rater), data = df_long_bias)
print(VarCorr(m_ICC2), comp = "Variance")
```

The residual variance is smaller in the model with the rater effect.
The model explains the data better, since it knows about the bias.

Look at the $\sigma_{\varepsilon}$ of the two models, they add up:

$$\sigma_{\varepsilon, ICC1}^2 = \sigma_{\varepsilon, ICC23}^2 + \sigma_{Rater}^2$$
$$53.75 = 47.557 + 6.193$$

Hence, we just split up the error differently by considering the bias.
The patient variability (`ID Variance` in the output) is slightly higher: 
$270.882$ compared to $267.79$ before.
In the first model, patient variability was conflated with rater variation because rater effects 
were not explicitly modeled. For this reason, `ID Variance` increases slightly.
It is a rather small increase, so ICC1 and ICC2 are not that different.

For a bias of 15 degrees, the additivity of the variances remains.
The patient variability increases from $223.89$ (ICC1) to $270.882$ (ICC2),
hence the difference in ICCs is larger ($ICC1=0.613$ vs. $ICC2=0.657$).

**ICC3** considers the bias in the model but does not include it in the measurement 
error since the raters are fixed.


### Difference between correlation and ICC

If we do not introduce a bias in the data, the correlation coefficient
is the same as the ICC (as seen above). On page 110, Figure 5.3, the authors show nicely 
what the difference is between the correlation coefficient and the ICC. 
It is also shown how $ICC_{agreement}$ and $ICC_{consistency}$ change with the bias.

$ICC_{consistency}$ stays $1$ if bias is introduced (assuming a hypothetical 
perfect agreement before). Peter and Mary still rank the patients
in the same order.

Correlation $r$ is always 1, no matter at what slope ($\ne 0$) the line is.
It measures the strength and direction of the *linear* relationship between two variables.

$ICC_{agreement}$ changes as soon as you depart from the 45 degree line
with respect to slope or 
shift the line up or down (i.e., introduce a bias).

This is a good point in time to think for a moment about the type of measurement
for agreement with respect to **costs**. The ICC below weights each measurement
equally, although one larger outlier (large discrepancy between Peter and Mary)
may influence the ICC notably (exercise later). One could easily think of a situation where 
overall agreement measure like the ICC is not adequate since, 
for instance, exceeding a certain
difference threshold could decide between life and death. 

### Bad news about the ICC?

Let's try to expand our intuitive understanding of what an ICC of 0.8 or so means.
For simplicity, we take the overall ICC, which is equal to the correlation coefficient,
if there is no bias present.

The following example demonstrates the meaning of the Test-Retest reliability of 
the Hospital Anxiety and Depression Scale - Anxiety subscale (HADS-A). We could take 
all kinds of other scores where ICC values and a 
[minimally clinically important difference (MCID)](https://en.wikipedia.org/wiki/Minimal_important_difference)
is given. Briefly, the MCID is the smallest change in a score that is considered
important to the patient. For example, a 5% change in BMI is (in some populations)
considered meaningful.

Specifically, we:

- Simulate two correlated measurement at two time points (TP1 and TP2) 
  to get predetermined ICC ($=\rho$).
- Calculate the Intraclass Correlation Coefficient (ICC).
- Compare the Minimal Clinically Important Difference (MCID) to prediction intervals.
- Visualize how often a clinically meaningful change of 1.68 points is 
  detected, even if no real change has occurred.

The code can be found [here](https://github.com/jdegenfellner/ICC_MCID/blob/main/ICC_MCID.R)
and in the github repo of the script. In the code, the sources for the ICC and MCID are
cited.

```{r}
# ICC and Test-Retest-Reliability

library(pacman)
p_load(tidyverse, lme4, conflicted, psych, MASS)

# MCID Minimal Clinically Important Difference---------

# HADS score---------
# The Hospital Anxiety and Depression Scale 

# Test-Retest-Reliability:
# https://doi.org/10.1016/S1361-9004(02)00029-8

# Use the numbers from here (Table 1):
# https://www.sciencedirect.com/science/article/abs/pii/S1361900402000298
# for demonstration purposes.

# Minimal Clinically Important Difference (MCID) for HADS-A:
# https://pmc.ncbi.nlm.nih.gov/articles/PMC2459149/
# Not exactly the same population as shift workers, but suffices for demonstration purposes.
# MCID HADS anxiety score and 1.68 (1.48â1.87)

# Simplification: Score is deemed to be continuous.

# Create 2 correlated measurements:
# Use HADS-A Anxiety subscale
# Use n=100, instead of n=24 for more stability

sigma1 <- 3.93  # Standard deviation of variable 1
sigma2 <- 3.52  # Standard deviation of variable 2
rho <- 0.82    # Correlation
cov_matrix <- matrix(c(sigma1^2, rho * sigma1 * sigma2,
                       rho * sigma1 * sigma2, sigma2^2), nrow = 2)
n <- 100

set.seed(188)  # For reproducibility
samples <- mvrnorm(n = n, mu = c(7.92, 7.83), Sigma = cov_matrix)
df <- as.data.frame(samples)
colnames(df) <- c("TP1", "TP2")

plot(df, main = "Scatterplot of Multivariate Normal Samples",
     xlab = "TP1", ylab = "TP2", pch = 19, col = rgb(0, 0, 1, alpha = 0.5))

# https://en.wikipedia.org/wiki/Intraclass_correlation#Modern_ICC_definitions:_simpler_formula_but_positive_bias
# ICC should be the correlation within the group (i.e. patient)

cor(df$TP1, df$TP2, method = "pearson") # ~0.8-0.9
ICC(df) # 0.82

# check manually
df_mod <- data.frame(id = 1:n, df$TP1, df$TP2)
names(df_mod) <- c("id", "TP1", "TP2")
df_mod_long <- df_mod %>% pivot_longer(cols = c(TP1, TP2), names_to = "time_point", values_to = "score")
mod <- lme4::lmer(score ~ time_point + (1|id), data = df_mod_long)
variance_df <- as.data.frame(summary(mod)$varcor)
# ICC=
variance_df$sdcor[1]^2 / (variance_df$sdcor[1]^2 + variance_df$sdcor[2]^2) # ~0.9
# cor and ICC are very very similar, should actually be identical.

# check
mean(df$TP1)
mean(df$TP2) 

#data.frame(TP1 = df$TP1, TP2 = df$TP2) %>% 
#  ggplot(aes(x = TP1, y = TP2)) +
#  geom_point() +
#  xlab("Measurement of patients at time point 1") +
#  ylab("Measurement of the same patients at time point 2")

# Range for HADS-A should be 0-21 
# (according to "The Hospital Anxiety and Depression Scale, Zigmond & Snaith, 1983")

df <- data.frame(TP1 = df$TP1, TP2 = df$TP2) %>% 
  dplyr::filter(TP1 >= 0) %>%
  dplyr::filter(TP2 >= 0) %>% # negative not possible
  dplyr::filter(TP1 <= 21) %>%
  dplyr::filter(TP2 <= 21) # max. score is 21
df

mod <- lm(TP2 ~ TP1, data = df)
pred <- predict(mod, df, interval = "prediction")

# How wide are the prediction intervals for a patient?
as.data.frame(pred) %>% mutate(width_prediction_interval = upr - lwr) # width of the prediction interval 8-10 points!

# Example:
predict(mod, newdata = data.frame(TP1 = 10), interval = "prediction") # 95% prediction interval for a patient with a score of 10 at time point 1.
(13.56369 - 5.226282)/1.68

# Prediction interval width is ~5 times our minimally clinically important 
# change of 1.68 for HADS-A.

df %>% 
  ggplot(aes(x = TP1, y = TP2)) +
  # Color points conditionally
  geom_point(aes(color = ifelse(TP2 > TP1 + 1.68 | TP2 < TP1 - 1.68, "red", "black"))) +
  scale_color_manual(values = c("red" = "red", "black" = "black"), guide = "none") +
  geom_abline(intercept = mod$coefficients[1], slope = mod$coefficients[2]) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_ribbon(aes(ymin = pred[,2], ymax = pred[,3]), alpha = 0.2) + 
  ggtitle("HADS-A and 95% Prediction Interval for TP2") +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_abline(intercept = 0, slope = 1, color = "green", linetype = "dashed") +
  geom_abline(intercept = 1.68, slope = 1, color = "red", linetype = "dashed") +
  geom_abline(intercept = -1.68, slope = 1, color = "red", linetype = "dashed")

# How often is TP2 within the MCIC of 1.68 points?---------
df$abs_diff <- abs(df$TP1 - df$TP2)
hist(df$abs_diff)
abline(v = 1.68, col = "red")

table(df$abs_diff > 1.68)/sum(table(df$abs_diff > 1.68)) # 
# -> in ~46% of cases we detect a minimally clinically important change of 1.68 points.
# even though the underlying truth did not change.

# This result is near random guessing
```

The example shows that for the given value of ICC ($>0.8$) and MCID, the test retest reliability 
is near random guessing with respect to detecting a change, since the second prediction is 
very often outside the MCID (red dashed lines). Or in short: Even if Mary measures twice 
with this ICC value, she will detect a clinically meaningful change in almost half of the cases, 
although the underlying
truth did not change. 

One could play around a little bit with the parameters here 
(and maybe make the argument more rigorous),
but my guess would be that the main message remains stable.

What do you think? Can you find literature or logical arguments against this
simulation? I would be pleased to hear your thoughts.


### Standard Error of Measurement (SEM)
 Above, we defined the overall ICC as:

$$ICC = \frac{\sigma_{\alpha}^2}{\sigma_{\alpha}^2 + \sigma_{\varepsilon}^2}$$

The SEM is defined as: $\sqrt{\sigma_{\varepsilon}^2}$.

This is a measure for the precision of the model. If this number is small,
we know the true but unknown score ($\eta$) very precisely. As we have seen, for the
$ICC_{agreement}$ the error term includes the rater variability $\sigma_{rater}$.

For ICC2 and 3 the model is the same and:

$$Y_{ij} = \eta_i + \beta_j + \varepsilon_i$$

For $ICC_{agreement}$, the SEM is then:
$$\sigma_{Y_i}^2 = \sigma_{\eta}^2 + \underbrace{\sigma_{\text{rater}}^2 + \sigma_{\varepsilon}^2}_{SEM_{agreement}^2}$$


For $ICC_{consistency}$, the SEM is then:
$$\sigma_{Y_i}^2 = \sigma_{\eta}^2 + \underbrace{\sigma_{\varepsilon}^2}_{SEM_{consistency}^2}$$

And since ICC 2 and 3 are based on the same model:
$$SEM_{consistency} \le SEM_{agreement}$$
(equality if there is no bias)

This yields the **first method** of getting the SEM: Estimate the 
model (as we did above) and take the square root of the respective error term
(with or without the rater variability).

One can also show (excercise...) that we can find the $SEM_{consistency}$ by using
the standard deviation of the differences between the two measurements
of Peter and Mary:

We verify this immediately:

$$SEM_{consistency} = \frac{SD_{difference}}{\sqrt{2}}$$

```{r}
mod <- lmer(ROM ~ (1 | ID) + (1 | Rater), data = df_long_bias)
print(VarCorr(mod))
```

$$SEM_{consistency} = \sigma_{residual} = 6.8961$$

Let's calculate the SEM from the differences between the two measurements:

```{r}
# SEM from differences
sd(df_long_bias$ROM[df_long_bias$Rater == "ROMas.Mary"] -
   df_long_bias$ROM[df_long_bias$Rater == "ROMas.Peter"]) / sqrt(2)
```

This is a perfect match! 

### Bland-Altman Plot

The Bland-Altman plot is a graphical method to compare two measurements
techniques or raters. The goal is to spot systematic differences between the two
methods/raters. On the x axis, we plot the average of the two measurements
and on the y axis the difference between the two measurements. So, we could 
for instance see larger variability in points with higher average values,
indicating that Peter and Mary disagree more for higher values.

Let's quickly draw one and explain it:

```{r}
library(BlandAltmanLeh)
library(data.table)
df_long_bias <- as.data.table(df_long_bias)
# https://cran.r-project.org/web/packages/BlandAltmanLeh/BlandAltmanLeh.pdf
bland.altman.plot(df_long_bias[Rater == "ROMas.Mary", ]$ROM,
                  df_long_bias[Rater == "ROMas.Peter", ]$ROM,
                  two = 1.96,
                  mode = 1,
                  conf.int = 0.95)
```

... describe.. LOA ... and so on.

## Validity
...

## TODOS

- [table 2 fallacy](https://onlinelibrary.wiley.com/doi/full/10.1111/cdoe.12617?casa_token=3IzTqt8cjzQAAAAA%3AhKH7i2yhPi9tIVbOz89JVMT2yuSuyhejGLcJqXVINfgo7kjxlkzzBhqu3iH0zW7ClZef0ivw790enq1I)
- mention missing values, missingness mechanisms -> Methodenvertiefung
- Logistic Regression, Poisson, -> Methodenvertiefung
- Exercise: Show by simulation what Gelman talks about with significant p values. So I scan the data
  for significant p values and then simulate data with the same effect size and see how often
  I get significant p values. Especially the next effect would be probably smaller,
  especially, if one did p-hacking! Calculate a priori probability for replication (def?).
- Chapter: Sample size calculations for multivariate regression, Proportions, ICCs, t.test
- Chapter about Reliability, Validity and ICCs (incl. simulation of what an ICC of 0.9 or so means), but maybe reduced
- Angenommen man hat ein masking eines Effekts und der Model fit ist aber gut (keine Voraussetzung verletzt), 
  ist diese Situation mÃ¶glich?
- What about papers? -> eLearning
- AIC, BIC, cross-validation, Model selection (best subset, leaps....), Variable selection
- More on bias variance tradeoff, show for polynomial regression?
- include eLearning tasks in script.

<!--chapter:end:04-Reliability_Validity.Rmd-->

